<!doctype html>
<html class="no-js">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

<meta property="og:title" content="Peripheral Blood Mononuclear Cells with 8 Antibodies" />
  
<meta property="og:type" content="article" />
  
<meta property="og:url" content="https://jlduan.github.io/fba/_build/html/tutorials/cell_hashing/PRJNA423077/tutorial.html" />
  
<meta property="og:description" content="Dataset: Cell hashing Stoeckius, M., Zheng, S., Houck-Loomis, B., Hao, S., Yeung, B.Z., Mauck, W.M., 3rd, Smibert, P., and Satija, R. (2018). Cell Hashing with barcoded antibodies enables multiplex..." />
  
<meta property="og:ignore_canonical" content="true" /><link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" /><link rel="next" title="MULTI-seq" href="../../multi-seq/index.html" /><link rel="prev" title="Cell hashing" href="../index.html" />

    <link rel="shortcut icon" href="../../../_static/logo.ico"/><meta name="generator" content="sphinx-4.5.0, furo 2021.10.09"/>
        <title>Peripheral Blood Mononuclear Cells with 8 Antibodies - FBA</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?digest=0254c309f5cadf746f1a613e7677379ac9c8cdcd" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?digest=16fb25fabf47304eee183a5e9af80b1ba98259b1" />
    
    


<style>
  body {
    --color-code-background: #eeeedd;
  --color-code-foreground: black;
  
  }
  body[data-theme="dark"] {
    --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
  }
  @media (prefers-color-scheme: dark) {
    body:not([data-theme="light"]) {
      --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
    }
  }
</style>


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-X0MEL2T23M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-X0MEL2T23M');
</script>

</head>
  <body>
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" />
      <line x1="4" y1="6" x2="20" y2="6" />
      <line x1="10" y1="12" x2="20" y2="12" />
      <line x1="6" y1="18" x2="20" y2="18" />
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">FBA</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo only-light" src="../../../_static/logo.svg" alt="Light Logo"/>
    <img class="sidebar-logo only-dark" src="../../../_static/logo_dark.svg" alt="Dark Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user_guide.html">User Guide</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 has-children"><a class="reference internal" href="../../crispr_screening/index.html">CRISPR screening</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/PRJNA358686/tutorial.html">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/PRJNA609688/tutorial.html">Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3' sgRNA-CR1&lt;sup&gt;cs1&lt;/sup&gt;</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_surface_protein_labeling/index.html">Cell surface protein labeling</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA393315/tutorial.html">CITE-seq; 8k Cord Blood Mononuclear Cells with 13 Antibodies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA658075/tutorial.html">ASAP-seq; Multiplexed CRISPR Perturbations in Primary T Cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/SC3_v3_NextGem_DI_PBMC_CSP_1K/tutorial.html">1k Human PBMCs Stained with a Panel of TotalSeq B Antibodies, Dual Indexed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../eccite-seq/index.html">ECCITE-seq</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../eccite-seq/PRJNA521522/tutorial.html">6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../phage-atac/index.html">PHAGE-ATAC</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../phage-atac/PRJNA661457/tutorial.html">PHAGE-ATAC; Anti-CD8 Phage Hashing Single-cell ATAC-seq Using CD8 T Cells from Four Human Donors</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cellplex/index.html">CellPlex</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Nuclei_30K_Multiplex/tutorial.html">30k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Nuclei Multiplexed, 12 CMOs</a></li>
</ul>
</li>
<li class="toctree-l2 current has-children"><a class="reference internal" href="../index.html">Cell hashing</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">Peripheral Blood Mononuclear cells with 8 Antibodies</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../multi-seq/index.html">MULTI-seq</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../multi-seq/PRJNA531855/tutorial.html">15k HEK293 and 40k HMECs Multiplexed by Lipid- and Cholesterol-tagged Indices</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../targeted_transcript_enrichment/index.html">Targeted transcript enrichment</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../targeted_transcript_enrichment/Targeted_NGSC3_DI_HodgkinsLymphoma_GeneSignature/tutorial.html">Hodgkin’s Lymphoma, Dissociated Tumor: Targeted, Gene Signature Panel</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../pseudo-bulk/index.html">Pseudo-bulk</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/jlduan/fba">GitHub Repository</a></li>
</ul>

</div>
</div>
      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="peripheral-blood-mononuclear-cells-with-8-antibodies">
<span id="tutorial-cell-hashing-prjna423077"></span><h1>Peripheral Blood Mononuclear Cells with 8 Antibodies<a class="headerlink" href="#peripheral-blood-mononuclear-cells-with-8-antibodies" title="Permalink to this headline">¶</a></h1>
<p><strong>Dataset</strong>: Cell hashing</p>
<p>Stoeckius, M., Zheng, S., Houck-Loomis, B., Hao, S., Yeung, B.Z., Mauck, W.M., 3rd, Smibert, P., and Satija, R. (2018). <a class="reference external" href="https://doi.org/10.1186/s13059-018-1603-1">Cell Hashing with barcoded antibodies enables multiplexing and doublet detection for single cell genomics</a>. <em>Genome Biol.</em> <strong>19</strong>, 224.</p>
<div class="line-block">
<div class="line"><br/></div>
</div>
<section id="preparation">
<h2>Preparation<a class="headerlink" href="#preparation" title="Permalink to this headline">¶</a></h2>
<p>Download fastq files from <a class="reference external" href="https://www.ebi.ac.uk/ena/browser/view/PRJNA423077">European Nucleotide Archive</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR828/007/SRR8281307/SRR8281307_1.fastq.gz

<span class="gp">$ </span>wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR828/007/SRR8281307/SRR8281307_2.fastq.gz
</pre></div>
</div>
<p>Download cell barcode info.</p>
<p>These are the cell-associated barcodes in this single cell RNA-Seq library (determined by the number of transcriptomic UMIs captured per barcode).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>curl -O https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2895nnn/GSM2895283/suppl/GSM2895283_Hashtag-HTO-count.csv.gz

<span class="gp">$ </span>gzip -dc GSM2895283_Hashtag-HTO-count.csv.gz <span class="p">|</span> head -1 <span class="p">|</span> sed <span class="s1">'s/,/\n/g'</span> <span class="p">|</span> grep -v <span class="s1">'^$'</span> &gt; cell_barcodes.txt
</pre></div>
</div>
<p>Inspect cell barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>head cell_barcodes.txt

<span class="go">GGCGACTAGAGGACGG</span>
<span class="go">CATCAAGGTCTTGTCC</span>
<span class="go">AAACCTGAGTGATCGG</span>
<span class="go">TGAGGGAGTACTTAGC</span>
<span class="go">CCTAAAGAGATGTGGC</span>
<span class="go">AGACGTTTCAGCCTAA</span>
<span class="go">TGGGAAGCAACACCCG</span>
<span class="go">CGATTGATCTTCGGTC</span>
<span class="go">CATCGAAGTGATGCCC</span>
<span class="go">TCAGATGCACGAGAGT</span>
</pre></div>
</div>
<p>Prepare feature barcodes (hashtag-oligos, HTO).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc GSM2895283_Hashtag-HTO-count.csv.gz <span class="p">|</span> cut -d <span class="s1">','</span> -f1 <span class="p">|</span> grep Batch <span class="p">|</span> gsed <span class="s1">'s/-/\t/g'</span> &gt; feature_barcodes.tsv
</pre></div>
</div>
<p>Inspect feature barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat feature_barcodes.tsv

<span class="go">BatchA  AGGACCATCCAA</span>
<span class="go">BatchB  ACATGTTACCGT</span>
<span class="go">BatchC  AGCTTACTATCC</span>
<span class="go">BatchD  TCGATAATGCGA</span>
<span class="go">BatchE  GAGGCTGAGCTA</span>
<span class="go">BatchF  GTGTGACGTATT</span>
<span class="go">BatchG  ACTGTCTAACGG</span>
<span class="go">BatchH  TATCACATCGGT</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="qc">
<h2>QC<a class="headerlink" href="#qc" title="Permalink to this headline">¶</a></h2>
<p>Sample the first 20,000 read pairs for quality control (set by <code class="docutils literal notranslate"><span class="pre">-n</span></code>, default <code class="docutils literal notranslate"><span class="pre">100,000</span></code>). Use <code class="docutils literal notranslate"><span class="pre">-t</span></code> to set the number of threads. The diagnostic results and plots are generated in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>). By default, full length of read 1 and read 2 are searched against reference cell and feature barcodes, respectively. The per base content of both read pairs and the distribution of matched barcode positions are summarized. Use <code class="docutils literal notranslate"><span class="pre">-r1_coords</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-r2_coords</span></code> to limit the search range. Use <code class="docutils literal notranslate"><span class="pre">-cb_n</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-fb_n</span></code> to set the mismatch tolerance for cell and feature barcode matching (default <code class="docutils literal notranslate"><span class="pre">3</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba qc <span class="se">\</span>
    -1 SRR8281307_1.fastq.gz <span class="se">\</span>
    -2 SRR8281307_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes.tsv <span class="se">\</span>
    --output_directory qc <span class="se">\</span>
    -n <span class="m">20000</span>
</pre></div>
</div>
<p>This library is constructed using Chromium Single Cell 3’ Reagent Kits (v2 Chemistry). The first 16 bases are cell barcodes and the following 10 bases are UMIs. Based on the base content plot, the GC content of cell barcodes and UMIs are quite even. Ploy-A/T tail starts at base 26.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_per_base_seq_content.png"><img alt="../../../_images/Pyplot_read1_per_base_seq_content.png" class="align-center" src="../../../_images/Pyplot_read1_per_base_seq_content.png" style="width: 350px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_barcodes_starting_ending.png"><img alt="../../../_images/Pyplot_read1_barcodes_starting_ending.png" class="align-center" src="../../../_images/Pyplot_read1_barcodes_starting_ending.png" style="width: 350px;"/></a>
<p>As for read 2, based on the per base content, it suggests that bases 0-11 are relatively GC balanced for the reads we have sampled. Starting from base 12, it is poly-A tail. Bases 0-11 are hashtag oligo sequences. Most of the reads have the correct structure.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_per_base_seq_content.png"><img alt="../../../_images/Pyplot_read2_per_base_seq_content.png" class="align-center" src="../../../_images/Pyplot_read2_per_base_seq_content.png" style="width: 800px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_barcodes_starting_ending.png"><img alt="../../../_images/Pyplot_read2_barcodes_starting_ending.png" class="align-center" src="../../../_images/Pyplot_read2_barcodes_starting_ending.png" style="width: 800px;"/></a>
<p>The detailed <code class="docutils literal notranslate"><span class="pre">qc</span></code> results are stored in <code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns indicate the matched positions on reads. <code class="docutils literal notranslate"><span class="pre">matching_description</span></code> columns indicate mismatches in substitutions:insertions:deletions format.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc qc/feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">NTCCGAACATATGAGAGCAATAGTCGTTT   CGAACATGTAAGAGAG        3:17    1:0:2   NCATGTTACCGTGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAGCAATTGTCACTTATAGGAGGAGAAGAAGGGAAGGGGGGGGGGGGGGGAAA     BatchB_ACATGTTACCGT     0:12    1:0:0</span>
<span class="go">NAACGGATCCACGAATGAAGGACGCCTTT   TACGGTATCCACGAAT        1:16    1:0:1   NNGNNAATGCGAGAAAAAAAAAAAAAAAAAAAAAAAAAAAAGGGGCGCTCTCTTCGGGGGGGCGGGGAGAGCGAAGGAGGGGGGGGGGGGGGGAAGGAG     no_match        NA      NA</span>
<span class="go">NGGCCAGTCTTCAACTGTTAACACTATTT   GTCCTCAAGCTGTCTA        6:20    1:0:2   NNNNNNNNNNNNNAAANNAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGGTTTAAAAAGTGAAAGAGGGACAAAACGGGAAAAACGGGGGTGGGGAAAA     no_match        NA      NA</span>
<span class="go">NATCCAGCAATACGCTTTCCACGACATTT   ATCCACCCATACGCTA        1:17    3:0:0   NNNNNNNNNNNNNAAANNAAAAAAAAAAAAAAAAAAAAAGTGGGGGGAAAGCGGTTTTGGGAGATAAAACGAAAAAGCGGCGGGGGGGGAAAAAGGTGA     no_match        NA      NA</span>
<span class="go">NTGCGATAGACACTAAGAGGAGTTCATTT   CGCGGTAAGACACTAA        1:16    2:0:1   NCGATAATGCGACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACCCCCTTTGTTTTTATCGTAAAGATGGGAAGGGGGCGGTGGAGGGAAAAA     BatchD_TCGATAATGCGA     0:12    1:0:0</span>
<span class="go">NTGATCCAGAAGGTGAGGGAGGCTGATTT   AGATTGCGTGAGGGAG        7:21    1:0:2   NNNNNNNNNNNNNNAANNAAAAAAAAAAAAAAAAAATCACCCCCCCCCCCCTTTTGGTTCAAAAACGGAAAAAGCGCCGCGGGGGGAAAGAGTGTAAAT     no_match        NA      NA</span>
<span class="go">NTGGGTCAGGCCGAATTGAAGGGATGTTT   GAAATGAAGTGAAGTT        12:28   3:0:0   NNNNNNCTATCCAAAANNAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACCCCTTCAATTGGCCCAGACCCAACACTCGAAGGGCCGGCTGGCAGCAAA     no_match        NA      NA</span>
<span class="go">NGAGAAGTCTCGATGAATCTAGCCGCTTT   CGATTGAAGCTAGCCC        10:25   2:0:1   NNNNNNNNNCTNCAAANNAAAAAAAAAAAAAAATAAAAAAAACGGGCTGATCCCAAGCAGACGTCACAAAGAAGCGAGAGAGTGGGATTGAGAAAAAGA     no_match        NA      NA</span>
<span class="go">NCACGGAGTTCCCTTGCCAATGTAGTTTT   AGGGAGTTCGTTTGCC        2:18    3:0:0   NGCTTACTATCCTAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATATGGGGGGGGGGAATCGGGGGGGAGGGGAAAGGGGGGGTGGGGGAAAAAAGA     BatchC_AGCTTACTATCC     0:12    1:0:0</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="barcode-extraction">
<h2>Barcode extraction<a class="headerlink" href="#barcode-extraction" title="Permalink to this headline">¶</a></h2>
<p>The lengths of cell and feature barcodes (hashtags) are all identical (16 and 12, respectively). And based on the <code class="docutils literal notranslate"><span class="pre">qc</span></code> results, the distributions of starting and ending positions of cell and feature barcodes are very uniform. Search ranges are set to <code class="docutils literal notranslate"><span class="pre">0,16</span></code> on read 1 and <code class="docutils literal notranslate"><span class="pre">0,12</span></code> on read 2. One mismatch for cell and feature barcodes (<code class="docutils literal notranslate"><span class="pre">-cb_m</span></code>, <code class="docutils literal notranslate"><span class="pre">-cf_m</span></code>) are allowed. By default, three ambiguous nucleotides (Ns) for read 1 and read2 (<code class="docutils literal notranslate"><span class="pre">-cb_n</span></code>, <code class="docutils literal notranslate"><span class="pre">-cf_n</span></code>) are allowed.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba extract <span class="se">\</span>
    -1 SRR8281307_1.fastq.gz <span class="se">\</span>
    -2 SRR8281307_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes.tsv <span class="se">\</span>
    -o feature_barcoding_output.tsv.gz <span class="se">\</span>
    -r1_coords <span class="m">0</span>,16 <span class="se">\</span>
    -r2_coords <span class="m">0</span>,12 <span class="se">\</span>
    -cb_m <span class="m">1</span> <span class="se">\</span>
    -fb_m <span class="m">1</span> <span class="se">\</span>
    -cb_n <span class="m">3</span> <span class="se">\</span>
    -fb_n <span class="m">3</span>
</pre></div>
</div>
<p>Preview of result.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_num_mismatches       read2_seq       feature_barcode fb_num_mismatches</span>
<span class="go">NTCCGAACATATGAGAgcaatagtcgttt   ATCCGAACATATGAGA        1       NCATGTTACCGTgaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaacagcaattgtcacttataggaggagaagaagggaagggggggggggggggaaa    BatchB_ACATGTTACCGT     1</span>
<span class="go">NTGCGATAGACACTAAgaggagttcattt   ATGCGATAGACACTAA        1       NCGATAATGCGAcaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaccccctttgtttttatcgtaaagatgggaagggggcggtggagggaaaaa    BatchD_TCGATAATGCGA     1</span>
<span class="go">NCACGGAGTTCCCTTGccaatgtagtttt   CCACGGAGTTCCCTTG        1       NGCTTACTATCCtaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaatatggggggggggaatcgggggggaggggaaagggggggtgggggaaaaaaga    BatchC_AGCTTACTATCC     1</span>
<span class="go">NGGGATGCAGCTTAACcgggcatcgcttt   AGGGATGCAGCTTAAC        1       NCATGTTACCGTcaaaaaaaaaaaaaaaaaaaaaaaaaaaaaatgaaatggaagtaggggtgtccctagtctgtagaagcggcgactggggaaatgtat    BatchB_ACATGTTACCGT     1</span>
<span class="go">NTTGTCACATACGCTAcgagcctgcattt   TTTGTCACATACGCTA        1       NATCACATCGGTtaaaaaaaaaaaaaaaaaaaaaaaaaaaagaaggccggggggggggggaaaaaaaaaaaaaaaaagggcggggtggggagagagtga    BatchH_TATCACATCGGT     1</span>
<span class="go">NGCTCTCGTTCCACGGaggttatcggttt   AGCTCTCGTTCCACGG        1       NCTGTCTAACGGgaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaacccccggggaggggaaaaaaagcaggaaaagcgccatgggggaaaaaaaaa    BatchG_ACTGTCTAACGG     1</span>
<span class="go">GATCTAGCAATGTTGCcaaccattttttt   GATCTAGCAATGTTGC        0       AGGACCATCCAAgaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaagatggaggaacttggttagaacagaaggaggaggggtggggggggaa    BatchA_AGGACCATCCAA     0</span>
<span class="go">NTTGCGCCATGGTCATagtaacaagattt   TTTGCGCCATGGTCAT        1       NCATGTTACCGTcaaaaaaaaaaaaaaaaaaaaaaaaaaaaatctttttcttttgccctgggcgaaaaagatgggaggagggggggggggggaaagggt    BatchB_ACATGTTACCGT     1</span>
<span class="go">CGCGGTAAGACACTAAcggccgtggtttt   CGCGGTAAGACACTAA        0       TATCACATCGGTtaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaacccgggcgggtggggttttacgaggaaggggagcagggggggtggaggaaaaaaa    BatchH_TATCACATCGGT     0</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>91.5% (67,916,430 out of 74,219,921) of total read pairs have valid cell and feature barcodes. Majority of the fragments in this library have the correct structure.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-02-17 16:16:13,003 - fba.__main__ - INFO - fba version: 0.0.7</span>
<span class="go">2021-02-17 16:16:13,003 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-02-17 16:16:13,003 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2021-02-17 16:16:13,003 - fba.__main__ - INFO - Using extract subcommand ...</span>
<span class="go">2021-02-17 16:16:13,026 - fba.levenshtein - INFO - Number of reference cell barcodes: 65,000</span>
<span class="go">2021-02-17 16:16:13,027 - fba.levenshtein - INFO - Number of reference feature barcodes: 8</span>
<span class="go">2021-02-17 16:16:13,027 - fba.levenshtein - INFO - Read 1 coordinates to search: [0, 16)</span>
<span class="go">2021-02-17 16:16:13,027 - fba.levenshtein - INFO - Read 2 coordinates to search: [0, 12)</span>
<span class="go">2021-02-17 16:16:13,027 - fba.levenshtein - INFO - Cell barcode maximum number of mismatches: 1</span>
<span class="go">2021-02-17 16:16:13,027 - fba.levenshtein - INFO - Feature barcode maximum number of mismatches: 1</span>
<span class="go">2021-02-17 16:16:13,027 - fba.levenshtein - INFO - Read 1 maximum number of N allowed: 3</span>
<span class="go">2021-02-17 16:16:13,027 - fba.levenshtein - INFO - Read 2 maximum number of N allowed: 3</span>
<span class="go">2021-02-17 16:16:15,500 - fba.levenshtein - INFO - Matching ...</span>
<span class="go">2021-02-17 16:28:20,306 - fba.levenshtein - INFO - Read pairs processed: 10,000,000</span>
<span class="go">2021-02-17 16:40:24,344 - fba.levenshtein - INFO - Read pairs processed: 20,000,000</span>
<span class="go">2021-02-17 16:52:14,506 - fba.levenshtein - INFO - Read pairs processed: 30,000,000</span>
<span class="go">2021-02-17 17:04:04,292 - fba.levenshtein - INFO - Read pairs processed: 40,000,000</span>
<span class="go">2021-02-17 17:15:52,792 - fba.levenshtein - INFO - Read pairs processed: 50,000,000</span>
<span class="go">2021-02-17 17:27:43,975 - fba.levenshtein - INFO - Read pairs processed: 60,000,000</span>
<span class="go">2021-02-17 17:39:35,941 - fba.levenshtein - INFO - Read pairs processed: 70,000,000</span>
<span class="go">2021-02-17 17:44:36,162 - fba.levenshtein - INFO - Number of read pairs processed: 74,219,921</span>
<span class="go">2021-02-17 17:44:36,162 - fba.levenshtein - INFO - Number of read pairs w/ valid barcodes: 67,916,430</span>
<span class="go">2021-02-17 17:44:36,264 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="matrix-generation">
<h2>Matrix generation<a class="headerlink" href="#matrix-generation" title="Permalink to this headline">¶</a></h2>
<p>Only fragments with valid (passed the criteria) cell and feature barcodes are included. UMI deduplication is powered by UMI-tools (<a class="reference external" href="http://www.genome.org/cgi/doi/10.1101/gr.209601.116">Smith, T., et al. 2017. Genome Res. 27, 491–499.</a>). Use <code class="docutils literal notranslate"><span class="pre">-us</span></code> to set the UMI starting position on read 1 (default <code class="docutils literal notranslate"><span class="pre">16</span></code>). Use <code class="docutils literal notranslate"><span class="pre">-ul</span></code> to set the UMI length (default <code class="docutils literal notranslate"><span class="pre">12</span></code>). Fragments with UMI length less than this value are discarded. Use <code class="docutils literal notranslate"><span class="pre">-um</span></code> to set mismatch threshold (default <code class="docutils literal notranslate"><span class="pre">1</span></code>). UMI deduplication method is set by <code class="docutils literal notranslate"><span class="pre">-ud</span></code> (default <code class="docutils literal notranslate"><span class="pre">directional</span></code>).</p>
<p>The generated feature count matrix can be easily imported into well-established single cell analysis packages: <a class="reference external" href="https://satijalab.org/seurat/">Seruat</a> and <a class="reference external" href="https://scanpy.readthedocs.io/en/stable">Scanpy</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba count <span class="se">\</span>
    -i feature_barcoding_output.tsv.gz <span class="se">\</span>
    -o matrix_featurecount.csv.gz <span class="se">\</span>
    -us <span class="m">16</span> <span class="se">\</span>
    -ul <span class="m">10</span> <span class="se">\</span>
    -um <span class="m">1</span> <span class="se">\</span>
    -ud directional
</pre></div>
</div>
<p>Result summary.</p>
<p>25.1% (17,022,991 out of 67,916,430) of read pairs with valid cell and feature barcodes are unique fragments. 22.9% (17,022,991 out of 74,219,921) of total sequenced read pairs contribute to the final matrix.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-02-17 17:44:43,315 - fba.__main__ - INFO - fba version: 0.0.7</span>
<span class="go">2021-02-17 17:44:43,315 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-02-17 17:44:43,315 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2021-02-17 17:44:43,315 - fba.__main__ - INFO - Using count subcommand ...</span>
<span class="go">2021-02-17 17:44:43,315 - fba.count - INFO - UMI-tools version: 1.1.1</span>
<span class="go">2021-02-17 17:44:43,318 - fba.count - INFO - UMI starting position on read 1: 16</span>
<span class="go">2021-02-17 17:44:43,318 - fba.count - INFO - UMI length: 10</span>
<span class="go">2021-02-17 17:44:43,318 - fba.count - INFO - UMI-tools deduplication threshold: 1</span>
<span class="go">2021-02-17 17:44:43,318 - fba.count - INFO - UMI-tools deduplication method: directional</span>
<span class="go">2021-02-17 17:44:43,318 - fba.count - INFO - Header line: read1_seq cell_barcode cb_num_mismatches read2_seq feature_barcode fb_num_mismatches</span>
<span class="go">2021-02-17 17:48:32,866 - fba.count - INFO - Number of lines processed: 67,916,430</span>
<span class="go">2021-02-17 17:48:33,127 - fba.count - INFO - Number of cell barcodes detected: 64,998</span>
<span class="go">2021-02-17 17:48:33,127 - fba.count - INFO - Number of features detected: 8</span>
<span class="go">2021-02-17 18:01:15,176 - fba.count - INFO - Total UMIs after deduplication: 17,022,991</span>
<span class="go">2021-02-17 18:01:15,298 - fba.count - INFO - Median number of UMIs per cell: 63.0</span>
<span class="go">2021-02-17 18:01:16,924 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="demultiplexing">
<h2>Demultiplexing<a class="headerlink" href="#demultiplexing" title="Permalink to this headline">¶</a></h2>
<section id="negative-binomial-distribution">
<h3>Negative binomial distribution<a class="headerlink" href="#negative-binomial-distribution" title="Permalink to this headline">¶</a></h3>
<p>Cells are classified based on the abundance of features (HTOs, no transcriptome information used). Demultiplexing method <code class="docutils literal notranslate"><span class="pre">1</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is implemented based on the method described in <a class="reference external" href="https://doi.org/10.1186/s13059-018-1603-1">Stoeckius, M., et al. (2018)</a> with some modifications. A cell identity matrix is generated in the output directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>, default <code class="docutils literal notranslate"><span class="pre">demultiplexed</span></code>): 0 means negative, 1 means positive. Use <code class="docutils literal notranslate"><span class="pre">-q</span></code> to set the quantile threshold for demulitplexing (Default <code class="docutils literal notranslate"><span class="pre">0.9999</span></code>). Set <code class="docutils literal notranslate"><span class="pre">-v</span></code> to create visualization plots.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    --output_directory demultiplexed <span class="se">\</span>
    -dm <span class="m">1</span> <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-02-18 01:27:19,172 - fba.__main__ - INFO - fba version: 0.0.7</span>
<span class="go">2021-02-18 01:27:19,173 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-02-18 01:27:19,173 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2021-02-18 01:27:19,173 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2021-02-18 01:27:19,177 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2021-02-18 01:27:19,177 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2021-02-18 01:27:22,932 - fba.demultiplex - INFO - Number of cells: 64,998</span>
<span class="go">2021-02-18 01:27:22,932 - fba.demultiplex - INFO - Number of features: 8</span>
<span class="go">2021-02-18 01:27:22,932 - fba.demultiplex - INFO - Total UMIs: 17,021,991</span>
<span class="go">2021-02-18 01:27:23,029 - fba.demultiplex - INFO - Median number of UMIs per cell: 63.0</span>
<span class="go">2021-02-18 01:27:23,029 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2021-02-18 03:19:27,245 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2021-02-18 03:20:37,827 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2021-02-18 03:21:21,120 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (HTOs) across all cells. Each column represents a single cell. This is a re-creation of <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1603-1/figures/1">Fig. 1c</a> in <a class="reference external" href="https://doi.org/10.1186/s13059-018-1603-1">Stoeckius, M., et al. (2018)</a>.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed.png" style="width: 700px;"/></a>
<p>t-SNE embedding of cells based on the abundance of features (HTOs, no transcriptome information used). Colors indicate the HTO status for each cell, as called by FBA. This is a re-creation of <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1603-1/figures/1">Fig. 1d</a> in <a class="reference external" href="https://doi.org/10.1186/s13059-018-1603-1">Stoeckius, M., et al. (2018)</a>.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed.png"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed.png" style="width: 500px;"/></a>
<p>Preview the demultiplexing result: the numbers of singlets. The result in <a class="reference external" href="https://doi.org/10.1186/s13059-018-1603-1">Stoeckius, M., et al. (2018)</a> can be found in <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1603-1#Sec26">Additional file 3</a>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"demultiplexed/matrix_cell_identity.csv.gz"</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
<span class="n">BatchA</span>    <span class="mi">2637</span>
<span class="n">BatchB</span>    <span class="mi">3019</span>
<span class="n">BatchC</span>    <span class="mi">2666</span>
<span class="n">BatchD</span>    <span class="mi">2441</span>
<span class="n">BatchE</span>    <span class="mi">2242</span>
<span class="n">BatchF</span>    <span class="mi">2234</span>
<span class="n">BatchG</span>    <span class="mi">2747</span>
<span class="n">BatchH</span>    <span class="mi">2719</span>
<span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="gaussian-mixture-model">
<h3>Gaussian mixture model<a class="headerlink" href="#gaussian-mixture-model" title="Permalink to this headline">¶</a></h3>
<p>Alternatively, cells can be demultiplexed using gaussian mixture model. The implementation of demultiplexing method <code class="docutils literal notranslate"><span class="pre">2</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is inspired by the method described on <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/crispr">10x Genomics’ website</a>. Use <code class="docutils literal notranslate"><span class="pre">-p</span></code> to set the probability threshold for demulitplexing (default <code class="docutils literal notranslate"><span class="pre">0.9</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    -dm <span class="m">2</span> <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-12-27 11:37:31,026 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2021-12-27 11:37:31,026 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-12-27 11:37:31,026 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2021-12-27 11:37:31,026 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2021-12-27 11:37:33,496 - fba.__main__ - INFO - Skipping arguments: "-q/--quantile", "-cm/--clustering_method"</span>
<span class="go">2021-12-27 11:37:33,496 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2021-12-27 11:37:33,496 - fba.demultiplex - INFO - Demultiplexing method: 2</span>
<span class="go">2021-12-27 11:37:33,496 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2021-12-27 11:37:33,496 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2021-12-27 11:37:33,496 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2021-12-27 11:37:33,496 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2021-12-27 11:37:34,111 - fba.demultiplex - INFO - Number of cells: 64,998</span>
<span class="go">2021-12-27 11:37:34,111 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2021-12-27 11:37:34,205 - fba.demultiplex - INFO - Number of features: 8 / 8 (after filtering / original in the matrix)</span>
<span class="go">2021-12-27 11:37:34,205 - fba.demultiplex - INFO - Features: BatchA BatchB BatchC BatchD BatchE BatchF BatchG BatchH</span>
<span class="go">2021-12-27 11:37:34,206 - fba.demultiplex - INFO - Total UMIs: 17,021,991 / 17,021,991</span>
<span class="go">2021-12-27 11:37:34,254 - fba.demultiplex - INFO - Median number of UMIs per cell: 63.0 / 63.0</span>
<span class="go">2021-12-27 11:37:34,254 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2021-12-27 11:37:48,810 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2021-12-27 11:38:10,642 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2021-12-27 11:38:54,942 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (HTOs) across all cells. Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_gm.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_gm.png" style="width: 700px;"/></a>
<p>t-SNE embedding of cells based on the abundance of features (HTOs, no transcriptome information used). Colors indicate the HTO status for each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed_gm.png"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed_gm.png" style="width: 500px;"/></a>
<p>Preview the demultiplexing result: the numbers of singlets.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"demultiplexed/matrix_cell_identity.csv.gz"</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
<span class="n">BatchA</span>    <span class="mi">2618</span>
<span class="n">BatchB</span>    <span class="mi">2979</span>
<span class="n">BatchC</span>    <span class="mi">2648</span>
<span class="n">BatchD</span>    <span class="mi">2368</span>
<span class="n">BatchE</span>    <span class="mi">2198</span>
<span class="n">BatchF</span>    <span class="mi">2201</span>
<span class="n">BatchG</span>    <span class="mi">2810</span>
<span class="n">BatchH</span>    <span class="mi">2721</span>
<span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../../multi-seq/index.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">MULTI-seq</div>
              </div>
              <svg><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../index.html">
              <svg><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Cell hashing</div>
                
              </div>
            </a>
        </div>

        <div class="related-information">
              Copyright &#169; 2020-2022, Jialei Duan
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            Contents
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Peripheral Blood Mononuclear Cells with 8 Antibodies</a><ul>
<li><a class="reference internal" href="#preparation">Preparation</a></li>
<li><a class="reference internal" href="#qc">QC</a></li>
<li><a class="reference internal" href="#barcode-extraction">Barcode extraction</a></li>
<li><a class="reference internal" href="#matrix-generation">Matrix generation</a></li>
<li><a class="reference internal" href="#demultiplexing">Demultiplexing</a><ul>
<li><a class="reference internal" href="#negative-binomial-distribution">Negative binomial distribution</a></li>
<li><a class="reference internal" href="#gaussian-mixture-model">Gaussian mixture model</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/scripts/main.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    </body>
</html>