<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />
<meta property="og:title" content="Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3’ sgRNA-CR1cs1" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://jlduan.github.io/fba/_build/html/tutorials/crispr_screening/PRJNA609688/tutorial.html" />
<meta property="og:site_name" content="fba" />
<meta property="og:description" content="Dataset: Direct-capture Perturb-seq ( sgRNA-CR1 cs1) Replogle, J.M., Norman, T.M., Xu, A., Hussmann, J.A., Chen, J., Cogan, J.Z., Meer, E.J., Terry, J.M., Riordan, D.P., Srinivas, N., et al. (2020). Combinatorial single-cell CRISPR screens by direct guide RNA capture and targeted sequencing. Nat...." />
<meta property="og:image" content="https://raw.githubusercontent.com/jlduan/fba/gh-pages/docs/_static/logo.svg" />
<meta property="og:image:alt" content="fba" />
<meta name="description" content="Dataset: Direct-capture Perturb-seq ( sgRNA-CR1 cs1) Replogle, J.M., Norman, T.M., Xu, A., Hussmann, J.A., Chen, J., Cogan, J.Z., Meer, E.J., Terry, J.M., Riordan, D.P., Srinivas, N., et al. (2020). Combinatorial single-cell CRISPR screens by direct guide RNA capture and targeted sequencing. Nat...." />
<meta property="og:ignore_canonical" content="true" />
<meta property="article:author" content="Jialei Duan" />
<meta property="og:site_name" content="FBA" />
<meta name="twitter:card" content="summary"></meta>
<meta name="twitter:creator" content="@jlduan" />
<meta name="twitter:image" content="https://github.com/jlduan.png" />
<link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" /><link rel="next" title="Cell surface protein labeling" href="../../cell_surface_protein_labeling/index.html" /><link rel="prev" title="CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)" href="../PRJNA358686/tutorial.html" />

    <link rel="shortcut icon" href="../../../_static/logo.ico"/><!-- Generated with Sphinx 5.3.0 and Furo 2023.03.27 -->
        <title>Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3’ sgRNA-CR1cs1 - FBA</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?digest=fad236701ea90a88636c2a8c73b44ae642ed2a53" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #eeeedd;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
      }
    }
  }
</style>


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-X0MEL2T23M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-X0MEL2T23M');
</script>

</head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">FBA</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo only-light" src="../../../_static/logo.svg" alt="Light Logo"/>
    <img class="sidebar-logo only-dark" src="../../../_static/logo_dark.svg" alt="Dark Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user_guide.html">User Guide</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current has-children"><a class="reference internal" href="../index.html">CRISPR screening</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../PRJNA358686/tutorial.html">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a></li>
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3' sgRNA-CR1&lt;sup&gt;cs1&lt;/sup&gt;</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_surface_protein_labeling/index.html">Cell surface protein labeling</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA393315/tutorial.html">CITE-seq; 8k Cord Blood Mononuclear Cells with 13 Antibodies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA658075/tutorial.html">ASAP-seq; Multiplexed CRISPR Perturbations in Primary T Cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/SC3_v3_NextGem_DI_PBMC_CSP_1K/tutorial.html">1k Human PBMCs Stained with a Panel of TotalSeq B Antibodies, Dual Indexed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../eccite-seq/index.html">ECCITE-seq</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../eccite-seq/PRJNA521522/tutorial.html">6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../phage-atac/index.html">PHAGE-ATAC</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../phage-atac/PRJNA661457/tutorial.html">PHAGE-ATAC; Anti-CD8 Phage Hashing Single-cell ATAC-seq Using CD8 T Cells from Four Human Donors</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cellplex/index.html">CellPlex</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Nuclei_30K_Multiplex/tutorial.html">30k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Nuclei Multiplexed, 12 CMOs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_hashing/index.html">Cell hashing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_hashing/PRJNA423077/tutorial.html">Peripheral Blood Mononuclear cells with 8 Antibodies</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../multi-seq/index.html">MULTI-seq</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../multi-seq/PRJNA531855/tutorial.html">15k HEK293 and 40k HMECs Multiplexed by Lipid- and Cholesterol-tagged Indices</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../targeted_transcript_enrichment/index.html">Targeted transcript enrichment</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../targeted_transcript_enrichment/Targeted_NGSC3_DI_HodgkinsLymphoma_GeneSignature/tutorial.html">Hodgkin’s Lymphoma, Dissociated Tumor: Targeted, Gene Signature Panel</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../pseudo-bulk/index.html">Pseudo-bulk</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/jlduan/fba">GitHub Repository</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/jlduan/fba/edit/gh-pages/docs/tutorials/crispr_screening/PRJNA609688/tutorial.rst" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="direct-capture-perturb-seq-crispri-based-screen-of-unfolded-protein-response-upr-using-3-sgrna-cr1cs1">
<span id="tutorial-crispr-screening-prjna609688"></span><h1>Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3’ sgRNA-CR1<sup>cs1</sup><a class="headerlink" href="#direct-capture-perturb-seq-crispri-based-screen-of-unfolded-protein-response-upr-using-3-sgrna-cr1cs1" title="Permalink to this heading">#</a></h1>
<p><strong>Dataset</strong>: Direct-capture Perturb-seq (sgRNA-CR1<sup>cs1</sup>)</p>
<p>Replogle, J.M., Norman, T.M., Xu, A., Hussmann, J.A., Chen, J., Cogan,
J.Z., Meer, E.J., Terry, J.M., Riordan, D.P., Srinivas, N., et al.
(2020). <a class="reference external" href="https://doi.org/10.1038/s41587-020-0470-y">Combinatorial single-cell CRISPR screens by direct guide RNA
capture and targeted sequencing</a>. <em>Nat. Biotechnol.</em> <strong>38</strong>, 954–961.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="preparation">
<h2>Preparation<a class="headerlink" href="#preparation" title="Permalink to this heading">#</a></h2>
<p>Download fastq files from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4367980">Gene Expression Omnibus</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat SRR11214033_2.fastq.gz SRR11214034_2.fastq.gz SRR11214035_2.fastq.gz SRR11214036_2.fastq.gz &gt; GSM4367980_1.fq.gz

<span class="gp">$ </span>cat SRR11214033_3.fastq.gz SRR11214034_3.fastq.gz SRR11214035_3.fastq.gz SRR11214036_3.fastq.gz &gt; GSM4367980_2.fq.gz
</pre></div>
</div>
<p>Prepare cell barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4367nnn/GSM4367979/suppl/GSM4367979_exp1-5.barcodes.tsv.gz

<span class="gp">$ </span>gzip -dc GSM4367979_exp1-5.barcodes.tsv.gz <span class="p">|</span> grep <span class="s1">&#39;\-4&#39;</span> &gt; cell_barcodes.txt
</pre></div>
</div>
<p>Inspect cell barcodes (<a class="reference external" href="https://www.nature.com/articles/s41587-020-0470-y/figures/1">Fig. 2d</a>, middle column).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>wc -l cell_barcodes.txt

<span class="go">8727</span>

<span class="gp">$ </span>head cell_barcodes.txt

<span class="go">AAACCCAAGAGGGCGA-4</span>
<span class="go">AAACCCACACCCTCTA-4</span>
<span class="go">AAACCCACATAGATGA-4</span>
<span class="go">AAACCCACATATCGGT-4</span>
<span class="go">AAACCCATCATGAGTC-4</span>
<span class="go">AAACCCATCCGGTAAT-4</span>
<span class="go">AAACCCATCCTGCCAT-4</span>
<span class="go">AAACCCATCTCACCCA-4</span>
<span class="go">AAACGAAAGCCTGACC-4</span>
<span class="go">AAACGAAAGTGCCGAA-4</span>
</pre></div>
</div>
<p>Prepare feature barcodes. sgRNA sequences can be found in <a class="reference external" href="https://www.nature.com/articles/s41587-020-0470-y#Sec18">Supplementary
Table 2</a> and are truncated to equal length.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat feature_barcodes_UPR_edited.tsv

<span class="go">NegCtrl2        GCGATGGGGGGGTGGGTAG</span>
<span class="go">NegCtrl3        GACGACTAGTTAGGCGTGT</span>
<span class="go">DDRGK1  GCGGTCCACAAAGGCTCAG</span>
<span class="go">UFL1    GTGACTCGCAGTAGACGCG</span>
<span class="go">UFM1    GCGGTAAGCAAACACTTAC</span>
<span class="go">SEC61G  GCTCCAGTGCTACGTGTCC</span>
<span class="go">SEC61A1 GCTGTGCAGTGGAACGCGC</span>
<span class="go">SRP68   GAGAAGCAGGTCCCAGGCG</span>
<span class="go">SRPRB   GGCCACCCGGCGCGAGTCC</span>
<span class="go">SRPR    GGCGAACGCGGCCTGAATT</span>
<span class="go">SRP72   GCCTCCAAGATGGCGAGCG</span>
<span class="go">TIMM23  GAAGTAGGCGCTGGCAACG</span>
<span class="go">ATP5B   GAGTCTCCGCAAGGCCCCG</span>
<span class="go">MRPL39  GTCTGGCTGGTCGCACCCG</span>
<span class="go">TMED2   GTGAGGCCGAAGCCAGGAC</span>
<span class="go">TMED10  GAGACTCGTTCACCACCGA</span>
<span class="go">CARS    GAGCCATGGCAGATTCCTC</span>
<span class="go">HARS    GCTCAAGTGGACAGCCGGG</span>
<span class="go">QARS    GCGCGCTCAGTGAGAGGAA</span>
<span class="go">TTI1    GAAGGCTGGAAGACGAGGT</span>
<span class="go">TELO2   GCCGCGGAGACCCGCCCCA</span>
<span class="go">TTI2    GTCCGGATCCTGTTAGACA</span>
<span class="go">TMEM167A        GCAGCCACATCACCCTTCC</span>
<span class="go">YIPF5   GGGTGCAGGGGACCGCGTC</span>
<span class="go">SCYL1   GGCCGGAGGACCCGGAGCT</span>
<span class="go">IER3IP1 GGGGCCCCATCGGCTTCCG</span>
<span class="go">DDOST   GTGGGTCCTTCGGCAGGAG</span>
<span class="go">DAD1    GACCTTGCGTGCAGTTATG</span>
<span class="go">OST4    GGCTTGTTCGCTGGTGGCG</span>
<span class="go">EIF2B4  GCTGAGGGCGATGGCTGCT</span>
<span class="go">EIF2B2  GTAGCTGCCTTCAGCCTTC</span>
<span class="go">EIF2B3  GCCATTGGGCTGTCAGTCA</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p>First we screen reads that have the constant sequence (<code class="docutils literal notranslate"><span class="pre">GTACATGGGG</span></code>)
upstream of sgRNAs on read 2 (<a class="reference external" href="https://github.com/marcelm/cutadapt">cutadapt</a>, version 3.7).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cutadapt <span class="se">\</span>
    --cores <span class="m">0</span> <span class="se">\</span>
    --front GTACATGGGG <span class="se">\</span>
    --length <span class="m">50</span> <span class="se">\</span>
    --minimum-length <span class="m">50</span>:26 <span class="se">\</span>
    --trimmed-only <span class="se">\</span>
    --output read_2_trimmed.fq.gz --paired-output read_1_trimmed.fq.gz <span class="se">\</span>
    GSM4367980_2.fq.gz GSM4367980_1.fq.gz
</pre></div>
</div>
<p>Preview the filtering result: 104,375,315 out of 404,963,129 (25.8%)
read pairs are kept for sgRNA identification.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">== Read fate breakdown ==</span>
<span class="go">Pairs that were too short:             377,332 (0.1%)</span>
<span class="go">Pairs discarded as untrimmed:      300,210,482 (74.1%)</span>
<span class="go">Pairs written (passing filters):   104,375,315 (25.8%)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="qc">
<h2>QC<a class="headerlink" href="#qc" title="Permalink to this heading">#</a></h2>
<p>The first 20,000 read pairs are sampled (set by <code class="docutils literal notranslate"><span class="pre">-n</span></code>, default
<code class="docutils literal notranslate"><span class="pre">100,000</span></code>) for quality control. The <code class="docutils literal notranslate"><span class="pre">-t</span></code> option can be used to set
the number of threads. By default, diagnostic results and plots are
generated in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>), and
the full length of read 1 and read 2 are searched against reference cell
and feature barcodes, respectively. The per base content of both read
pairs and the distribution of matched barcode positions are summarized.
Use <code class="docutils literal notranslate"><span class="pre">-r1_c</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-r2_c</span></code> to limit the search range, and <code class="docutils literal notranslate"><span class="pre">-cb_n</span></code>
and/or <code class="docutils literal notranslate"><span class="pre">-fb_n</span></code> to set the mismatch tolerance for cell and feature
barcode matching (default <code class="docutils literal notranslate"><span class="pre">3</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba qc <span class="se">\</span>
    -1 read_1_trimmed.fq.gz <span class="se">\</span>
    -2 read_2_trimmed.fq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes_UPR_edited.tsv <span class="se">\</span>
    -r1_c <span class="m">0</span>,16 <span class="se">\</span>
    -n <span class="m">200000</span>
</pre></div>
</div>
<p>This library is built using the Chromium Single Cell 3’ Solution v3 and
sequenced on Illumina NovaSeq 6000. The first 16 bases are cell barcodes
and the following 10 bases are UMIs. Based on the base content plot, the
GC content of cell barcodes are quite even. The UMIs are slightly T
enriched.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_per_base_seq_content5.webp"><img alt="../../../_images/Pyplot_read1_per_base_seq_content5.webp" class="align-center" src="../../../_images/Pyplot_read1_per_base_seq_content5.webp" style="width: 350px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_barcodes_starting_ending3.webp"><img alt="../../../_images/Pyplot_read1_barcodes_starting_ending3.webp" class="align-center" src="../../../_images/Pyplot_read1_barcodes_starting_ending3.webp" style="width: 350px;" /></a>
<p>As for read 2, based on the per base content, it suggests that read 2 is
slightly A enriched.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_per_base_seq_content5.webp"><img alt="../../../_images/Pyplot_read2_per_base_seq_content5.webp" class="align-center" src="../../../_images/Pyplot_read2_per_base_seq_content5.webp" style="width: 425px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_barcodes_starting_ending5.webp"><img alt="../../../_images/Pyplot_read2_barcodes_starting_ending5.webp" class="align-center" src="../../../_images/Pyplot_read2_barcodes_starting_ending5.webp" style="width: 425px;" /></a>
<p>The detailed <code class="docutils literal notranslate"><span class="pre">qc</span></code> results are stored in
<code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns
indicate the matched positions on reads. <code class="docutils literal notranslate"><span class="pre">matching_description</span></code>
columns indicate mismatches in substitutions:insertions:deletions
format.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc feature_barcoding_output.tsv.gz <span class="p">|</span> grep -v no_ <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">GTGTCCTGTCGCGCATaggacttccg      GTGTCCTCACGCGCAT        0:16    2:0:0   GTGACTCGCAGTAGACGCGGGTTTAAGAGCTAAGCTGGAAACAGCATAGC      UFL1_GTGACTCGCAGTAGACGCG        0:19    0:0:0</span>
<span class="go">CGGAGAAAGACCTGTCggtatgggac      CGGAGAATCACCTGTC        0:16    2:0:0   GGCTTGTTCGCTGGTGGCGTGTTTAAGAGCTAAGCTGGAAACAGCATAGC      OST4_GGCTTGTTCGCTGGTGGCG        0:19    0:0:0</span>
<span class="go">TAGTGCAGTGCATGCCccgaatgttt      TAGTGCAGTGGTATGG        0:15    2:0:1   GGGCCGGAGGACCCGGAGCTAGTTTAAGAGCTAAGCTGGAAACAGCATAG      SCYL1_GGCCGGAGGACCCGGAGCT       1:20    0:0:0</span>
<span class="go">GTCATCCGTTGACTACgggggccact      ATCCTATGTTGACTAC        3:16    0:0:3   GGCTTGTTCGCTGGTGGCGTGTTTAAGAGCTAAGCTGGAAACAGCATAGC      OST4_GGCTTGTTCGCTGGTGGCG        0:19    0:0:0</span>
<span class="go">ATCGCCTCAAGGATATttcagattaa      TCGACCTCAAGAATGT        1:16    2:0:1   GAGCCATGGCAGATTCCTCCGTTTAAGAGCTAAGCTGGAAACAGCATAGC      CARS_GAGCCATGGCAGATTCCTC        0:19    0:0:0</span>
<span class="go">GTTGCGGGTCGCCACAgtacatactt      GTTGCGGCACGCCACA        0:16    2:0:0   GCGGTCCACAAAGGCTCAGAGTTTAAGAGCTAAGCTGGAAACAGCATAGC      DDRGK1_GCGGTCCACAAAGGCTCAG      0:19    0:0:0</span>
<span class="go">TGCATGATCGTGATCGtggagaaagt      AGTGATCAGGTGATCG        3:16    0:0:3   CTCCAGTGCTACGTGTCCCGTTTAAGAGCTAAGCTGGAAACAGCATAGCA      SEC61G_GCTCCAGTGCTACGTGTCC      0:18    0:0:1</span>
<span class="go">CTCCCAAAGCCGTGTTcatcgatatt      CTCCCAAAGACCTTTG        0:14    1:0:2   GTGACTCGCAGTAGACGCGGGTTTAAGAGCTAAGCTGGAAACAGCATAGC      UFL1_GTGACTCGCAGTAGACGCG        0:19    0:0:0</span>
<span class="go">TTTGGTTGTCGACAGAttacgcgttt      TTGGTTTGTCGCACAC        1:15    1:0:2   GACGACTAGTTAGGCGTGTAGTTTAAGAGCTAAGCTGGAAACAGCATAGC      NegCtrl3_GACGACTAGTTAGGCGTGT    0:19    0:0:0</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="barcode-extraction">
<h2>Barcode extraction<a class="headerlink" href="#barcode-extraction" title="Permalink to this heading">#</a></h2>
<p>Search ranges are set to <code class="docutils literal notranslate"><span class="pre">0,16</span></code> on read 1 and <code class="docutils literal notranslate"><span class="pre">0,19</span></code> on read 2. Two
mismatches for cell and feature barcodes (<code class="docutils literal notranslate"><span class="pre">-cb_m</span></code>, <code class="docutils literal notranslate"><span class="pre">-cf_m</span></code>) are
allowed.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba extract <span class="se">\</span>
    -1 read_1_trimmed.fq.gz <span class="se">\</span>
    -2 read_2_trimmed.fq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes_UPR_edited.tsv <span class="se">\</span>
    -o feature_barcoding_output.tsv.gz <span class="se">\</span>
    -r1_c <span class="m">0</span>,16 <span class="se">\</span>
    -r2_c <span class="m">0</span>,19 <span class="se">\</span>
    -cb_m <span class="m">2</span> <span class="se">\</span>
    -fb_m <span class="m">2</span>
</pre></div>
</div>
<p>Preview of result.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_num_mismatches       read2_seq       feature_barcode fb_num_mismatches</span>
<span class="go">GTGTCCTGTCGCGCATaggacttccg      GTGTCCTCACGCGCAT        2       GTGACTCGCAGTAGACGCGggtttaagagctaagctggaaacagcatagc    UFL1_GTGACTCGCAGTAGACGCG        0</span>
<span class="go">CGGAGAAAGACCTGTCggtatgggac      CGGAGAATCACCTGTC        2       GGCTTGTTCGCTGGTGGCGtgtttaagagctaagctggaaacagcatagc    OST4_GGCTTGTTCGCTGGTGGCG        0</span>
<span class="go">GTTGCGGGTCGCCACAgtacatactt      GTTGCGGCACGCCACA        2       GCGGTCCACAAAGGCTCAGagtttaagagctaagctggaaacagcatagc    DDRGK1_GCGGTCCACAAAGGCTCAG      0</span>
<span class="go">TTTGGTTGTCGACAGAttacgcgttt      TTTGGTTCACGACAGA        2       GACGACTAGTTAGGCGTGTagtttaagagctaagctggaaacagcatagc    NegCtrl3_GACGACTAGTTAGGCGTGT    0</span>
<span class="go">TCGTGGGAGGGAAACGcatggtcgaa      TCGTGGGTCGGAAACG        2       GTCTGGCTGGTCGCACCCGggtttaagagctaagctggaaacagcatagc    MRPL39_GTCTGGCTGGTCGCACCCG      0</span>
<span class="go">ACAGCCGTCTTGCTCAtttaacaggc      ACAGCCGAGTTGCTCA        2       GACCTTGCGTGCAGTTATGtgtttaagagctaagctggaaacagcatagc    DAD1_GACCTTGCGTGCAGTTATG        0</span>
<span class="go">CCGTAGGAGTGCGGCAgccgagcaac      CCGTAGGTCTGCGGCA        2       GACGACTAGTTAGGCGTGTagtttaagagctaagctggaaacagcatagc    NegCtrl3_GACGACTAGTTAGGCGTGT    0</span>
<span class="go">GTCTCACTCAGGACTCtatccatcca      GTCTCACAGAGGACTC        2       GCGAACGCGGCCTGAATTCcgtttaagagctaagctggaaacagcatagc    SRPR_GGCGAACGCGGCCTGAATT        2</span>
<span class="go">AAGACAACATTCGCTCtctaactgca      AAGACAAGTTTCGCTC        2       GAGTCTCCGCAAGGCCCCGggtttaagagctaagctggaaacagcatagc    ATP5B_GAGTCTCCGCAAGGCCCCG       0</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>52,352,330 out of 104,375,315 read pairs have valid cell and feature
barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-06 03:44:57,488 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-06 03:44:57,488 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-06 03:44:57,489 - fba.__main__ - INFO - Python version: 3.10</span>
<span class="go">2022-03-06 03:44:57,489 - fba.__main__ - INFO - Using extract subcommand ...</span>
<span class="go">2022-03-06 03:44:57,504 - fba.levenshtein - INFO - Number of reference cell barcodes: 8,727</span>
<span class="go">2022-03-06 03:44:57,504 - fba.levenshtein - INFO - Number of reference feature barcodes: 32</span>
<span class="go">2022-03-06 03:44:57,504 - fba.levenshtein - INFO - Read 1 coordinates to search: [0, 16)</span>
<span class="go">2022-03-06 03:44:57,504 - fba.levenshtein - INFO - Read 2 coordinates to search: [0, 19)</span>
<span class="go">2022-03-06 03:44:57,504 - fba.levenshtein - INFO - Cell barcode maximum number of mismatches: 2</span>
<span class="go">2022-03-06 03:44:57,504 - fba.levenshtein - INFO - Feature barcode maximum number of mismatches: 2</span>
<span class="go">2022-03-06 03:44:57,504 - fba.levenshtein - INFO - Read 1 maximum number of N allowed: 3</span>
<span class="go">2022-03-06 03:44:57,504 - fba.levenshtein - INFO - Read 2 maximum number of N allowed: 3</span>
<span class="go">2022-03-06 03:44:58,965 - fba.levenshtein - INFO - Matching ...</span>
<span class="go">2022-03-06 04:02:50,201 - fba.levenshtein - INFO - Read pairs processed: 10,000,000</span>
<span class="go">2022-03-06 04:21:17,938 - fba.levenshtein - INFO - Read pairs processed: 20,000,000</span>
<span class="go">2022-03-06 04:40:47,371 - fba.levenshtein - INFO - Read pairs processed: 30,000,000</span>
<span class="go">2022-03-06 05:00:15,184 - fba.levenshtein - INFO - Read pairs processed: 40,000,000</span>
<span class="go">2022-03-06 05:19:43,813 - fba.levenshtein - INFO - Read pairs processed: 50,000,000</span>
<span class="go">2022-03-06 05:39:14,583 - fba.levenshtein - INFO - Read pairs processed: 60,000,000</span>
<span class="go">2022-03-06 05:58:41,750 - fba.levenshtein - INFO - Read pairs processed: 70,000,000</span>
<span class="go">2022-03-06 06:18:09,714 - fba.levenshtein - INFO - Read pairs processed: 80,000,000</span>
<span class="go">2022-03-06 06:37:33,602 - fba.levenshtein - INFO - Read pairs processed: 90,000,000</span>
<span class="go">2022-03-06 06:56:58,484 - fba.levenshtein - INFO - Read pairs processed: 100,000,000</span>
<span class="go">2022-03-06 07:05:24,748 - fba.levenshtein - INFO - Number of read pairs processed: 104,375,315</span>
<span class="go">2022-03-06 07:05:24,771 - fba.levenshtein - INFO - Number of read pairs w/ valid barcodes: 52,352,330</span>
<span class="go">2022-03-06 07:05:24,834 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="matrix-generation">
<h2>Matrix generation<a class="headerlink" href="#matrix-generation" title="Permalink to this heading">#</a></h2>
<p>Only fragments with correctly matched cell and feature barcodes are
included, while fragments with UMI lengths less than the specified value
are discarded. UMI removal is performed using UMI-tools (<a class="reference external" href="http://www.genome.org/cgi/doi/10.1101/gr.209601.116">Smith, T., et
al. 2017. Genome Res. 27, 491–499.</a>), with the starting position on
read 1 set by <code class="docutils literal notranslate"><span class="pre">-us</span></code> (default <code class="docutils literal notranslate"><span class="pre">16</span></code>) and the length set by <code class="docutils literal notranslate"><span class="pre">-ul</span></code>
(default <code class="docutils literal notranslate"><span class="pre">12</span></code>). The UMI deduplication method can be set using <code class="docutils literal notranslate"><span class="pre">-ud</span></code>
(default <code class="docutils literal notranslate"><span class="pre">directional</span></code>), and the UMI deduplication mismatch threshold
can be specified using <code class="docutils literal notranslate"><span class="pre">-um</span></code> (default <code class="docutils literal notranslate"><span class="pre">1</span></code>).</p>
<p>The generated feature count matrix can be easily imported into
well-established single cell analysis packages: <a class="reference external" href="https://satijalab.org/seurat/">Seurat</a> and <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba count <span class="se">\</span>
    -i feature_barcoding_output.tsv.gz <span class="se">\</span>
    -o matrix_featurecount.csv.gz <span class="se">\</span>
    -us <span class="m">16</span> <span class="se">\</span>
    -ul <span class="m">10</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>10.7% (5,581,448 out of 52,352,330) of read pairs with valid cell and
feature barcodes are unique fragments. 1.4% (5,581,448 out of
404,963,129) of total sequenced read pairs contribute to the final
matrix. The meidan number of UMIs of sgRNA per cell is 413.0.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-06 07:05:24,972 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-06 07:05:24,973 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-06 07:05:24,973 - fba.__main__ - INFO - Python version: 3.10</span>
<span class="go">2022-03-06 07:05:24,973 - fba.__main__ - INFO - Using count subcommand ...</span>
<span class="go">2022-03-06 07:05:26,523 - fba.count - INFO - UMI-tools version: 1.1.2</span>
<span class="go">2022-03-06 07:05:26,525 - fba.count - INFO - UMI starting position on read 1: 16</span>
<span class="go">2022-03-06 07:05:26,526 - fba.count - INFO - UMI length: 10</span>
<span class="go">2022-03-06 07:05:26,526 - fba.count - INFO - UMI-tools deduplication threshold: 1</span>
<span class="go">2022-03-06 07:05:26,526 - fba.count - INFO - UMI-tools deduplication method: directional</span>
<span class="go">2022-03-06 07:05:26,526 - fba.count - INFO - Header line: read1_seq cell_barcode cb_num_mismatches read2_seq feature_barcode fb_num_mismatches</span>
<span class="go">2022-03-06 07:06:52,385 - fba.count - INFO - Number of lines processed: 52,352,330</span>
<span class="go">2022-03-06 07:06:52,394 - fba.count - INFO - Number of cell barcodes detected: 8,670</span>
<span class="go">2022-03-06 07:06:52,394 - fba.count - INFO - Number of features detected: 32</span>
<span class="go">2022-03-06 07:11:58,774 - fba.count - INFO - Total UMIs after deduplication: 5,581,448</span>
<span class="go">2022-03-06 07:11:58,776 - fba.count - INFO - Median number of UMIs per cell: 413.0</span>
<span class="go">2022-03-06 07:11:59,297 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="demultiplexing">
<h2>Demultiplexing<a class="headerlink" href="#demultiplexing" title="Permalink to this heading">#</a></h2>
<section id="poisson-gaussian-mixture-model">
<h3>Poisson-Gaussian mixture model<a class="headerlink" href="#poisson-gaussian-mixture-model" title="Permalink to this heading">#</a></h3>
<p>The implementation of demultiplexing method <code class="docutils literal notranslate"><span class="pre">3</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is
inspired by <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2021.12.16.473013">Replogle, M., et al. (2021)</a>. The probability threshold
for demultiplexing can be set using <code class="docutils literal notranslate"><span class="pre">-p</span></code> (default <code class="docutils literal notranslate"><span class="pre">0.5</span></code>). Use
<code class="docutils literal notranslate"><span class="pre">-nc</span></code> to specify the minimum number of positive cells for a given
feature to be considered during demultiplexing, with a default value of
<code class="docutils literal notranslate"><span class="pre">200</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    -dm <span class="m">3</span> <span class="se">\</span>
    -v <span class="se">\</span>
    -nc <span class="m">0</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-06 14:07:02,328 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-06 14:07:02,328 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-06 14:07:02,328 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2022-03-06 14:07:02,328 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2022-03-06 14:07:04,814 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;</span>
<span class="go">2022-03-06 14:07:04,814 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2022-03-06 14:07:04,814 - fba.demultiplex - INFO - Demultiplexing method: 3</span>
<span class="go">2022-03-06 14:07:04,814 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2022-03-06 14:07:04,814 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2022-03-06 14:07:04,814 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2022-03-06 14:07:04,814 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2022-03-06 14:07:04,902 - fba.demultiplex - INFO - Number of cells: 8,670</span>
<span class="go">2022-03-06 14:07:04,902 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 0</span>
<span class="go">2022-03-06 14:07:04,916 - fba.demultiplex - INFO - Number of features: 32 / 32 (after filtering / original in the matrix)</span>
<span class="go">2022-03-06 14:07:04,935 - fba.demultiplex - INFO - Features: ATP5B CARS DAD1 DDOST DDRGK1 EIF2B2 EIF2B3 EIF2B4 HARS IER3IP1 MRPL39 NegCtrl2 NegCtrl3 OST4 QARS SCYL1 SEC61A1 SEC61G SRP68 SRP72 SRPRB SRPR TELO2 TIMM23 TMED10 TMED2 TMEM167A TTI1 TTI2 UFL1 UFM1 YIPF5</span>
<span class="go">2022-03-06 14:07:04,935 - fba.demultiplex - INFO - Total UMIs: 5,581,448 / 5,581,448</span>
<span class="go">2022-03-06 14:07:04,943 - fba.demultiplex - INFO - Median number of UMIs per cell: 413.0 / 413.0</span>
<span class="go">2022-03-06 14:07:04,957 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2022-03-06 14:07:47,811 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2022-03-06 14:07:57,340 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2022-03-06 14:08:12,180 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_pgm.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_pgm.png" style="width: 700px;" /></a>
<p>Preview the demultiplexing result (<a class="reference external" href="https://www.nature.com/articles/s41587-020-0470-y/figures/1">Fig. 2d</a>, middle column): the
numbers of singlets, multiplets and negatives are 6,618 (76.3%), 1,171
(13.5%), and 881 (10.1%), respectively.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;demultiplexed/matrix_cell_identity.csv.gz&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
<span class="n">ATP5B</span>       <span class="mi">130</span>
<span class="n">CARS</span>        <span class="mi">115</span>
<span class="n">DAD1</span>        <span class="mi">163</span>
<span class="n">DDOST</span>       <span class="mi">186</span>
<span class="n">DDRGK1</span>      <span class="mi">200</span>
<span class="n">EIF2B2</span>      <span class="mi">104</span>
<span class="n">EIF2B3</span>      <span class="mi">129</span>
<span class="n">EIF2B4</span>      <span class="mi">160</span>
<span class="n">HARS</span>        <span class="mi">136</span>
<span class="n">IER3IP1</span>     <span class="mi">162</span>
<span class="n">MRPL39</span>      <span class="mi">153</span>
<span class="n">NegCtrl2</span>    <span class="mi">777</span>
<span class="n">NegCtrl3</span>    <span class="mi">898</span>
<span class="n">OST4</span>        <span class="mi">167</span>
<span class="n">QARS</span>        <span class="mi">113</span>
<span class="n">SCYL1</span>       <span class="mi">108</span>
<span class="n">SEC61A1</span>     <span class="mi">183</span>
<span class="n">SEC61G</span>      <span class="mi">256</span>
<span class="n">SRP68</span>       <span class="mi">212</span>
<span class="n">SRP72</span>       <span class="mi">191</span>
<span class="n">SRPRB</span>       <span class="mi">184</span>
<span class="n">SRPR</span>        <span class="mi">175</span>
<span class="n">TELO2</span>       <span class="mi">152</span>
<span class="n">TIMM23</span>      <span class="mi">211</span>
<span class="n">TMED10</span>      <span class="mi">156</span>
<span class="n">TMED2</span>       <span class="mi">152</span>
<span class="n">TMEM167A</span>    <span class="mi">127</span>
<span class="n">TTI1</span>        <span class="mi">132</span>
<span class="n">TTI2</span>        <span class="mi">183</span>
<span class="n">UFL1</span>        <span class="mi">204</span>
<span class="n">UFM1</span>        <span class="mi">223</span>
<span class="n">YIPF5</span>       <span class="mi">176</span>
<span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="mi">6618</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">5</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">5</span><span class="p">]:</span> <span class="mi">1171</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">6</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">6</span><span class="p">]:</span> <span class="mi">881</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">7</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">shape</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">7</span><span class="p">]:</span> <span class="p">(</span><span class="mi">32</span><span class="p">,</span> <span class="mi">8670</span><span class="p">)</span>
</pre></div>
</div>
<p>t-SNE embedding of cells based on the abundance of features (sgRNAs, no
transcriptome information used). Colors indicate the sgRNA status for
each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed_pgm.webp"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed_pgm.webp" style="width: 500px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../../cell_surface_protein_labeling/index.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Cell surface protein labeling</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../PRJNA358686/tutorial.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2020-2023, Jialei Duan
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/jlduan/fba" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3’ sgRNA-CR1<sup>cs1</sup></a><ul>
<li><a class="reference internal" href="#preparation">Preparation</a></li>
<li><a class="reference internal" href="#qc">QC</a></li>
<li><a class="reference internal" href="#barcode-extraction">Barcode extraction</a></li>
<li><a class="reference internal" href="#matrix-generation">Matrix generation</a></li>
<li><a class="reference internal" href="#demultiplexing">Demultiplexing</a><ul>
<li><a class="reference internal" href="#poisson-gaussian-mixture-model">Poisson-Gaussian mixture model</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/scripts/furo.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    </body>
</html>