<!doctype html>
<html class="no-js" lang="en" data-content_root="../../../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://jlduan.github.io/fba/_build/html/tutorials/pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html" />
<meta property="og:site_name" content="fba" />
<meta property="og:description" content="Dataset: 10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed The detailed description of this dataset can be found here., Preparation: Fastq files and feature barcodes are prepared as described here., QC: Threshold: one mismatch: When running the qc subcommand, omitti..." />
<meta property="og:image" content="https://raw.githubusercontent.com/jlduan/fba/gh-pages/docs/_static/logo.svg" />
<meta property="og:image:alt" content="fba" />
<meta name="description" content="Dataset: 10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed The detailed description of this dataset can be found here., Preparation: Fastq files and feature barcodes are prepared as described here., QC: Threshold: one mismatch: When running the qc subcommand, omitti..." />
<meta property="og:ignore_canonical" content="true" />
<meta property="article:author" content="Jialei Duan" />
<meta property="og:site_name" content="FBA" />
<meta name="twitter:card" content="summary"></meta>
<meta name="twitter:creator" content="@jlduan" />
<meta name="twitter:image" content="https://github.com/jlduan.png" />
<link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" /><link rel="next" title="10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed" href="../SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html" /><link rel="prev" title="Pseudo-bulk" href="../index.html" />

    <link rel="shortcut icon" href="../../../_static/logo.ico"/><!-- Generated with Sphinx 7.4.5 and Furo 2024.07.18 -->
        <title>10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA - FBA</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a3258c64" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=613ab9ff" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #eeeedd;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
      }
    }
  }
</style>


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-X0MEL2T23M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-X0MEL2T23M');
</script>

</head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">FBA</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo only-light" src="../../../_static/logo.svg" alt="Light Logo"/>
    <img class="sidebar-logo only-dark" src="../../../_static/logo_dark.svg" alt="Dark Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user_guide.html">User Guide</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 has-children"><a class="reference internal" href="../../crispr_screening/index.html">CRISPR screening</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of CRISPR screening</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/PRJNA358686/tutorial.html">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/PRJNA609688/tutorial.html">Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3' sgRNA-CR1&lt;sup&gt;cs1&lt;/sup&gt;</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_surface_protein_labeling/index.html">Cell surface protein labeling</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Cell surface protein labeling</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA393315/tutorial.html">CITE-seq; 8k Cord Blood Mononuclear Cells with 13 Antibodies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA658075/tutorial.html">ASAP-seq; Multiplexed CRISPR Perturbations in Primary T Cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/SC3_v3_NextGem_DI_PBMC_CSP_1K/tutorial.html">1k Human PBMCs Stained with a Panel of TotalSeq B Antibodies, Dual Indexed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../eccite-seq/index.html">ECCITE-seq</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of ECCITE-seq</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../eccite-seq/PRJNA521522/tutorial.html">6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../phage-atac/index.html">PHAGE-ATAC</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of PHAGE-ATAC</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../phage-atac/PRJNA661457/tutorial.html">PHAGE-ATAC; Anti-CD8 Phage Hashing Single-cell ATAC-seq Using CD8 T Cells from Four Human Donors</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cellplex/index.html">CellPlex</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of CellPlex</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Nuclei_30K_Multiplex/tutorial.html">30k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Nuclei Multiplexed, 12 CMOs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_hashing/index.html">Cell hashing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of Cell hashing</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_hashing/PRJNA423077/tutorial.html">Peripheral Blood Mononuclear cells with 8 Antibodies</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../multi-seq/index.html">MULTI-seq</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of MULTI-seq</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../multi-seq/PRJNA531855/tutorial.html">15k HEK293 and 40k HMECs Multiplexed by Lipid- and Cholesterol-tagged Indices</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../targeted_transcript_enrichment/index.html">Targeted transcript enrichment</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of Targeted transcript enrichment</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../targeted_transcript_enrichment/Targeted_NGSC3_DI_HodgkinsLymphoma_GeneSignature/tutorial.html">Hodgkin’s Lymphoma, Dissociated Tumor: Targeted, Gene Signature Panel</a></li>
</ul>
</li>
<li class="toctree-l2 current has-children"><a class="reference internal" href="../index.html">Pseudo-bulk</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of Pseudo-bulk</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/jlduan/fba">GitHub Repository</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="https://github.com/jlduan/fba/blob/gh-pages/docs/tutorials/pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.rst?plain=true" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div><div class="edit-this-page">
  <a class="muted-link" href="https://github.com/jlduan/fba/edit/gh-pages/docs/tutorials/pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.rst" title="Edit this page">
    <svg><use href="#svg-pencil"></use></svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="k-a375-cells-transduced-with-1-non-target-and-1-target-sgrna">
<span id="tutorial-pseudo-bulk-sc3-v3-nextgem-di-crispr-10k"></span><h1>10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA<a class="headerlink" href="#k-a375-cells-transduced-with-1-non-target-and-1-target-sgrna" title="Link to this heading">¶</a></h1>
<p><strong>Dataset</strong>: 10k A375 Cells Transduced with (1) Non-Target and (1)
Target sgRNA, Dual Indexed</p>
<p>The detailed description of this dataset can be found <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/4.0.0/SC3_v3_NextGem_DI_CRISPR_10K">here</a>.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="preparation">
<h2>Preparation<a class="headerlink" href="#preparation" title="Link to this heading">¶</a></h2>
<p>Fastq files and feature barcodes are prepared as described <a class="reference internal" href="../../crispr_screening/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html#tutorial-crispr-screening-sc3-v3-nextgem-di-crispr-10k"><span class="std std-ref">here</span></a>.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="qc">
<h2>QC<a class="headerlink" href="#qc" title="Link to this heading">¶</a></h2>
<section id="threshold-one-mismatch">
<h3>Threshold: one mismatch<a class="headerlink" href="#threshold-one-mismatch" title="Link to this heading">¶</a></h3>
<p>When running the <code class="docutils literal notranslate"><span class="pre">qc</span></code> subcommand, omitting the <code class="docutils literal notranslate"><span class="pre">-1</span></code> (read 1) option
activates bulk mode, which is useful for designing and testing feature
barcoding assays prior to conducting single cell experiments. For
example, bulk mode can be used to estimate: 1) the number of reads with
valid feature barcodes, which can indicate primer specificity and
suggest the number of reads required for sequencing; and 2) the
distribution of feature barcodes, which reflects the biological aspect
of the assay design.</p>
<p>To specify read 2 and feature barcodes, use <code class="docutils literal notranslate"><span class="pre">-2</span></code> and <code class="docutils literal notranslate"><span class="pre">-f</span></code> options,
respectively. The search range for read 2 can be controlled with
<code class="docutils literal notranslate"><span class="pre">-r2_c</span></code>. In this example, a single mismatch is allowed for feature
barcode matching, set by <code class="docutils literal notranslate"><span class="pre">-fb_m</span></code>. Use <code class="docutils literal notranslate"><span class="pre">-n</span></code> to specify the number of
reads to analyze, with <code class="docutils literal notranslate"><span class="pre">None</span></code> indicating that all reads in the fastq
file should be analyzed. By default, the distribution of detected
feature barcodes is summarized in <code class="docutils literal notranslate"><span class="pre">qc/feature_barcode_frequency.csv</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>qc<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-2<span class="w"> </span>SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_combined_R2_001.fastq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-f<span class="w"> </span>SC3_v3_NextGem_DI_CRISPR_10K_feature_ref_edited.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-r2_c<span class="w"> </span><span class="m">31</span>,51<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-fb_m<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-n<span class="w"> </span>None
</pre></div>
</div>
<p>The content of <code class="docutils literal notranslate"><span class="pre">qc/feature_barcode_frequency.csv</span></code>.</p>
<div class="table-wrapper docutils container">
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>feature barcode</p></td>
<td><p>num_reads</p></td>
<td><p>percentage</p></td>
</tr>
<tr class="row-even"><td><p>NON_TARGET-1_AACGTGCTGACGATGCGGGC</p></td>
<td><p>59,310,228</p></td>
<td><p>0.6979885931379053</p></td>
</tr>
<tr class="row-odd"><td><p>RAB1A-2_GCCGGCGAACCAGGAAATAG</p></td>
<td><p>25,662,834</p></td>
<td><p>0.3020114068620947</p></td>
</tr>
</tbody>
</table>
</div>
<p>Result summary.</p>
<p>58.59% (84,973,062 / 145,032,428) of reads have valid feature barcodes
(sgRNAs). Although the valid read ratio is 1.663170816
(NON_TARGET-1_AACGTGCTGACGATGCGGGC / RAB1A-2_GCCGGCGAACCAGGAAATAG;
59,310,228 / 25,662,834), cells transduced with them separately are
mixed at 1: 1 ratio. See <a class="reference internal" href="../../crispr_screening/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html#tutorial-crispr-screening-sc3-v3-nextgem-di-crispr-10k"><span class="std std-ref">here</span></a> for more
details.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-02-17 16:12:35,393 - fba.__main__ - INFO - fba version: 0.0.7</span>
<span class="go">2021-02-17 16:12:35,393 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-02-17 16:12:35,393 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2021-02-17 16:12:35,393 - fba.__main__ - INFO - Using qc subcommand ...</span>
<span class="go">2021-02-17 16:12:35,394 - fba.__main__ - INFO - Bulk mode enabled: only feature barcodes on reads 2 are analyzed</span>
<span class="go">2021-02-17 16:12:35,394 - fba.__main__ - INFO - Skipping arguments: &quot;-w/--whitelist&quot;, &quot;-cb_m/--cb_mismatches&quot;, &quot;-r1_c/--read1_coordinate&quot;</span>
<span class="go">2021-02-17 16:12:35,395 - fba.qc - INFO - Number of reference feature barcodes: 2</span>
<span class="go">2021-02-17 16:12:35,395 - fba.qc - INFO - Read 2 coordinates to search: [31, 51)</span>
<span class="go">2021-02-17 16:12:35,395 - fba.qc - INFO - Feature barcode maximum number of mismatches: 1</span>
<span class="go">2021-02-17 16:12:35,395 - fba.qc - INFO - Read 2 maximum number of N allowed: inf</span>
<span class="go">2021-02-17 16:12:35,395 - fba.qc - INFO - Number of read pairs to analyze: all</span>
<span class="go">2021-02-17 16:12:35,395 - fba.qc - INFO - Matching ...</span>
<span class="go">2021-02-17 16:15:07,684 - fba.qc - INFO - Reads processed: 10,000,000</span>
<span class="go">2021-02-17 16:17:39,083 - fba.qc - INFO - Reads processed: 20,000,000</span>
<span class="go">2021-02-17 16:20:09,116 - fba.qc - INFO - Reads processed: 30,000,000</span>
<span class="go">2021-02-17 16:22:38,981 - fba.qc - INFO - Reads processed: 40,000,000</span>
<span class="go">2021-02-17 16:25:11,671 - fba.qc - INFO - Reads processed: 50,000,000</span>
<span class="go">2021-02-17 16:27:44,790 - fba.qc - INFO - Reads processed: 60,000,000</span>
<span class="go">2021-02-17 16:30:18,110 - fba.qc - INFO - Reads processed: 70,000,000</span>
<span class="go">2021-02-17 16:32:51,391 - fba.qc - INFO - Reads processed: 80,000,000</span>
<span class="go">2021-02-17 16:35:24,625 - fba.qc - INFO - Reads processed: 90,000,000</span>
<span class="go">2021-02-17 16:37:57,678 - fba.qc - INFO - Reads processed: 100,000,000</span>
<span class="go">2021-02-17 16:40:30,706 - fba.qc - INFO - Reads processed: 110,000,000</span>
<span class="go">2021-02-17 16:43:03,867 - fba.qc - INFO - Reads processed: 120,000,000</span>
<span class="go">2021-02-17 16:45:37,197 - fba.qc - INFO - Reads processed: 130,000,000</span>
<span class="go">2021-02-17 16:48:10,511 - fba.qc - INFO - Reads processed: 140,000,000</span>
<span class="go">2021-02-17 16:49:27,662 - fba.qc - INFO - Number of reads processed: 145,032,428</span>
<span class="go">2021-02-17 16:49:27,663 - fba.qc - INFO - Number of reads w/ valid feature barcodes: 84,973,062</span>
<span class="go">2021-02-17 16:49:27,664 - fba.__main__ - INFO - Output file: qc/feature_barcode_frequency.csv</span>
<span class="go">2021-02-17 16:49:27,689 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="threshold-two-mismatches">
<h3>Threshold: two mismatches<a class="headerlink" href="#threshold-two-mismatches" title="Link to this heading">¶</a></h3>
<p>Let’s relax the threshold to allow 2 mismatches for feature barcode
matching (set by <code class="docutils literal notranslate"><span class="pre">-fb_m</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>qc<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-2<span class="w"> </span>SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_combined_R2_001.fastq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-f<span class="w"> </span>SC3_v3_NextGem_DI_CRISPR_10K_feature_ref_edited.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-r2_c<span class="w"> </span><span class="m">31</span>,51<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-fb_m<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-n<span class="w"> </span>None
</pre></div>
</div>
<p>The content of <code class="docutils literal notranslate"><span class="pre">qc/feature_barcode_frequency.csv</span></code>.</p>
<div class="table-wrapper docutils container">
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>feature barcode</p></td>
<td><p>num_reads</p></td>
<td><p>percentage</p></td>
</tr>
<tr class="row-even"><td><p>NON_TARGET-1_AACGTGCTGACGATGCGGGC</p></td>
<td><p>66,334,740</p></td>
<td><p>0.6613115326075217</p></td>
</tr>
<tr class="row-odd"><td><p>RAB1A-2_GCCGGCGAACCAGGAAATAG</p></td>
<td><p>33,973,113</p></td>
<td><p>0.3386884673924782</p></td>
</tr>
</tbody>
</table>
</div>
<p>Result summary.</p>
<p>69.16% (100,307,853 / 145,032,428) of reads have valid feature barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-02-17 16:12:00,407 - fba.__main__ - INFO - fba version: 0.0.7</span>
<span class="go">2021-02-17 16:12:00,407 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-02-17 16:12:00,408 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2021-02-17 16:12:00,408 - fba.__main__ - INFO - Using qc subcommand ...</span>
<span class="go">2021-02-17 16:12:00,408 - fba.__main__ - INFO - Bulk mode enabled: only feature barcodes on reads 2 are analyzed</span>
<span class="go">2021-02-17 16:12:00,408 - fba.__main__ - INFO - Skipping arguments: &quot;-w/--whitelist&quot;, &quot;-cb_m/--cb_mismatches&quot;, &quot;-r1_c/--read1_coordinate&quot;</span>
<span class="go">2021-02-17 16:12:00,426 - fba.qc - INFO - Number of reference feature barcodes: 2</span>
<span class="go">2021-02-17 16:12:00,426 - fba.qc - INFO - Read 2 coordinates to search: [31, 51)</span>
<span class="go">2021-02-17 16:12:00,426 - fba.qc - INFO - Feature barcode maximum number of mismatches: 2</span>
<span class="go">2021-02-17 16:12:00,426 - fba.qc - INFO - Read 2 maximum number of N allowed: inf</span>
<span class="go">2021-02-17 16:12:00,426 - fba.qc - INFO - Number of read pairs to analyze: all</span>
<span class="go">2021-02-17 16:12:00,426 - fba.qc - INFO - Matching ...</span>
<span class="go">2021-02-17 16:28:02,710 - fba.qc - INFO - Reads processed: 10,000,000</span>
<span class="go">2021-02-17 16:44:07,554 - fba.qc - INFO - Reads processed: 20,000,000</span>
<span class="go">2021-02-17 17:00:13,431 - fba.qc - INFO - Reads processed: 30,000,000</span>
<span class="go">2021-02-17 17:16:17,034 - fba.qc - INFO - Reads processed: 40,000,000</span>
<span class="go">2021-02-17 17:32:21,635 - fba.qc - INFO - Reads processed: 50,000,000</span>
<span class="go">2021-02-17 17:48:26,948 - fba.qc - INFO - Reads processed: 60,000,000</span>
<span class="go">2021-02-17 18:04:31,050 - fba.qc - INFO - Reads processed: 70,000,000</span>
<span class="go">2021-02-17 18:20:34,413 - fba.qc - INFO - Reads processed: 80,000,000</span>
<span class="go">2021-02-17 18:36:38,778 - fba.qc - INFO - Reads processed: 90,000,000</span>
<span class="go">2021-02-17 18:52:44,033 - fba.qc - INFO - Reads processed: 100,000,000</span>
<span class="go">2021-02-17 19:08:49,500 - fba.qc - INFO - Reads processed: 110,000,000</span>
<span class="go">2021-02-17 19:24:56,356 - fba.qc - INFO - Reads processed: 120,000,000</span>
<span class="go">2021-02-17 19:41:02,072 - fba.qc - INFO - Reads processed: 130,000,000</span>
<span class="go">2021-02-17 19:57:09,967 - fba.qc - INFO - Reads processed: 140,000,000</span>
<span class="go">2021-02-17 20:05:15,665 - fba.qc - INFO - Number of reads processed: 145,032,428</span>
<span class="go">2021-02-17 20:05:15,666 - fba.qc - INFO - Number of reads w/ valid feature barcodes: 100,307,853</span>
<span class="go">2021-02-17 20:05:15,667 - fba.__main__ - INFO - Output file: qc/feature_barcode_frequency.csv</span>
<span class="go">2021-02-17 20:05:15,701 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Pseudo-bulk</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2020-2024, Jialei Duan
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/jlduan/fba" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA</a><ul>
<li><a class="reference internal" href="#preparation">Preparation</a></li>
<li><a class="reference internal" href="#qc">QC</a><ul>
<li><a class="reference internal" href="#threshold-one-mismatch">Threshold: one mismatch</a></li>
<li><a class="reference internal" href="#threshold-two-mismatches">Threshold: two mismatches</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../../../_static/documentation_options.js?v=b6a3c597"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/furo.js?v=5fa4622c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=3b13dea8"></script>
    </body>
</html>