<!doctype html>
<html class="no-js">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta property="og:title" content="CROP-Seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)" />
  <meta property="og:type" content="website" />
  <meta property="og:url" content="tutorials/crispr_screening/PRJNA358686/tutorial.html" />
  <meta property="og:description" content="Dataset: CROP-Seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T) Datlinger, P., Rendeiro, A.F., Schmidl, C., Krausgruber, T., Traxler, P., Klughammer, J., Schuster, L.C., Ku..." />
  <link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" /><link rel="next" title="Cell surface protein labeling" href="../../cell_surface_protein_labeling/index.html" /><link rel="prev" title="10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA" href="../SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html" />

    <link rel="shortcut icon" href="../../../_static/logo.svg"/><meta name="generator" content="sphinx-4.3.2, furo 2021.10.09"/>
        <title>CROP-Seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T) - FBA</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?digest=0254c309f5cadf746f1a613e7677379ac9c8cdcd" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?digest=16fb25fabf47304eee183a5e9af80b1ba98259b1" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  body[data-theme="dark"] {
    --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
  }
  @media (prefers-color-scheme: dark) {
    body:not([data-theme="light"]) {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
  }
</style>


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-X0MEL2T23M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-X0MEL2T23M');
</script>

</head>
  <body>
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" />
      <line x1="4" y1="6" x2="20" y2="6" />
      <line x1="10" y1="12" x2="20" y2="12" />
      <line x1="6" y1="18" x2="20" y2="18" />
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">FBA</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand centered" href="../../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../../_static/logo.svg" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user_guide.html">User Guide</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current has-children"><a class="reference internal" href="../index.html">CRISPR screening</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">CROP-Seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_surface_protein_labeling/index.html">Cell surface protein labeling</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA393315/tutorial.html">CITE-seq; 8k Cord Blood Mononuclear Cells with 13 Antibodies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/SC3_v3_NextGem_DI_PBMC_CSP_1K/tutorial.html">1k Human PBMCs Stained with a Panel of TotalSeq B Antibodies, Dual Indexed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cellplex/index.html">CellPlex</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Nuclei_30K_Multiplex/tutorial.html">30k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Nuclei Multiplexed, 12 CMOs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_hashing/index.html">Cell hashing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_hashing/PRJNA423077/tutorial.html">Peripheral Blood Mononuclear cells with 8 Antibodies</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../multi-seq/index.html">MULTI-seq</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../multi-seq/PRJNA531855/tutorial.html">15k HEK293 and 40k HMECs Multiplexed by Lipid- and Cholesterol-tagged Indices</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../eccite-seq/index.html">ECCITE-seq</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../eccite-seq/PRJNA521522/tutorial.html">6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../targeted_transcript_enrichment/index.html">Targeted transcript enrichment</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../targeted_transcript_enrichment/Targeted_NGSC3_DI_HodgkinsLymphoma_GeneSignature/tutorial.html">Hodgkin’s Lymphoma, Dissociated Tumor: Targeted, Gene Signature Panel</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../pseudo-bulk/index.html">Pseudo-bulk</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/jlduan/fba">GitHub Repository</a></li>
</ul>

</div>
</div>
      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="crop-seq-1-1-1-mixture-of-dnmt3b-mbd1-and-tet2-knockout-cell-lines-hek293t">
<span id="tutorial-crispr-screening-prjna358686"></span><h1>CROP-Seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)<a class="headerlink" href="#crop-seq-1-1-1-mixture-of-dnmt3b-mbd1-and-tet2-knockout-cell-lines-hek293t" title="Permalink to this headline">¶</a></h1>
<p>Dataset: CROP-Seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</p>
<p>Datlinger, P., Rendeiro, A.F., Schmidl, C., Krausgruber, T., Traxler, P., Klughammer, J., Schuster, L.C., Kuchler, A., Alpar, D., and Bock, C. (2017). <a class="reference external" href="https://doi.org/10.1038/nmeth.4177">Pooled CRISPR screening with single-cell transcriptome readout</a>. Nat. Methods <em>14</em>, 297–301.</p>
<div class="line-block">
<div class="line"><br/></div>
</div>
<section id="preparation">
<h2>Preparation<a class="headerlink" href="#preparation" title="Permalink to this headline">¶</a></h2>
<p>Download fastq files from <a class="reference external" href="https://www.ebi.ac.uk/ena/browser/view/PRJNA521522?show=reads">European Nucleotide Archive</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>curl -O ftp.sra.ebi.ac.uk/vol1/fastq/SRR516/009/SRR5163029/SRR5163029_1.fastq.gz

<span class="gp">$ </span>curl -O ftp.sra.ebi.ac.uk/vol1/fastq/SRR516/009/SRR5163029/SRR5163029_2.fastq.gz
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="clustering">
<h2>Clustering<a class="headerlink" href="#clustering" title="Permalink to this headline">¶</a></h2>
<p>This dataset is single-cell RNA-seq on HEK293T cell lines mixed in equal proportions knocked out for DNMT3B, MBD1, and TET2. We need to re-process these raw reads to get cell-associated barcodes for downstream analysis. The platform of this dataset is Drop-Seq. The details about the original data processing can be found <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2450588">here</a>. Briefly, raw data were process with Drop-seq Tools v1.12 software. The first 12 bases on read 1 are cell barcodes, followed by 8 bases UMIs. Captured single cell transcripts are on read 2.</p>
<p>Inspect cell barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat cell_barcodes.txt

<span class="go">AACGGGCATGGG</span>
<span class="go">AACTGGGCATGG</span>
<span class="go">AAGACAGCGTGT</span>
<span class="go">AAGGGCGTACTC</span>
<span class="go">AATAAATACAAA</span>
<span class="go">AATAAATACAAC</span>
<span class="go">AATAAATACAAG</span>
<span class="go">AATAAATACAAT</span>
<span class="go">AATCAATCGCAC</span>
<span class="go">AATCAATCGCAG</span>
</pre></div>
</div>
<p>Prepare feature barcodes.</p>
<p>gRNA sequences can be found in Supplementary Table 1.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat feature_barcodes.tsv

<span class="go">DNMT3B  CAGGATTGGGGGCGAGTCGG</span>
<span class="go">MBD1    ATAGGTGTCTGAGCGTCCAC</span>
<span class="go">TET2    CAGGACTCACACGACTATTC</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="barcode-extraction">
<h2>Barcode extraction<a class="headerlink" href="#barcode-extraction" title="Permalink to this headline">¶</a></h2>
<p>The transcripts derived from CROPseq-Guide-Puro and captured by Drop-Seq beads contain sgRNA sequences. There are no secondary libraries built on top of this single-cell RNA-Seq library for gRNA enrichment. The transcripts derived from CROPseq-Guide-Puro are captured by the ployA tails. Therefore, the locations of sgRNA on read 2 vary. We need to extract the sgRNA sequences from read 2.</p>
<p><code class="docutils literal notranslate"><span class="pre">qc</span></code> mode is used for gRNA extraction. Use <code class="docutils literal notranslate"><span class="pre">-n</span></code> to specify the number of reads to analyze, <code class="docutils literal notranslate"><span class="pre">None</span></code> means all the reads. Use <code class="docutils literal notranslate"><span class="pre">-t</span></code> to set the number of threads. By default, the diagnostic results and plots are generated in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>), and full length of read 1 and read 2 are searched against reference cell and feature barcodes, respectively. The per base content of both read pairs and the distribution of matched barcode positions are summarized. Use <code class="docutils literal notranslate"><span class="pre">-r1_c</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-r2_c</span></code> to limit the search range for read 1 and read 2 respectively. Use <code class="docutils literal notranslate"><span class="pre">-cb_n</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-fb_n</span></code> to set the mismatch tolerance for cell and feature barcode matching (default <code class="docutils literal notranslate"><span class="pre">3</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba qc <span class="se">\</span>
    -1 SRR5163029_1.fastq.gz <span class="se">\</span>
    -2 SRR5163029_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes.tsv <span class="se">\</span>
    -cb_m <span class="m">2</span> <span class="se">\</span>
    -fb_m <span class="m">2</span> <span class="se">\</span>
    -cb_n <span class="m">15</span> <span class="se">\</span>
    -fb_n <span class="m">15</span> <span class="se">\</span>
    -r1_c <span class="m">0</span>,12 <span class="se">\</span>
    -t <span class="nv">$SLURM_CPUS_ON_NODE</span> <span class="se">\</span>
    --num_reads None
</pre></div>
</div>
<p>The first 12 bases are cell barcodes and the following 8 bases are UMIs. Based on the base content plot, the GC content of cell barcodes and UMIs are quite even.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_per_base_seq_content4.png"><img alt="../../../_images/Pyplot_read1_per_base_seq_content4.png" class="align-center" src="../../../_images/Pyplot_read1_per_base_seq_content4.png" style="width: 1200px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_barcodes_starting_ending1.png"><img alt="../../../_images/Pyplot_read1_barcodes_starting_ending1.png" class="align-center" src="../../../_images/Pyplot_read1_barcodes_starting_ending1.png" style="width: 1200px;"/></a>
<p>As for read 2, based on the per base content, after the first 12 bases, the reads are A enriched.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_per_base_seq_content4.png"><img alt="../../../_images/Pyplot_read2_per_base_seq_content4.png" class="align-center" src="../../../_images/Pyplot_read2_per_base_seq_content4.png" style="width: 1200px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_barcodes_starting_ending5.png"><img alt="../../../_images/Pyplot_read2_barcodes_starting_ending5.png" class="align-center" src="../../../_images/Pyplot_read2_barcodes_starting_ending5.png" style="width: 1200px;"/></a>
<p>The detailed <code class="docutils literal notranslate"><span class="pre">qc</span></code> results are stored in <code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns indicate the matched positions on reads. <code class="docutils literal notranslate"><span class="pre">matching_description</span></code> columns indicate mismatches in substitutions:insertions:deletions format.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc qc/feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">GATATCGCACCGgggagggtttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttctttctctattcataccttttctctccatacccaa no_match        NA      NA      AATTTGATGTACACAACATTAAAAGTACTGACACCTGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACACAAAAAAAAAATCACATTATAACACAACAAAAAAA NA      NA      NA</span>
<span class="go">TTTAGGATCGTTtgatgtattttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttcttctttcttttttattctttacaacatcctaccataacata no_match        NA      NA      ATTAAAAATATTGTGGCAGGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAACAAAAAAAAATCAGCTATATAACCACTAATACTTCTA NA      NA      NA</span>
<span class="go">GTCGAAACTCTTaacgggatttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt no_match        NA      NA      TTATAATGGTTACAAATAAAGCAATAGCATCACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA NA      NA      NA</span>
<span class="go">GTTTACGTGTTCatgggcgattttttttttttttttttttttttttttttaaaaaagttaaaagggggcccgtggggggacaaatagaggggcctagagttccaccccccatcccacaaaaaaaaccctcaccgcacagggcctcgcccct GTTTACGTGTTC    0:12    0:0:0   GGAGTACGGAGAATTCTATAAGAGCTTGACCAATGACTGGGAAGATCACTTGGCAGTGAAGCATTTTTCAGTTGAAGGACAGTTGGAATTCAGAGCCCTTCTATTTGTCCCACGACGTGCTCCTTTTGATCTGTTTGAAAAAAAAAAAAAA no_match        NA      NA</span>
<span class="go">CCGTCCTAGTTGgtgtatattttttttgtttttttttttttttcaccgggtcagagctgcccctaagtaccacgtcccgtcccacctttatcggacctcggccaccacaaattgcttatccagagtgcccccctccgcccatcccagactc CCGTCCTAGTTG    0:12    0:0:0   AATTAAGTCTCGTAAAGAACGAGAAGCTGAACTTGGACCTAGGGCAACCGACTTCACCAATGTTTACAGCGAGAATCTTGGTGACGACGTGGATGATGAGCGCCTTAAGGTTCTCTTTGGCAAGTTTGGGCCTGCCTTGAGTGTGCGACTT no_match        NA      NA</span>
<span class="go">TAGCAAGCCCTTttgagcacttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttattttttcttttttttatcataaatat TAGCAAGCCCTT    0:12    0:0:0   CAAACGTTTTATACTAAATAAATATCAAACTACATTCTTCTGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGACAAAAAAAAACAAAAAAAAA no_match        NA      NA</span>
<span class="go">ATCCAAATCTAGgtatagggtttttttttttttttttttttttttttttttaattttttttttttttttttattttttctttttattgggtgtaactcctgcgtacaaaaacctaccctgggggggaaggggattttctcattatagctgg ATCCAAATCTAG    0:12    0:0:0   AAATCGACGGGGACAGGCTCGCCGACTTGCCCCTACTTTTCGATGGGCCATACCCAATAGGCAGATCAATGTTGGGGTGGGTGGTGATGGAGATGATATGGACATATTCATGGAGGAGATGAGAGAAATCAGAAGAAAACTTAGGGAGCTG no_match        NA      NA</span>
<span class="go">TTAACCTTAGCGgactctttttttttttttttttttttttggaaacttaaaatccccaaaaaactgtttattatacaagggagttttgagtcacggagggcttaccggtaggagttctggttgggggcgcgggcaccagggcctccaaact TTACACCATTAG    0:10    0:0:2   TCCGTGTCCGTGTAAAGGGTGGTGGTCACGTGGCCCAGATTTATGCTATCCGTCAGTCCATCTCCAAAGCCCTGGTGGCCTATTACCAGAACTATGTGGCTGAGGCTTCCAAGAAGGAGATCAAAGACATCCTCATCCAGTATGACCGGAC no_match        NA      NA</span>
<span class="go">ACGTGTCGTCCGgtgctattttttttttttttttttttttttttttttttttttttttttttattttttttttttttttctagttaacagtttgttacttcacaaaatgaaacagtaagagggtaaaaaaaaaatatttagaaaagaaaaa ACGTGATCGTCG    0:11    1:0:1   TTATCACATAGTGTTGTATTTGAAACATAGTTCATGGTTTTATCAAAGAACTGAAGATGAGAATACTGGTCATCTAACTTTGTAATTTGATTTGATTATACTGTAAAGTTTGACAGTCTCATTTTATCACTGCGTTTGTATCTATTACTAA no_match        NA      NA</span>
<span class="go">TTTCAGTATTGGggcgaattttttttttttttttttttttttttttttttttttttttttttttggctagtttttttgtggtttttgcttttggttctctcgtttgccctggagctcccaggtccctttcttgtcctaccataggtaaccc TTTCAGTATTGG    0:12    0:0:0   GGACGAAACACCGATAGGTGTCTGAGCGTCCACGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAGCTTGGCGTAACTAGATCTTGAGACACTGCTTTT MBD1_ATAGGTGTCTGAGCGTCCAC       13:33 0:0:0</span>
<span class="go">AATCACCTACCTtccatcagtttttttttttttttttttttttttttattactctatgttatctttttttttgctttttttgttggttctctttttttcccgagagctcccctgctcccatctgggttaaccagagagacccactacaagc AATCACCTACCT    0:12    0:0:0   GGACTCACACGACTATTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAGCTTGGCGTAACTAGATCTTGAGACCCTGCTTTTTGCTTGTACTGGGTC TET2_CAGGACTCACACGACTATTC       0:18  0:0:2</span>
<span class="go">GCCTGTCTCATCacatccttttttttttttttttttttttttttttttttttttttttttttgttttttttttttgtttttttgtgtgttttctttgtttggccaggagtctccggcccgcatttgtgttttctcgcgaagacccagtcaa CCGTGTCTCCTC    1:12    1:0:1   GACTATCATATGCTTACCGTAACTTGAAAGTATTTCGATTTCTTGGCTTTATATATCTTGTGGAAAGGACGAAACACCGCGGGTCTCACACGACTATTCGTTTTAGAGCTAGTTATCGCATGTTCAAATAAGGCTAGTCCGTTATCAACTT TET2_CAGGACTCACACGACTATTC       79:99 2:0:0</span>
<span class="go">CTAGGTACCACTagacagtttttttttttttttttttttttttttttttttttttttttttctctatgtgtgcttttttttggctttagtctgtgggtccctagttagccccggcgcccccacgcgcagaacgtgtcttaccacaagaacc CTAGGTACCACT    0:12    0:0:0   TTCTTGGGTAGTTTGCAGTTTTTAAAATTATGTTTTAAAATGGACTATCATATGCTTACCGTAACTTGAAAGTATTTCGATTTCTTGGCTTTATATATCTTGTGGAACGGACGAAACACCGATAGGTGTCTGAGCGTCCACGTTTTAGAGC MBD1_ATAGGTGTCTGAGCGTCCAC       121:1410:0:0</span>
<span class="go">TCTTCCACTACCgtcccgtcttttttttttttttttttttttttttttttttttttttctttatgtcagttttttttgtgctttagtattgggttcccttgtttgcccgagggctcccaggcccagatttgggctaaccaaagggaccccg TCTTCCACTACC    0:12    0:0:0   ACCGATAGGTGTCTGAGCGTCCACGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAGCTTGGCGTAACTAGATCTTGAGACACTGCTTTTTGCTTGTAC MBD1_ATAGGTGTCTGAGCGTCCAC       4:24  0:0:0</span>
<span class="go">GTGTCGCTCTGAcctatttttttttttttttttttttctttttacctttaacaatactgttctcatgttatatgctgttattaaatgtgccccccatgtctctagagatcctacccctgctgatcccgtcccaacaaaaaatactagtaca GTGTACCGCTCT    0:10    0:0:2   AAGGAAAAGACGAAACACCGATAGGAGTCTGAGCGTCCACAAATTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTTGTCCGTTAGCAACTTGAAACACTGGCACCGAGTCGGTCCTTTTTTAAGCTTGGCGTAACTGGATCTTGAGCCTC MBD1_ATAGGTGTCTGAGCGTCCAC       20:40 1:0:0</span>
<span class="go">CTTAATTTGGTGggaagattttttttttttttttttttttttttttttaagtactttaagtaagctttttttaggctttagccgtgggttcccctgttagcccgggaggtccccgggcccaatctgggcctaacagagaggccccgtacaa CTTAATTTGGTG    0:12    0:0:0   CCGTAACTTGAAAGTATTTCGATTTCTTGGCTTTATATATCTTGTGGAAAGGACGAAACACCGCAGGACTCACACGACTCTTCGTTTTAGAGCTAGCAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGT TET2_CAGGACTCACACGACTATTC       63:83 1:0:0</span>
<span class="go">GGCGGACAATACccctacgattttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt GGCGGACAATAC    0:12    0:0:0   CCGCAGGACTCACACGACTATTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAGCTTGGCGTAACTAGATCTTGAGACACTGCTTTTTGCTTGTTCT TET2_CAGGACTCACACGACTATTC       3:23  0:0:0</span>
<span class="go">TAAATGTTTAGGtgtccggattttttttttttttttttttttttttttttttttttttttttttttttttttttttctttttcttttttttttctgtttatcatgtggggttttgttttttccttagatttctcccactgtcgtcttgttt TAAATGTTTAGG    0:12    0:0:0   ATCATATGCTTACCGTAACTTGAAAGTATTTCGATTTCTTTGCTTTATATATCTTGTGGAAAGGACGAAACTCCGGTAGGTGTCTGAGCGTCCTCGTTTTTGAGCTTGAAATAGCCAGTTAAAATAAGGCTAGTCCGTTATCAACTTGACA MBD1_ATAGGTGTCTGAGCGTCCAC       75:95 2:0:0</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="matrix-generation">
<h2>Matrix generation<a class="headerlink" href="#matrix-generation" title="Permalink to this headline">¶</a></h2>
<p>Only fragments with correct (passed the criteria) cell and feature barcodes are included. UMI removal is powered by UMI-tools (<a class="reference external" href="http://www.genome.org/cgi/doi/10.1101/gr.209601.116">Smith, T., et al. 2017. Genome Res. 27, 491–499.</a>). Use <code class="docutils literal notranslate"><span class="pre">-us</span></code> to set the UMI starting position on read 1 (default <code class="docutils literal notranslate"><span class="pre">16</span></code>). Use <code class="docutils literal notranslate"><span class="pre">-ul</span></code> to set the UMI length (default <code class="docutils literal notranslate"><span class="pre">12</span></code>). Fragments with UMI length less than this value are discarded. UMI deduplication method is set by <code class="docutils literal notranslate"><span class="pre">-ud</span></code> (default <code class="docutils literal notranslate"><span class="pre">directional</span></code>). Use <code class="docutils literal notranslate"><span class="pre">-um</span></code> to set UMI deduplication mismatch threshold (default <code class="docutils literal notranslate"><span class="pre">1</span></code>).</p>
<p>The generated feature count matrix can be easily imported into well-established single cell analysis packages: <a class="reference external" href="https://satijalab.org/seurat/">Seruat</a> and <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba count <span class="se">\</span>
    -i feature_barcoding_output.tsv.gz <span class="se">\</span>
    -o matrix_featurecount.csv.gz <span class="se">\</span>
    -us <span class="m">12</span> <span class="se">\</span>
    -ul <span class="m">8</span> <span class="se">\</span>
    -um <span class="m">1</span> <span class="se">\</span>
    -ud directional
</pre></div>
</div>
<p>Result summary.</p>
<p>7.6% (7,145,799 out of 93,795,979) of read pairs with valid cell and feature barcodes are unique fragments. 4.9% (7,143,943 out of 145,032,428) of total sequenced read pairs contribute to the final matrix.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2020-10-20 04:47:32,738 - fba.__main__ - INFO - fba version: 0.0.7</span>
<span class="go">2020-10-20 04:47:32,738 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2020-10-20 04:47:32,738 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2020-10-20 04:47:32,738 - fba.__main__ - INFO - Using count subcommand ...</span>
<span class="go">2020-10-20 04:47:32,738 - fba.count - INFO - UMI-tools version: 1.0.1</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - UMI starting position on read 1: 16</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - UMI length: 12</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - UMI-tools deduplication threshold: 1</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - UMI-tools deduplication method: directional</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - Header line: read1_seq cell_barcode cb_num_mismatches read2_seq feature_barcode fb_num_mismatches</span>
<span class="go">2020-10-20 04:51:50,886 - fba.count - INFO - Number of lines processed: 93,795,979</span>
<span class="go">2020-10-20 04:51:50,893 - fba.count - INFO - Number of cell barcodes detected: 11,758</span>
<span class="go">2020-10-20 04:51:50,894 - fba.count - INFO - Number of features detected: 2</span>
<span class="go">2020-10-20 05:00:42,298 - fba.count - INFO - Total UMIs after deduplication: 7,145,799</span>
<span class="go">2020-10-20 05:00:42,320 - fba.count - INFO - Median number of UMIs per cell: 477.0</span>
<span class="go">2020-10-20 05:00:42,434 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="demultiplexing">
<h2>Demultiplexing<a class="headerlink" href="#demultiplexing" title="Permalink to this headline">¶</a></h2>
<section id="gaussian-mixture-model">
<h3>Gaussian mixture model<a class="headerlink" href="#gaussian-mixture-model" title="Permalink to this headline">¶</a></h3>
<p>The implementation of demultiplexing method <code class="docutils literal notranslate"><span class="pre">2</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is inspired by the method described on <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/crispr">10x Genomics’ website</a>. Use <code class="docutils literal notranslate"><span class="pre">-p</span></code> to set the probability threshold for demulitplexing (default <code class="docutils literal notranslate"><span class="pre">0.9</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    -dm <span class="m">2</span> <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-10-04 14:14:15,659 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2021-10-04 14:14:15,659 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-10-04 14:14:15,659 - fba.__main__ - INFO - Python version: 3.8</span>
<span class="go">2021-10-04 14:14:15,659 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2021-10-04 14:14:36,166 - fba.__main__ - INFO - Skipping arguments: "-q/--quantile", "-cm/--clustering_method"</span>
<span class="go">2021-10-04 14:14:36,166 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2021-10-04 14:14:36,166 - fba.demultiplex - INFO - Demultiplexing method: 2</span>
<span class="go">2021-10-04 14:14:36,166 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2021-10-04 14:14:36,167 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2021-10-04 14:14:36,167 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2021-10-04 14:14:36,167 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2021-10-04 14:14:37,875 - fba.demultiplex - INFO - Number of cells: 11,758</span>
<span class="go">2021-10-04 14:14:37,875 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2021-10-04 14:14:37,920 - fba.demultiplex - INFO - Number of features: 2 / 2 (after filtering / original in the matrix)</span>
<span class="go">2021-10-04 14:14:37,920 - fba.demultiplex - INFO - Features: NON_TARGET-1 RAB1A-2</span>
<span class="go">2021-10-04 14:14:37,920 - fba.demultiplex - INFO - Total UMIs: 7,145,799 / 7,145,799</span>
<span class="go">2021-10-04 14:14:37,942 - fba.demultiplex - INFO - Median number of UMIs per cell: 477.0 / 477.0</span>
<span class="go">2021-10-04 14:14:37,942 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2021-10-04 14:14:38,418 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2021-10-04 14:14:42,078 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2021-10-04 14:15:24,288 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells. Each column represents a single cell.</p>
<a class="reference internal image-reference" href="tutorials/crispr_screening/PRJNA358686/Pyplot_heatmap_cells_demultiplexed_gm.png"><img alt="Heatmap" class="align-center" src="tutorials/crispr_screening/PRJNA358686/Pyplot_heatmap_cells_demultiplexed_gm.png" style="width: 700px;"/></a>
<p>t-SNE embedding of cells based on the abundance of features (sgRNAs, no transcriptome information used). Colors indicate the sgRNA status for each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="tutorials/crispr_screening/PRJNA358686/Pyplot_embedding_cells_demultiplexed_gm.png"><img alt="t-SNE embedding" class="align-center" src="tutorials/crispr_screening/PRJNA358686/Pyplot_embedding_cells_demultiplexed_gm.png" style="width: 500px;"/></a>
<p>UMI distribution and model fitting threshold:</p>
<a class="reference internal image-reference" href="tutorials/crispr_screening/PRJNA358686/Pyplot_feature_umi_distribution_gm.png"><img alt="UMI distribution" class="align-center" src="tutorials/crispr_screening/PRJNA358686/Pyplot_feature_umi_distribution_gm.png" style="width: 800px;"/></a>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../../cell_surface_protein_labeling/index.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Cell surface protein labeling</div>
              </div>
              <svg><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">
              <svg><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA</div>
                
              </div>
            </a>
        </div>

        <div class="related-information">
              Copyright &#169; 2020-2022, Jialei Duan
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            Contents
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">CROP-Seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a><ul>
<li><a class="reference internal" href="#preparation">Preparation</a></li>
<li><a class="reference internal" href="#clustering">Clustering</a></li>
<li><a class="reference internal" href="#barcode-extraction">Barcode extraction</a></li>
<li><a class="reference internal" href="#matrix-generation">Matrix generation</a></li>
<li><a class="reference internal" href="#demultiplexing">Demultiplexing</a><ul>
<li><a class="reference internal" href="#gaussian-mixture-model">Gaussian mixture model</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/scripts/main.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    </body>
</html>