<!doctype html>
<html class="no-js" lang="en" data-content_root="../../../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://jlduan.github.io/fba/_build/html/tutorials/crispr_screening/PRJNA358686/tutorial.html" />
<meta property="og:site_name" content="fba" />
<meta property="og:description" content="Dataset: CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T) Datlinger, P., Rendeiro, A.F., Schmidl, C., Krausgruber, T., Traxler, P., Klughammer, J., Schuster, L.C., Kuchler, A., Alpar, D., and Bock, C. (2017). Pooled CRISPR screening with single-cell transcriptome re..." />
<meta property="og:image" content="https://raw.githubusercontent.com/jlduan/fba/gh-pages/docs/_static/logo.svg" />
<meta property="og:image:alt" content="fba" />
<meta name="description" content="Dataset: CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T) Datlinger, P., Rendeiro, A.F., Schmidl, C., Krausgruber, T., Traxler, P., Klughammer, J., Schuster, L.C., Kuchler, A., Alpar, D., and Bock, C. (2017). Pooled CRISPR screening with single-cell transcriptome re..." />
<meta property="og:ignore_canonical" content="true" />
<meta property="article:author" content="Jialei Duan" />
<meta property="og:site_name" content="FBA" />
<meta name="twitter:card" content="summary"></meta>
<meta name="twitter:creator" content="@jlduan" />
<meta name="twitter:image" content="https://github.com/jlduan.png" />
<link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" /><link rel="next" title="Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3’ sgRNA-CR1cs1" href="../PRJNA609688/tutorial.html" /><link rel="prev" title="10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA" href="../SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html" />

    <link rel="shortcut icon" href="../../../_static/logo.ico"/><!-- Generated with Sphinx 7.4.5 and Furo 2024.07.18 -->
        <title>CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T) - FBA</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a3258c64" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=613ab9ff" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #eeeedd;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
      }
    }
  }
</style>


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-X0MEL2T23M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-X0MEL2T23M');
</script>

</head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">FBA</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo only-light" src="../../../_static/logo.svg" alt="Light Logo"/>
    <img class="sidebar-logo only-dark" src="../../../_static/logo_dark.svg" alt="Dark Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user_guide.html">User Guide</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current has-children"><a class="reference internal" href="../index.html">CRISPR screening</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of CRISPR screening</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../PRJNA609688/tutorial.html">Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3' sgRNA-CR1&lt;sup&gt;cs1&lt;/sup&gt;</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_surface_protein_labeling/index.html">Cell surface protein labeling</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Cell surface protein labeling</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA393315/tutorial.html">CITE-seq; 8k Cord Blood Mononuclear Cells with 13 Antibodies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA658075/tutorial.html">ASAP-seq; Multiplexed CRISPR Perturbations in Primary T Cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/SC3_v3_NextGem_DI_PBMC_CSP_1K/tutorial.html">1k Human PBMCs Stained with a Panel of TotalSeq B Antibodies, Dual Indexed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../eccite-seq/index.html">ECCITE-seq</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of ECCITE-seq</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../eccite-seq/PRJNA521522/tutorial.html">6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../phage-atac/index.html">PHAGE-ATAC</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of PHAGE-ATAC</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../phage-atac/PRJNA661457/tutorial.html">PHAGE-ATAC; Anti-CD8 Phage Hashing Single-cell ATAC-seq Using CD8 T Cells from Four Human Donors</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cellplex/index.html">CellPlex</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of CellPlex</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Nuclei_30K_Multiplex/tutorial.html">30k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Nuclei Multiplexed, 12 CMOs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_hashing/index.html">Cell hashing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of Cell hashing</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_hashing/PRJNA423077/tutorial.html">Peripheral Blood Mononuclear cells with 8 Antibodies</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../multi-seq/index.html">MULTI-seq</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of MULTI-seq</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../multi-seq/PRJNA531855/tutorial.html">15k HEK293 and 40k HMECs Multiplexed by Lipid- and Cholesterol-tagged Indices</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../targeted_transcript_enrichment/index.html">Targeted transcript enrichment</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of Targeted transcript enrichment</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../targeted_transcript_enrichment/Targeted_NGSC3_DI_HodgkinsLymphoma_GeneSignature/tutorial.html">Hodgkin’s Lymphoma, Dissociated Tumor: Targeted, Gene Signature Panel</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../pseudo-bulk/index.html">Pseudo-bulk</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of Pseudo-bulk</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/jlduan/fba">GitHub Repository</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="https://github.com/jlduan/fba/blob/gh-pages/docs/tutorials/crispr_screening/PRJNA358686/tutorial.rst?plain=true" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div><div class="edit-this-page">
  <a class="muted-link" href="https://github.com/jlduan/fba/edit/gh-pages/docs/tutorials/crispr_screening/PRJNA358686/tutorial.rst" title="Edit this page">
    <svg><use href="#svg-pencil"></use></svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="crop-seq-1-1-1-mixture-of-dnmt3b-mbd1-and-tet2-knockout-cell-lines-hek293t">
<span id="tutorial-crispr-screening-prjna358686"></span><h1>CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)<a class="headerlink" href="#crop-seq-1-1-1-mixture-of-dnmt3b-mbd1-and-tet2-knockout-cell-lines-hek293t" title="Link to this heading">¶</a></h1>
<p><strong>Dataset</strong>: CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout
Cell Lines (HEK293T)</p>
<p>Datlinger, P., Rendeiro, A.F., Schmidl, C., Krausgruber, T., Traxler,
P., Klughammer, J., Schuster, L.C., Kuchler, A., Alpar, D., and Bock, C.
(2017). <a class="reference external" href="https://doi.org/10.1038/nmeth.4177">Pooled CRISPR screening with single-cell transcriptome
readout</a>. <em>Nat. Methods</em> <strong>14</strong>, 297–301.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="preparation">
<h2>Preparation<a class="headerlink" href="#preparation" title="Link to this heading">¶</a></h2>
<p>Download fastq files from <a class="reference external" href="https://www.ebi.ac.uk/ena/browser/view/PRJNA521522?show=reads">European Nucleotide Archive</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>curl<span class="w"> </span>-O<span class="w"> </span>ftp.sra.ebi.ac.uk/vol1/fastq/SRR516/009/SRR5163029/SRR5163029_1.fastq.gz

<span class="gp">$ </span>curl<span class="w"> </span>-O<span class="w"> </span>ftp.sra.ebi.ac.uk/vol1/fastq/SRR516/009/SRR5163029/SRR5163029_2.fastq.gz
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Link to this heading">¶</a></h2>
<p>This dataset comprises single cell RNA-seq data obtained from HEK293T
cell lines that had been knocked out for DNMT3B, MBD1, and TET2,
separately and were mixed in equal proportions (<a class="reference external" href="https://www.nature.com/articles/nmeth.4177/figures/1">Fig. 1h</a>). The
platform used for this dataset is Drop-seq, and more details about the
original data processing can be found <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2450588">here</a>. In brief, the raw data was
initially processed with Drop-seq Tools v1.12 pipeline. The first 12
bases of read 1 are cell barcodes, followed by 8 bases of UMIs, while
captured single cell transcripts are in read 2.</p>
<p>To perform the downstream analysis, I reprocessed the raw reads to obtain
cell-associated barcodes. The raw reads were trimmed using <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore">Trim
Galore!</a> and then mapped to the human reference genome
<a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/release-notes/build#GRCh38_2020A">refdata-gex-GRCh38-2020-A</a> using STAR (version 2.7.8a). The plasmid
<a class="reference external" href="https://www.addgene.org/86708/">CROPseq-Guide-Puro</a> sequence was not included in the reference.</p>
<div class="table-wrapper colwidths-given docutils container" id="id10">
<table class="docutils align-default" id="id10">
<caption><span class="caption-text">Mapping statistics</span><a class="headerlink" href="#id10" title="Link to this table">¶</a></caption>
<colgroup>
<col style="width: 77.8%" />
<col style="width: 22.2%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Number of Read Pairs</p></td>
<td><p>227,621,653</p></td>
</tr>
<tr class="row-even"><td><p>Number of Read Pairs, After Trimming</p></td>
<td><p>201,527,916</p></td>
</tr>
<tr class="row-odd"><td><p>Reads Mapped to Genome: Unique+Multiple</p></td>
<td><p>0.7654</p></td>
</tr>
<tr class="row-even"><td><p>Reads Mapped to Genome: Unique</p></td>
<td><p>0.696169</p></td>
</tr>
<tr class="row-odd"><td><p>Estimated Number of Cells</p></td>
<td><p>1,199</p></td>
</tr>
<tr class="row-even"><td><p>Fraction of Reads in Cells</p></td>
<td><p>0.479281</p></td>
</tr>
<tr class="row-odd"><td><p>Mean Reads per Cell</p></td>
<td><p>45,290</p></td>
</tr>
<tr class="row-even"><td><p>Median Reads per Cell</p></td>
<td><p>16,329</p></td>
</tr>
<tr class="row-odd"><td><p>Mean UMI per Cell</p></td>
<td><p>2,391</p></td>
</tr>
<tr class="row-even"><td><p>Median UMI per Cell</p></td>
<td><p>1,118</p></td>
</tr>
<tr class="row-odd"><td><p>Mean Genes per Cell</p></td>
<td><p>1,042</p></td>
</tr>
<tr class="row-even"><td><p>Median Genes per Cell</p></td>
<td><p>716</p></td>
</tr>
<tr class="row-odd"><td><p>Total Genes Detected</p></td>
<td><p>16,068</p></td>
</tr>
</tbody>
</table>
</div>
<p>Inspect cell barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat<span class="w"> </span>cell_barcodes.txt

<span class="go">AACGGGCATGGG</span>
<span class="go">AACTGGGCATGG</span>
<span class="go">AAGACAGCGTGT</span>
<span class="go">AAGGGCGTACTC</span>
<span class="go">AATAAATACAAA</span>
<span class="go">AATAAATACAAC</span>
<span class="go">AATAAATACAAG</span>
<span class="go">AATAAATACAAT</span>
<span class="go">AATCAATCGCAA</span>
<span class="go">AATCAATCGCAC</span>
</pre></div>
</div>
<p>Prepare feature barcodes. sgRNA sequences can be found in Supplementary
Table 1.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat<span class="w"> </span>feature_barcodes.tsv

<span class="go">DNMT3B  CAGGATTGGGGGCGAGTCGG</span>
<span class="go">MBD1    ATAGGTGTCTGAGCGTCCAC</span>
<span class="go">TET2    CAGGACTCACACGACTATTC</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="approach-a">
<h2>Approach A<a class="headerlink" href="#approach-a" title="Link to this heading">¶</a></h2>
<p>As sgRNAs are captured together with other transcripts through polyA
tails, their locations on read 2 may vary (it should be noted that this
is not an sgRNA enrichment library). To expedite processing, we first
screened reads that contained the constant sequence (<code class="docutils literal notranslate"><span class="pre">GACGAAACACCG</span></code>)
upstream of sgRNAs using <a class="reference external" href="https://github.com/marcelm/cutadapt">cutadapt</a> (version 3.7).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cutadapt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cores<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--front<span class="w"> </span>GACGAAACACCG<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--length<span class="w"> </span><span class="m">25</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--minimum-length<span class="w"> </span><span class="m">25</span>:25<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--trimmed-only<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output<span class="w"> </span>read_2_trimmed.fq.gz<span class="w"> </span>--paired-output<span class="w"> </span>read_1_trimmed.fq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>../SRR5163029_2.fastq.gz<span class="w"> </span>../SRR5163029_1.fastq.gz
</pre></div>
</div>
<p>Preview the filtering result: 1,429,437 out of 227,621,653 (0.6%) read
pairs are kept for sgRNA identification.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">== Read fate breakdown ==</span>
<span class="go">Pairs that were too short:              25,972 (0.0%)</span>
<span class="go">Pairs discarded as untrimmed:      226,166,244 (99.4%)</span>
<span class="go">Pairs written (passing filters):     1,429,437 (0.6%)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="qc">
<h3>QC<a class="headerlink" href="#qc" title="Link to this heading">¶</a></h3>
<p>The first 100,000 read pairs are sampled (default, set by <code class="docutils literal notranslate"><span class="pre">-n</span></code>) for
quality control. By default, diagnostic results and plots are generated
in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>), and the full
length of read 1 and read 2 are searched against reference cell and
feature barcodes, respectively. The per base content of both read pairs
and the distribution of matched barcode positions are summarized. Use
<code class="docutils literal notranslate"><span class="pre">-r1_c</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-r2_c</span></code> to limit the search range, and <code class="docutils literal notranslate"><span class="pre">-cb_n</span></code>
and/or <code class="docutils literal notranslate"><span class="pre">-fb_n</span></code> to set the mismatch tolerance for cell and/or feature
barcode matching (default <code class="docutils literal notranslate"><span class="pre">3</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>qc<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-1<span class="w"> </span>read_1_trimmed.fq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-2<span class="w"> </span>read_2_trimmed.fq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-w<span class="w"> </span>cell_barcodes.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-f<span class="w"> </span>feature_barcodes.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-r1_c<span class="w"> </span><span class="m">0</span>,12
</pre></div>
</div>
<p>This library was constructed using the Drop-seq platform, where the
first 12 bases correspond to cell barcodes, and the following 8 bases
represent UMIs. Based on the base content plot, the GC content of the
cell barcodes is evenly distributed. However, the UMIs show a slight
T-enrichment.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_per_base_seq_content4.webp"><img alt="../../../_images/Pyplot_read1_per_base_seq_content4.webp" class="align-center" src="../../../_images/Pyplot_read1_per_base_seq_content4.webp" style="width: 350px;" />
</a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_barcodes_starting_ending2.webp"><img alt="../../../_images/Pyplot_read1_barcodes_starting_ending2.webp" class="align-center" src="../../../_images/Pyplot_read1_barcodes_starting_ending2.webp" style="width: 350px;" />
</a>
<p>Regarding read 2, the GC content of sgRNAs is uniformly distributed. It
should be noted that the first 20 bases correspond to the sgRNA
sequences.</p>
<p><a class="reference internal" href="../../../_images/Pyplot_read2_per_base_seq_content4.webp"><img alt="pic1" src="../../../_images/Pyplot_read2_per_base_seq_content4.webp" style="width: 48%;" /></a> <a class="reference internal" href="../../../_images/Pyplot_read2_barcodes_starting_ending4.webp"><img alt="pic2" src="../../../_images/Pyplot_read2_barcodes_starting_ending4.webp" style="width: 48%;" /></a></p>
<p>The detailed <code class="docutils literal notranslate"><span class="pre">qc</span></code> results are stored in the
<code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. The <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns
indicate the matched positions on reads, while the
<code class="docutils literal notranslate"><span class="pre">matching_description</span></code> columns indicate mismatches in the format of
substitutions:insertions:deletions.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>zcat<span class="w"> </span>feature_barcoding_output.tsv.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span>no_match<span class="w"> </span><span class="p">|</span><span class="w"> </span>head

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">GCTGCATAGTCGggggggatttttt       TTCATAGCTCCG    2:12    1:0:2   CAGGACTCACACGACTATTCGTTTT       TET2_CAGGACTCACACGACTATTC       0:20    0:0:0</span>
<span class="go">GTTGCTCCTCACggtgatttttttt       GTTCCCTCCCAC    0:12    1:1:1   CAGGACTCACACGACTATTCGTTTT       TET2_CAGGACTCACACGACTATTC       0:20    0:0:0</span>
<span class="go">TAATGTTTAGGGagggcgctttttt       TAATGTTTAGGG    0:12    0:0:0   ATAGGTGTCTGAGCGTCCACGTTTT       MBD1_ATAGGTGTCTGAGCGTCCAC       0:20    0:0:0</span>
<span class="go">TCTTCCACTACCggtatgacttttt       TCTTCCACTACC    0:12    0:0:0   CAGGATTGGGGGCGAGTCGGGTTTT       DNMT3B_CAGGATTGGGGGCGAGTCGG     0:20    0:0:0</span>
<span class="go">GGAATGCCTTGAgtatacttttttt       GGAATGCCTTGA    0:12    0:0:0   CAGGACTCACACGACTATTCGTTTT       TET2_CAGGACTCACACGACTATTC       0:20    0:0:0</span>
<span class="go">GCGATCACAATGtaatagatttttt       GCGATCACAATG    0:12    0:0:0   CAGGATTGGGGGCGAGTCGGGTTTT       DNMT3B_CAGGATTGGGGGCGAGTCGG     0:20    0:0:0</span>
<span class="go">CGCCGTCGGACAcgaatcctttttt       CCGTAGCGGGCA    2:12    1:0:2   ATAGGTGTCTGAGCGTCCACGTTTT       MBD1_ATAGGTGTCTGAGCGTCCAC       0:20    0:0:0</span>
<span class="go">CCGTCCTAGTTGatcccagtttttt       CCGTCCTAGTTG    0:12    0:0:0   CAGGACTCACACGACTATTCGTTTT       TET2_CAGGACTCACACGACTATTC       0:20    0:0:0</span>
<span class="go">ATTGTTCCATCTgtcggcttttttt       ACTGTTTGATCT    0:12    3:0:0   ATAGGTGTCTGAGCGTCCACGTTTT       MBD1_ATAGGTGTCTGAGCGTCCAC       0:20    0:0:0</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="barcode-extraction">
<h3>Barcode extraction<a class="headerlink" href="#barcode-extraction" title="Link to this heading">¶</a></h3>
<p>Search ranges are set to <code class="docutils literal notranslate"><span class="pre">0,12</span></code> on read 1 and <code class="docutils literal notranslate"><span class="pre">0,20</span></code> on read 2. One
mismatch for cell and feature barcodes (<code class="docutils literal notranslate"><span class="pre">-cb_m</span></code>, <code class="docutils literal notranslate"><span class="pre">-cf_m</span></code>) are
allowed.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>extract<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-1<span class="w"> </span>read_1_trimmed.fq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-2<span class="w"> </span>read_2_trimmed.fq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-w<span class="w"> </span>cell_barcodes.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-f<span class="w"> </span>feature_barcodes.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-o<span class="w"> </span>feature_barcoding_output.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-r1_c<span class="w"> </span><span class="m">0</span>,12<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-r2_c<span class="w"> </span><span class="m">0</span>,20<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-cb_m<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-fb_m<span class="w"> </span><span class="m">1</span>
</pre></div>
</div>
<p>Preview of result.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip<span class="w"> </span>-dc<span class="w"> </span>feature_barcoding_output.tsv.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>head

<span class="go">read1_seq       cell_barcode    cb_num_mismatches       read2_seq       feature_barcode fb_num_mismatches</span>
<span class="go">TAATGTTTAGGGagggcgctttttt       TAATGTTTAGGG    0       ATAGGTGTCTGAGCGTCCACgtttt       MBD1_ATAGGTGTCTGAGCGTCCAC       0</span>
<span class="go">TCTTCCACTACCggtatgacttttt       TCTTCCACTACC    0       CAGGATTGGGGGCGAGTCGGgtttt       DNMT3B_CAGGATTGGGGGCGAGTCGG     0</span>
<span class="go">GGAATGCCTTGAgtatacttttttt       GGAATGCCTTGA    0       CAGGACTCACACGACTATTCgtttt       TET2_CAGGACTCACACGACTATTC       0</span>
<span class="go">GCGATCACAATGtaatagatttttt       GCGATCACAATG    0       CAGGATTGGGGGCGAGTCGGgtttt       DNMT3B_CAGGATTGGGGGCGAGTCGG     0</span>
<span class="go">CCGTCCTAGTTGatcccagtttttt       CCGTCCTAGTTG    0       CAGGACTCACACGACTATTCgtttt       TET2_CAGGACTCACACGACTATTC       0</span>
<span class="go">ATTATATGTGAGcagactttttttt       ATTATATGTGAG    0       ATAGGTGTCTGAGCGTCCACgtttt       MBD1_ATAGGTGTCTGAGCGTCCAC       0</span>
<span class="go">TTTCAGTATTGGggcgaattttttt       TTTCAGTATTGG    0       ATAGGTGTCTGAGCGTCCACgtttt       MBD1_ATAGGTGTCTGAGCGTCCAC       0</span>
<span class="go">GTTCCCTCCCAAacatgagtttttt       GTTCCCTCCCAA    0       CAGGATTGGGGGCGAGTCGGgtttt       DNMT3B_CAGGATTGGGGGCGAGTCGG     0</span>
<span class="go">GCTCCGCTTTTAactcaagtttttt       GCTCCGCTTTTA    0       CAGGATTGGGGCCGAGTCGGgactt       DNMT3B_CAGGATTGGGGGCGAGTCGG     1</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>9,213 out of 1,429,437 read pairs have valid cell and feature barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-07 16:11:53,295 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-07 16:11:53,295 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-07 16:11:53,295 - fba.__main__ - INFO - Python version: 3.10</span>
<span class="go">2022-03-07 16:11:53,295 - fba.__main__ - INFO - Using extract subcommand ...</span>
<span class="go">2022-03-07 16:11:53,310 - fba.levenshtein - INFO - Number of reference cell barcodes: 1,199</span>
<span class="go">2022-03-07 16:11:53,310 - fba.levenshtein - INFO - Number of reference feature barcodes: 3</span>
<span class="go">2022-03-07 16:11:53,310 - fba.levenshtein - INFO - Read 1 coordinates to search: [0, 12)</span>
<span class="go">2022-03-07 16:11:53,310 - fba.levenshtein - INFO - Read 2 coordinates to search: [0, 20)</span>
<span class="go">2022-03-07 16:11:53,310 - fba.levenshtein - INFO - Cell barcode maximum number of mismatches: 1</span>
<span class="go">2022-03-07 16:11:53,310 - fba.levenshtein - INFO - Feature barcode maximum number of mismatches: 1</span>
<span class="go">2022-03-07 16:11:53,312 - fba.levenshtein - INFO - Read 1 maximum number of N allowed: 3</span>
<span class="go">2022-03-07 16:11:53,312 - fba.levenshtein - INFO - Read 2 maximum number of N allowed: 3</span>
<span class="go">2022-03-07 16:11:53,337 - fba.levenshtein - INFO - Matching ...</span>
<span class="go">2022-03-07 16:12:13,951 - fba.levenshtein - INFO - Number of read pairs processed: 1,429,437</span>
<span class="go">2022-03-07 16:12:13,952 - fba.levenshtein - INFO - Number of read pairs w/ valid barcodes: 9,213</span>
<span class="go">2022-03-07 16:12:13,954 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="matrix-generation">
<h3>Matrix generation<a class="headerlink" href="#matrix-generation" title="Link to this heading">¶</a></h3>
<p>Only fragments with correctly matched cell and feature barcodes are
included, while fragments with UMI lengths less than the specified value
are discarded. UMI removal is performed using UMI-tools (<a class="reference external" href="http://www.genome.org/cgi/doi/10.1101/gr.209601.116">Smith, T., et
al. 2017. Genome Res. 27, 491–499.</a>), with the starting position on
read 1 set by <code class="docutils literal notranslate"><span class="pre">-us</span></code> (default <code class="docutils literal notranslate"><span class="pre">16</span></code>) and the length set by <code class="docutils literal notranslate"><span class="pre">-ul</span></code>
(default <code class="docutils literal notranslate"><span class="pre">12</span></code>). The UMI deduplication method can be set using <code class="docutils literal notranslate"><span class="pre">-ud</span></code>
(default <code class="docutils literal notranslate"><span class="pre">directional</span></code>), and the UMI deduplication mismatch threshold
can be specified using <code class="docutils literal notranslate"><span class="pre">-um</span></code> (default <code class="docutils literal notranslate"><span class="pre">1</span></code>).</p>
<p>The generated feature count matrix can be easily imported into
well-established single cell analysis packages such as <a class="reference external" href="https://satijalab.org/seurat/">Seurat</a> and <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-i<span class="w"> </span>feature_barcoding_output.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-o<span class="w"> </span>matrix_featurecount.csv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-us<span class="w"> </span><span class="m">12</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-ul<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>Result summary.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-08 13:43:27,499 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-08 13:43:27,499 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-08 13:43:27,499 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2022-03-08 13:43:27,499 - fba.__main__ - INFO - Using count subcommand ...</span>
<span class="go">2022-03-08 13:43:28,183 - fba.count - INFO - UMI-tools version: 1.1.1</span>
<span class="go">2022-03-08 13:43:28,184 - fba.count - INFO - UMI starting position on read 1: 12</span>
<span class="go">2022-03-08 13:43:28,184 - fba.count - INFO - UMI length: 8</span>
<span class="go">2022-03-08 13:43:28,184 - fba.count - INFO - UMI-tools deduplication threshold: 1</span>
<span class="go">2022-03-08 13:43:28,184 - fba.count - INFO - UMI-tools deduplication method: directional</span>
<span class="go">2022-03-08 13:43:28,184 - fba.count - INFO - Header line: read1_seq cell_barcode cb_num_mismatches read2_seq feature_barcode fb_num_mismatches</span>
<span class="go">2022-03-08 13:43:28,194 - fba.count - INFO - Number of lines processed: 9,213</span>
<span class="go">2022-03-08 13:43:28,194 - fba.count - INFO - Number of cell barcodes detected: 420</span>
<span class="go">2022-03-08 13:43:28,194 - fba.count - INFO - Number of features detected: 3</span>
<span class="go">2022-03-08 13:43:28,194 - fba.count - INFO - UMI deduplicating ...</span>
<span class="go">2022-03-08 13:43:28,202 - fba.count - INFO - Total UMIs after deduplication: 1,089</span>
<span class="go">2022-03-08 13:43:28,202 - fba.count - INFO - Median number of UMIs per cell: 1.0</span>
<span class="go">2022-03-08 13:43:28,204 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="demultiplexing">
<h3>Demultiplexing<a class="headerlink" href="#demultiplexing" title="Link to this heading">¶</a></h3>
<section id="gaussian-mixture-model">
<h4>Gaussian mixture model<a class="headerlink" href="#gaussian-mixture-model" title="Link to this heading">¶</a></h4>
<p>The implementation of demultiplexing method <code class="docutils literal notranslate"><span class="pre">2</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is
inspired by the method described on the <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/crispr">10x Genomics’ website</a>. To set
the probability threshold for demultiplexing, use <code class="docutils literal notranslate"><span class="pre">-p</span></code> (default
<code class="docutils literal notranslate"><span class="pre">0.9</span></code>). To specify the minimum number of positive cells for a given
feature to be considered during demultiplexing, use <code class="docutils literal notranslate"><span class="pre">-nc</span></code> (default
<code class="docutils literal notranslate"><span class="pre">200</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>demultiplex<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-i<span class="w"> </span>matrix_featurecount.csv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-dm<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-v<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-nc<span class="w"> </span><span class="m">0</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-07 19:57:14,925 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-07 19:57:14,925 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-07 19:57:14,925 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2022-03-07 19:57:14,925 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2022-03-07 19:57:17,564 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;</span>
<span class="go">2022-03-07 19:57:17,564 - fba.demultiplex - INFO - Output directory: demultiplexed_gm</span>
<span class="go">2022-03-07 19:57:17,564 - fba.demultiplex - INFO - Demultiplexing method: 2</span>
<span class="go">2022-03-07 19:57:17,564 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2022-03-07 19:57:17,564 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2022-03-07 19:57:17,564 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2022-03-07 19:57:17,564 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2022-03-07 19:57:17,571 - fba.demultiplex - INFO - Number of cells: 420</span>
<span class="go">2022-03-07 19:57:17,571 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 0</span>
<span class="go">2022-03-07 19:57:17,572 - fba.demultiplex - INFO - Number of features: 3 / 3 (after filtering / original in the matrix)</span>
<span class="go">2022-03-07 19:57:17,572 - fba.demultiplex - INFO - Features: DNMT3B MBD1 TET2</span>
<span class="go">2022-03-07 19:57:17,572 - fba.demultiplex - INFO - Total UMIs: 1,081 / 1,081</span>
<span class="go">2022-03-07 19:57:17,573 - fba.demultiplex - INFO - Median number of UMIs per cell: 1.0 / 1.0</span>
<span class="go">2022-03-07 19:57:17,573 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2022-03-07 19:57:18,277 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2022-03-07 19:57:18,423 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2022-03-07 19:57:21,922 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_trimmed_gm.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_trimmed_gm.png" style="width: 700px;" />
</a>
<p>Preview the demultiplexing result: the numbers of singlets and
multiplets.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;demultiplexed/matrix_cell_identity.csv.gz&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
<span class="n">DNMT3B</span>    <span class="mi">141</span>
<span class="n">MBD1</span>      <span class="mi">150</span>
<span class="n">TET2</span>      <span class="mi">158</span>
<span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="mi">74</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="knee-point">
<h4>Knee point<a class="headerlink" href="#knee-point" title="Link to this heading">¶</a></h4>
<p>Cells are demultiplexed according to the abundance of features,
specifically sgRNAs. Demultiplexing method <code class="docutils literal notranslate"><span class="pre">5</span></code> is implemented to use
the local maxima on the difference curve to detemine the knee point on
the UMI saturation curve.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>demultiplex<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-i<span class="w"> </span>matrix_featurecount.csv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-dm<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-v<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-nc<span class="w"> </span><span class="m">0</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-05 01:52:38,900 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-05 01:52:38,900 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-05 01:52:38,900 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2022-03-05 01:52:38,900 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2022-03-05 01:52:41,396 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;, &quot;-p/--prob&quot;</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Demultiplexing method: 5</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2022-03-05 01:52:41,403 - fba.demultiplex - INFO - Number of cells: 523</span>
<span class="go">2022-03-05 01:52:41,403 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 0</span>
<span class="go">2022-03-05 01:52:41,404 - fba.demultiplex - INFO - Number of features: 3 / 3 (after filtering / original in the matrix)</span>
<span class="go">2022-03-05 01:52:41,404 - fba.demultiplex - INFO - Features: DNMT3B MBD1 TET2</span>
<span class="go">2022-03-05 01:52:41,404 - fba.demultiplex - INFO - Total UMIs: 1,364 / 1,364</span>
<span class="go">2022-03-05 01:52:41,405 - fba.demultiplex - INFO - Median number of UMIs per cell: 1.0 / 1.0</span>
<span class="go">2022-03-05 01:52:41,405 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2022-03-05 01:52:41,810 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2022-03-05 01:52:41,979 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2022-03-05 01:52:44,840 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_trimmed_knee.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_trimmed_knee.png" style="width: 700px;" />
</a>
<p>Preview the demultiplexing result: the numbers of singlets and
multiplets.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;demultiplexed/matrix_cell_identity.csv.gz&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
<span class="n">DNMT3B</span>    <span class="mi">141</span>
<span class="n">MBD1</span>      <span class="mi">150</span>
<span class="n">TET2</span>      <span class="mi">158</span>
<span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="mi">74</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
</section>
<section id="approach-b">
<h2>Approach B<a class="headerlink" href="#approach-b" title="Link to this heading">¶</a></h2>
<p>Rather than pre-filtering read 2 for the constant upstream region of
sgRNA, we conduct a search for sgRNAs throughout the entire read 2.
Although this approach may take longer, it provides more comprehensive
results. To optimize speed, we suggest splitting fastq files and running
them on different nodes simultaneously.</p>
<section id="id1">
<h3>Barcode extraction<a class="headerlink" href="#id1" title="Link to this heading">¶</a></h3>
<p>The <a class="reference external" href="https://www.addgene.org/86708/">CROPseq-Guide-Puro</a> transcripts captured by Drop-seq beads in
single cell RNA-seq library contain sgRNA sequences. As there are no
secondary libraries for sgRNA enrichment, we need to extract the sgRNA
sequences from read 2. However, the locations of sgRNAs on read 2 vary
because the transcripts are captured by polyA tails. Therefore, we use
the <code class="docutils literal notranslate"><span class="pre">qc</span></code> mode for sgRNA extraction, which is capable of handling the
variable locations of sgRNAs on read 2.</p>
<p>To specify the number of reads to analyze, use <code class="docutils literal notranslate"><span class="pre">--num_reads</span></code>, where
<code class="docutils literal notranslate"><span class="pre">None</span></code> means all reads. To set the number of threads, use <code class="docutils literal notranslate"><span class="pre">-t</span></code>. By
default, the diagnostic results and plots are generated in the <code class="docutils literal notranslate"><span class="pre">qc</span></code>
directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>), and the full length of read 1
and read 2 are searched against reference cell and feature barcodes,
respectively. The per base content of both read pairs and the
distribution of matched barcode positions are summarized. To limit the
search range for read 1 and/or read 2, use <code class="docutils literal notranslate"><span class="pre">-r1_c</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-r2_c</span></code>,
respectively. Use <code class="docutils literal notranslate"><span class="pre">-cb_n</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-fb_n</span></code> to set the mismatch
tolerance for cell and/or feature barcode matching (default <code class="docutils literal notranslate"><span class="pre">3</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>qc<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-1<span class="w"> </span>SRR5163029_1.fastq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-2<span class="w"> </span>SRR5163029_2.fastq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-w<span class="w"> </span>cell_barcodes.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-f<span class="w"> </span>feature_barcodes.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-cb_m<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-fb_m<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-cb_n<span class="w"> </span><span class="m">15</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-fb_n<span class="w"> </span><span class="m">15</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-r1_c<span class="w"> </span><span class="m">0</span>,12<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-t<span class="w"> </span><span class="nv">$SLURM_CPUS_ON_NODE</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--num_reads<span class="w"> </span>None
</pre></div>
</div>
<p>The detailed <code class="docutils literal notranslate"><span class="pre">qc</span></code> results are stored in the
<code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. The <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns
indicate the matched positions on reads, while the
<code class="docutils literal notranslate"><span class="pre">matching_description</span></code> columns indicate mismatches in the format of
substitutions:insertions:deletions.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip<span class="w"> </span>-dc<span class="w"> </span>qc/feature_barcoding_output.tsv.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>head

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">TTTAGGATCGTTtgatgtattttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttcttctttcttttttattctttacaacatcctaccataacata no_match        NA      NA      ATTAAAAATATTGTGGCAGGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAACAAAAAAAAATCAGCTATATAACCACTAATACTTCTA    NA      NA      NA</span>
<span class="go">GTCGAAACTCTTaacgggatttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt no_match        NA      NA      TTATAATGGTTACAAATAAAGCAATAGCATCACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA    NA      NA      NA</span>
<span class="go">GTTTACGTGTTCatgggcgattttttttttttttttttttttttttttttaaaaaagttaaaagggggcccgtggggggacaaatagaggggcctagagttccaccccccatcccacaaaaaaaaccctcaccgcacagggcctcgcccct GTTTACGTGTTC    0:12    0:0:0   GGAGTACGGAGAATTCTATAAGAGCTTGACCAATGACTGGGAAGATCACTTGGCAGTGAAGCATTTTTCAGTTGAAGGACAGTTGGAATTCAGAGCCCTTCTATTTGTCCCACGACGTGCTCCTTTTGATCTGTTTGAAAAAAAAAAAAAA    no_match        NA      NA</span>
<span class="go">CCGTCCTAGTTGgtgtatattttttttgtttttttttttttttcaccgggtcagagctgcccctaagtaccacgtcccgtcccacctttatcggacctcggccaccacaaattgcttatccagagtgcccccctccgcccatcccagactc CCGTCCTAGTTG    0:12    0:0:0   AATTAAGTCTCGTAAAGAACGAGAAGCTGAACTTGGACCTAGGGCAACCGACTTCACCAATGTTTACAGCGAGAATCTTGGTGACGACGTGGATGATGAGCGCCTTAAGGTTCTCTTTGGCAAGTTTGGGCCTGCCTTGAGTGTGCGACTT    no_match        NA      NA</span>
<span class="go">TTTCAGTATTGGggcgaattttttttttttttttttttttttttttttttttttttttttttttggctagtttttttgtggtttttgcttttggttctctcgtttgccctggagctcccaggtccctttcttgtcctaccataggtaaccc TTTCAGTATTGG    0:12    0:0:0   GGACGAAACACCGATAGGTGTCTGAGCGTCCACGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAGCTTGGCGTAACTAGATCTTGAGACACTGCTTTT MBD1_ATAGGTGTCTGAGCGTCCAC       13:33 0:0:0</span>
<span class="go">CTAGGTACCACTagacagtttttttttttttttttttttttttttttttttttttttttttctctatgtgtgcttttttttggctttagtctgtgggtccctagttagccccggcgcccccacgcgcagaacgtgtcttaccacaagaacc CTAGGTACCACT    0:12    0:0:0   TTCTTGGGTAGTTTGCAGTTTTTAAAATTATGTTTTAAAATGGACTATCATATGCTTACCGTAACTTGAAAGTATTTCGATTTCTTGGCTTTATATATCTTGTGGAACGGACGAAACACCGATAGGTGTCTGAGCGTCCACGTTTTAGAGC MBD1_ATAGGTGTCTGAGCGTCCAC       121:1410:0:0</span>
<span class="go">TCTTCCACTACCgtcccgtcttttttttttttttttttttttttttttttttttttttctttatgtcagttttttttgtgctttagtattgggttcccttgtttgcccgagggctcccaggcccagatttgggctaaccaaagggaccccg TCTTCCACTACC    0:12    0:0:0   ACCGATAGGTGTCTGAGCGTCCACGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTAAGCTTGGCGTAACTAGATCTTGAGACACTGCTTTTTGCTTGTAC MBD1_ATAGGTGTCTGAGCGTCCAC       4:24  0:0:0</span>
<span class="go">CTTAATTTGGTGggaagattttttttttttttttttttttttttttttaagtactttaagtaagctttttttaggctttagccgtgggttcccctgttagcccgggaggtccccgggcccaatctgggcctaacagagaggccccgtacaa CTTAATTTGGTG    0:12    0:0:0   CCGTAACTTGAAAGTATTTCGATTTCTTGGCTTTATATATCTTGTGGAAAGGACGAAACACCGCAGGACTCACACGACTCTTCGTTTTAGAGCTAGCAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGT TET2_CAGGACTCACACGACTATTC       63:83 1:0:0</span>
<span class="go">TCGTACATACGGtggtttttttttttttttttttttttttttttttttttttttttttttttttttgtttttttttttttttgtttttttttttgtgtcctttgttttcactggggctcccaggtccatatccggtgttaccagagaaacc TCGTACATACGG    0:12    0:0:0   ATCATATGCTTACCGTAACTTGAAAGTATTTCGATTTCTTGGCTTTATATATCTTGTGGAAAGGACGAAACACCGCAGGATTGGGGGCGAGTCGGGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAA DNMT3B_CAGGATTGGGGGCGAGTCGG     75:95 0:0:0</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id2">
<h3>Matrix generation<a class="headerlink" href="#id2" title="Link to this heading">¶</a></h3>
<p>Only fragments with correctly matched cell and feature barcodes are
included, while fragments with UMI lengths less than the specified value
are discarded. UMI removal is performed using UMI-tools (<a class="reference external" href="http://www.genome.org/cgi/doi/10.1101/gr.209601.116">Smith, T., et
al. 2017. Genome Res. 27, 491–499.</a>), with the starting position on
read 1 set by <code class="docutils literal notranslate"><span class="pre">-us</span></code> (default <code class="docutils literal notranslate"><span class="pre">16</span></code>) and the length set by <code class="docutils literal notranslate"><span class="pre">-ul</span></code>
(default <code class="docutils literal notranslate"><span class="pre">12</span></code>). The UMI deduplication method can be set using <code class="docutils literal notranslate"><span class="pre">-ud</span></code>
(default <code class="docutils literal notranslate"><span class="pre">directional</span></code>), and the UMI deduplication mismatch threshold
can be specified using <code class="docutils literal notranslate"><span class="pre">-um</span></code> (default <code class="docutils literal notranslate"><span class="pre">1</span></code>).</p>
<p>The generated feature count matrix can be easily imported into
well-established single cell analysis packages: <a class="reference external" href="https://satijalab.org/seurat/">Seurat</a> and <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-i<span class="w"> </span>feature_barcoding_output.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-o<span class="w"> </span>matrix_featurecount.csv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-us<span class="w"> </span><span class="m">12</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-ul<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>11.76 % (1,364 out of 11,597) of read pairs with valid cell and feature
barcodes are unique fragments.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-04 23:18:27,501 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-04 23:18:27,501 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-04 23:18:27,501 - fba.__main__ - INFO - Python version: 3.10</span>
<span class="go">2022-03-04 23:18:27,501 - fba.__main__ - INFO - Using count subcommand ...</span>
<span class="go">2022-03-04 23:18:31,494 - fba.count - INFO - UMI-tools version: 1.1.2</span>
<span class="go">2022-03-04 23:18:31,495 - fba.count - INFO - UMI start position on read 1 auto-detected, overriding -us</span>
<span class="go">2022-03-04 23:18:31,495 - fba.count - INFO - UMI length: 8</span>
<span class="go">2022-03-04 23:18:31,496 - fba.count - INFO - UMI-tools deduplication threshold: 1</span>
<span class="go">2022-03-04 23:18:31,496 - fba.count - INFO - UMI-tools deduplication method: directional</span>
<span class="go">2022-03-04 23:18:31,496 - fba.count - INFO - Header line: read1_seq cell_barcode cb_matching_pos cb_matching_description read2_seq feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">2022-03-04 23:18:31,581 - fba.count - INFO - Number of lines processed: 11,597</span>
<span class="go">2022-03-04 23:18:31,581 - fba.count - INFO - Number of cell barcodes detected: 523</span>
<span class="go">2022-03-04 23:18:31,582 - fba.count - INFO - Number of features detected: 3</span>
<span class="go">2022-03-04 23:18:31,608 - fba.count - INFO - Total UMIs after deduplication: 1,364</span>
<span class="go">2022-03-04 23:18:31,609 - fba.count - INFO - Median number of UMIs per cell: 1.0</span>
<span class="go">2022-03-04 23:18:31,615 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id6">
<h3>Demultiplexing<a class="headerlink" href="#id6" title="Link to this heading">¶</a></h3>
<section id="id7">
<h4>Gaussian mixture model<a class="headerlink" href="#id7" title="Link to this heading">¶</a></h4>
<p>The implementation of demultiplexing method <code class="docutils literal notranslate"><span class="pre">2</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is
inspired by the method described on the <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/crispr">10x Genomics’ website</a>. To set
the probability threshold for demultiplexing, use <code class="docutils literal notranslate"><span class="pre">-p</span></code> (default
<code class="docutils literal notranslate"><span class="pre">0.9</span></code>). To specify the minimum number of positive cells for a given
feature to be considered during demultiplexing, use <code class="docutils literal notranslate"><span class="pre">-nc</span></code> (default
<code class="docutils literal notranslate"><span class="pre">200</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>demultiplex<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-i<span class="w"> </span>matrix_featurecount.csv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-dm<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-v<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-nc<span class="w"> </span><span class="m">0</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-04 23:19:05,218 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-04 23:19:05,219 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-04 23:19:05,219 - fba.__main__ - INFO - Python version: 3.10</span>
<span class="go">2022-03-04 23:19:05,219 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2022-03-04 23:19:15,199 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;</span>
<span class="go">2022-03-04 23:19:15,200 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2022-03-04 23:19:15,201 - fba.demultiplex - INFO - Demultiplexing method: 2</span>
<span class="go">2022-03-04 23:19:15,201 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2022-03-04 23:19:15,201 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2022-03-04 23:19:15,201 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2022-03-04 23:19:15,201 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2022-03-04 23:19:15,219 - fba.demultiplex - INFO - Number of cells: 523</span>
<span class="go">2022-03-04 23:19:15,219 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 0</span>
<span class="go">2022-03-04 23:19:15,222 - fba.demultiplex - INFO - Number of features: 3 / 3 (after filtering / original in the matrix)</span>
<span class="go">2022-03-04 23:19:15,222 - fba.demultiplex - INFO - Features: DNMT3B MBD1 TET2</span>
<span class="go">2022-03-04 23:19:15,222 - fba.demultiplex - INFO - Total UMIs: 1,364 / 1,364</span>
<span class="go">2022-03-04 23:19:15,223 - fba.demultiplex - INFO - Median number of UMIs per cell: 1.0 / 1.0</span>
<span class="go">2022-03-04 23:19:15,223 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2022-03-04 23:19:17,319 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2022-03-04 23:19:17,784 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2022-03-04 23:19:32,256 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_original_gm.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_original_gm.png" style="width: 700px;" />
</a>
<p>Preview the demultiplexing result: the numbers of singlets and
multiplets.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;demultiplexed/matrix_cell_identity.csv.gz&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
<span class="n">DNMT3B</span>    <span class="mi">141</span>
<span class="n">MBD1</span>      <span class="mi">150</span>
<span class="n">TET2</span>      <span class="mi">158</span>
<span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="mi">74</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="id9">
<h4>Knee point<a class="headerlink" href="#id9" title="Link to this heading">¶</a></h4>
<p>Cells are demultiplexed according to the abundance of features,
specifically sgRNAs. Demultiplexing method <code class="docutils literal notranslate"><span class="pre">5</span></code> is implemented to use
the local maxima on the difference curve to detemine the knee point on
the UMI saturation curve.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba<span class="w"> </span>demultiplex<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-i<span class="w"> </span>matrix_featurecount.csv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-dm<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-v<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-nc<span class="w"> </span><span class="m">0</span>
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-03-05 01:52:38,900 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-03-05 01:52:38,900 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-03-05 01:52:38,900 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2022-03-05 01:52:38,900 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2022-03-05 01:52:41,396 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;, &quot;-p/--prob&quot;</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Demultiplexing method: 5</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2022-03-05 01:52:41,396 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2022-03-05 01:52:41,403 - fba.demultiplex - INFO - Number of cells: 523</span>
<span class="go">2022-03-05 01:52:41,403 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 0</span>
<span class="go">2022-03-05 01:52:41,404 - fba.demultiplex - INFO - Number of features: 3 / 3 (after filtering / original in the matrix)</span>
<span class="go">2022-03-05 01:52:41,404 - fba.demultiplex - INFO - Features: DNMT3B MBD1 TET2</span>
<span class="go">2022-03-05 01:52:41,404 - fba.demultiplex - INFO - Total UMIs: 1,364 / 1,364</span>
<span class="go">2022-03-05 01:52:41,405 - fba.demultiplex - INFO - Median number of UMIs per cell: 1.0 / 1.0</span>
<span class="go">2022-03-05 01:52:41,405 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2022-03-05 01:52:41,810 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2022-03-05 01:52:41,979 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2022-03-05 01:52:44,840 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_original_knee.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_original_knee.png" style="width: 700px;" />
</a>
<p>Preview the demultiplexing result: the numbers of singlets and
multiplets.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;demultiplexed/matrix_cell_identity.csv.gz&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
<span class="n">DNMT3B</span>    <span class="mi">141</span>
<span class="n">MBD1</span>      <span class="mi">150</span>
<span class="n">TET2</span>      <span class="mi">158</span>
<span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="mi">74</span>
</pre></div>
</div>
<p>UMI distribution and knee point detection:</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_knee_DNMT3B.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_knee_DNMT3B.webp" style="width: 400px;" />
</a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_knee_MBD1.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_knee_MBD1.webp" style="width: 400px;" />
</a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_knee_TET2.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_knee_TET2.webp" style="width: 400px;" />
</a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../PRJNA609688/tutorial.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3’ sgRNA-CR1<sup>cs1</sup></div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2020-2024, Jialei Duan
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/jlduan/fba" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a><ul>
<li><a class="reference internal" href="#preparation">Preparation</a></li>
<li><a class="reference internal" href="#mapping">Mapping</a></li>
<li><a class="reference internal" href="#approach-a">Approach A</a><ul>
<li><a class="reference internal" href="#qc">QC</a></li>
<li><a class="reference internal" href="#barcode-extraction">Barcode extraction</a></li>
<li><a class="reference internal" href="#matrix-generation">Matrix generation</a></li>
<li><a class="reference internal" href="#demultiplexing">Demultiplexing</a><ul>
<li><a class="reference internal" href="#gaussian-mixture-model">Gaussian mixture model</a></li>
<li><a class="reference internal" href="#knee-point">Knee point</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#approach-b">Approach B</a><ul>
<li><a class="reference internal" href="#id1">Barcode extraction</a></li>
<li><a class="reference internal" href="#id2">Matrix generation</a></li>
<li><a class="reference internal" href="#id6">Demultiplexing</a><ul>
<li><a class="reference internal" href="#id7">Gaussian mixture model</a></li>
<li><a class="reference internal" href="#id9">Knee point</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../../../_static/documentation_options.js?v=b6a3c597"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/furo.js?v=5fa4622c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=3b13dea8"></script>
    </body>
</html>