<!doctype html>
<html class="no-js">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta property="og:title" content="6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs" />
  <meta property="og:type" content="website" />
  <meta property="og:url" content="tutorials/eccite-seq/PRJNA521522/tutorial.html" />
  <meta property="og:description" content="Dataset: ECCITE-seq Mimitou, E.P., Cheng, A., Montalbano, A., Hao, S., Stoeckius, M., Legut, M., Roush, T., Herrera, A., Papalexi, E., Ouyang, Z., et al. (2019). Multiplexed detection of proteins, ..." />
  <link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" /><link rel="next" title="Targeted transcript enrichment" href="../../targeted_transcript_enrichment/index.html" /><link rel="prev" title="ECCITE-Seq" href="../index.html" />

    <link rel="shortcut icon" href="../../../_static/logo.svg"/><meta name="generator" content="sphinx-4.3.2, furo 2021.10.09"/>
        <title>6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs - FBA</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?digest=0254c309f5cadf746f1a613e7677379ac9c8cdcd" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?digest=16fb25fabf47304eee183a5e9af80b1ba98259b1" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  body[data-theme="dark"] {
    --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
  }
  @media (prefers-color-scheme: dark) {
    body:not([data-theme="light"]) {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
  }
</style>


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-X0MEL2T23M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-X0MEL2T23M');
</script>

</head>
  <body>
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" />
      <line x1="4" y1="6" x2="20" y2="6" />
      <line x1="10" y1="12" x2="20" y2="12" />
      <line x1="6" y1="18" x2="20" y2="18" />
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">FBA</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand centered" href="../../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../../_static/logo.svg" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user_guide.html">User Guide</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 has-children"><a class="reference internal" href="../../crispr_screening/index.html">CRISPR screening</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/PRJNA358686/tutorial.html">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../crispr_screening/PRJNA609688/tutorial.html">Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3' sgRNA-CR1&lt;sup&gt;cs1&lt;/sup&gt;</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_surface_protein_labeling/index.html">Cell surface protein labeling</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA393315/tutorial.html">CITE-seq; 8k Cord Blood Mononuclear Cells with 13 Antibodies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/SC3_v3_NextGem_DI_PBMC_CSP_1K/tutorial.html">1k Human PBMCs Stained with a Panel of TotalSeq B Antibodies, Dual Indexed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../phage-atac/index.html">PHAGE-ATAC</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../phage-atac/PRJNA661457/tutorial.html">PHAGE-ATAC; Anti-CD8 Phage Hashing Single-cell ATAC-seq Using CD8 T Cells from Four Human Donors</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cellplex/index.html">CellPlex</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Nuclei_30K_Multiplex/tutorial.html">30k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Nuclei Multiplexed, 12 CMOs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_hashing/index.html">Cell hashing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_hashing/PRJNA423077/tutorial.html">Peripheral Blood Mononuclear cells with 8 Antibodies</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../multi-seq/index.html">MULTI-seq</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../multi-seq/PRJNA531855/tutorial.html">15k HEK293 and 40k HMECs Multiplexed by Lipid- and Cholesterol-tagged Indices</a></li>
</ul>
</li>
<li class="toctree-l2 current has-children"><a class="reference internal" href="../index.html">ECCITE-seq</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../targeted_transcript_enrichment/index.html">Targeted transcript enrichment</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../targeted_transcript_enrichment/Targeted_NGSC3_DI_HodgkinsLymphoma_GeneSignature/tutorial.html">Hodgkin’s Lymphoma, Dissociated Tumor: Targeted, Gene Signature Panel</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../pseudo-bulk/index.html">Pseudo-bulk</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/jlduan/fba">GitHub Repository</a></li>
</ul>

</div>
</div>
      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="k-single-cell-multimodal-readout-of-nih-3t3-myla-sez4-and-pbmcs">
<span id="tutorial-eccite-seq-prjna521522"></span><h1>6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs<a class="headerlink" href="#k-single-cell-multimodal-readout-of-nih-3t3-myla-sez4-and-pbmcs" title="Permalink to this headline">¶</a></h1>
<p>Dataset: ECCITE-seq</p>
<p>Mimitou, E.P., Cheng, A., Montalbano, A., Hao, S., Stoeckius, M., Legut, M., Roush, T., Herrera, A., Papalexi, E., Ouyang, Z., et al. (2019). <a class="reference external" href="https://doi.org/10.1038/s41592-019-0392-0">Multiplexed detection of proteins, transcriptomes, clonotypes and CRISPR perturbations in single cells</a>. Nat. Methods <em>16</em>, 409–412.</p>
<div class="line-block">
<div class="line"><br/></div>
</div>
<section id="preparation">
<h2>Preparation<a class="headerlink" href="#preparation" title="Permalink to this headline">¶</a></h2>
<p>Download fastq files from <a class="reference external" href="https://www.ebi.ac.uk/ena/browser/view/PRJNA521522?show=reads">European Nucleotide Archive</a>.</p>
<p>Hashtag (Cell hashing):</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>curl -O ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR855/007/SRR8550947/SRR8550947_1.fastq.gz

<span class="gp">$ </span>curl -O ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR855/007/SRR8550947/SRR8550947_2.fastq.gz
</pre></div>
</div>
<p>Protein-tag (Cite-seq):</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>curl -O ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR855/006/SRR8550946/SRR8550946_1.fastq.gz

<span class="gp">$ </span>curl -O ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR855/006/SRR8550946/SRR8550946_2.fastq.gz
</pre></div>
</div>
<p>Guide-tag (sgRNAs):</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>curl -O ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR855/008/SRR8550948/SRR8550948_1.fastq.gz

<span class="gp">$ </span>curl -O ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR855/008/SRR8550948/SRR8550948_2.fastq.gz
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
<p>Download pre-processed transcriptome matrix from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3596084">Gene Expression Omnibus</a>. We will need the cell-associated barcodes, which are determined by the transcriptomes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3596nnn/GSM3596084/suppl/GSM3596084_mx-cDNA.txt.gz
</pre></div>
</div>
<p>Inspect cell barcodes:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc GSM3596084_mx-cDNA.txt.gz <span class="p">|</span> head -1 <span class="p">|</span> sed <span class="s1">'s/"//g'</span> <span class="p">|</span> sed <span class="s1">'s/ /\n/g'</span> <span class="p">|</span> sort <span class="p">|</span> grep -B1 <span class="m">1</span>

<span class="go">CAGATCACACGTAAGG</span>
<span class="go">CAGATCACACGTAAGG.1</span>
<span class="go">--</span>
<span class="go">CGCTATCAGCCGCCTA</span>
<span class="go">CGCTATCAGCCGCCTA.1</span>
<span class="go">--</span>
<span class="go">TTTGCGCCAGTTCATG</span>
<span class="go">TTTGCGCCAGTTCATG.1</span>
</pre></div>
</div>
<p>It seems there are 3 colliding barcodes. We will use the first ones.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc GSM3596084_mx-cDNA.txt.gz <span class="p">|</span> head -1 <span class="p">|</span> sed <span class="s1">'s/"//g'</span> <span class="p">|</span> sed <span class="s1">'s/ /\n/g'</span> <span class="p">|</span> sort <span class="p">|</span> grep -v <span class="m">1</span> &gt; cell_barcodes.txt

<span class="gp">$ </span>head cell_barcodes.txt

<span class="go">AAACCTGAGTGGTAGC</span>
<span class="go">AAACCTGAGTGTTTGC</span>
<span class="go">AAACCTGAGTTAGGTA</span>
<span class="go">AAACCTGCAAGTTGTC</span>
<span class="go">AAACCTGCAATTCCTT</span>
<span class="go">AAACCTGCACAGACTT</span>
<span class="go">AAACCTGCACATCCGG</span>
<span class="go">AAACCTGCACGGTAAG</span>
<span class="go">AAACCTGCAGACACTT</span>
<span class="go">AAACCTGCATCCGCGA</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
<section id="hashtag">
<h2>Hashtag<a class="headerlink" href="#hashtag" title="Permalink to this headline">¶</a></h2>
<section id="id1">
<h3>Preparation<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>Prepare feature barcodes (hashtag-oligo sequences, from <a class="reference external" href="https://www.nature.com/articles/s41592-019-0392-0#MOESM1">Supplementary Table 4 and 5</a>, legend of <a class="reference external" href="https://www.nature.com/articles/s41592-019-0392-0#MOESM1">Supplementary Figure 1</a>):</p>
<blockquote>
<div><p>NIH-3T3 cells were split into 7 tubes and stained with 7 barcoded hashing antibodies (Hashtag-A to Hashtag-G), followed by washing and pooling. MyLa, Sez4 and PBMCs were stained with Hashtag_1, Hashtag_2 and Hashtag_3 respectively.</p>
</div></blockquote>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat feature_barcodes_hashtag.tsv

<span class="go">Hashtag_1       ACATGTTACCGT</span>
<span class="go">Hashtag_2       AGCTTACTATCC</span>
<span class="go">Hashtag_3       TATCACATCGGT</span>
<span class="go">Hashtag_A       AGGACCATCCAA</span>
<span class="go">Hashtag_B       TCGATAATGCGA</span>
<span class="go">Hashtag_C       GAGGCTGAGCTA</span>
<span class="go">Hashtag_D       GTGTGACGTATT</span>
<span class="go">Hashtag_E       ACTGTCTAACGG</span>
<span class="go">Hashtag_F       CACATAATGACG</span>
<span class="go">Hashtag_G       TAACGACGTGGT</span>
</pre></div>
</div>
</section>
<section id="qc">
<h3>QC<a class="headerlink" href="#qc" title="Permalink to this headline">¶</a></h3>
<p>Sample the first 100,000 (set by <code class="docutils literal notranslate"><span class="pre">-n</span></code>) read pairs for quality control. Use <code class="docutils literal notranslate"><span class="pre">-t</span></code> to set the number of threads. The diagnostic results and plots are generated in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>, default <code class="docutils literal notranslate"><span class="pre">qc</span></code>). By default, full length of read 1 and read 2 are searched against reference cell and feature barcodes, respectively. The per base content of both read pairs and the distribution of matched barcode positions are summarized. Use <code class="docutils literal notranslate"><span class="pre">-r1_c</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-r2_c</span></code> to limit the search range. Use <code class="docutils literal notranslate"><span class="pre">-cb_n</span></code> and/or <code class="docutils literal notranslate"><span class="pre">-fb_n</span></code> to set the mismatch tolerance for cell and feature barcode matching (default <code class="docutils literal notranslate"><span class="pre">3</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba qc <span class="se">\</span>
    -1 SRR8550947_1.fastq.gz <span class="se">\</span>
    -2 SRR8550947_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes_hashtag.tsv <span class="se">\</span>
    --output_directory qc
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 16:32:43,465 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 16:32:43,465 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 16:32:43,465 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 16:32:43,465 - fba.__main__ - INFO - Using qc subcommand ...</span>
<span class="go">2022-01-08 16:32:44,108 - fba.qc - INFO - Summarizing per base read content ...</span>
<span class="go">2022-01-08 16:32:44,108 - fba.qc - INFO - Number of read pairs to analyze: 100,000</span>
<span class="go">2022-01-08 16:32:44,108 - fba.qc - INFO - Output directory: qc</span>
<span class="go">2022-01-08 16:32:44,429 - fba.qc - INFO - Number of reads processed: 100,000</span>
<span class="go">2022-01-08 16:32:47,187 - fba.regex - INFO - regex version: 2.5.91</span>
<span class="go">2022-01-08 16:32:47,192 - fba.regex - INFO - Number of reference cell barcodes: 6,871</span>
<span class="go">2022-01-08 16:32:47,192 - fba.regex - INFO - Number of reference feature barcodes: 10</span>
<span class="go">2022-01-08 16:32:47,192 - fba.regex - INFO - Cell barcode maximum number of mismatches: 3</span>
<span class="go">2022-01-08 16:32:47,192 - fba.regex - INFO - Feature barcode maximum number of mismatches: 3</span>
<span class="go">2022-01-08 16:32:47,192 - fba.regex - INFO - Read 1 maximum number of N allowed: inf</span>
<span class="go">2022-01-08 16:32:47,192 - fba.regex - INFO - Read 2 maximum number of N allowed: inf</span>
<span class="go">2022-01-08 16:32:47,192 - fba.regex - INFO - Number of read pairs to analyze: 100,000</span>
<span class="go">2022-01-08 16:32:48,448 - fba.regex - INFO - Number of threads: 72</span>
<span class="go">2022-01-08 16:32:48,448 - fba.regex - INFO - Chunk size: 50,000</span>
<span class="go">2022-01-08 16:32:48,449 - fba.regex - INFO - Matching ...</span>
<span class="go">2022-01-08 16:33:55,391 - fba.regex - INFO - Read pairs processed: 50,000</span>
<span class="go">2022-01-08 16:35:02,689 - fba.regex - INFO - Read pairs processed: 100,000</span>
<span class="go">2022-01-08 16:35:04,179 - fba.qc - INFO - Summarizing barcode coordinates ...</span>
<span class="go">2022-01-08 16:35:04,179 - fba.qc - INFO - Output directory: qc</span>
<span class="go">2022-01-08 16:35:05,431 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>For read 1, the first 16 bases are cell barcodes and the following 10 bases are UMIs (Read 1 length is 26). Based on the base content plot, the GC content of cell barcodes are quite even. The UMIs are slightly T enriched.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_per_base_seq_content_hashtag.png"><img alt="../../../_images/Pyplot_read1_per_base_seq_content_hashtag.png" class="align-center" src="../../../_images/Pyplot_read1_per_base_seq_content_hashtag.png" style="width: 390px;"/></a>
<p>As for read 2, based on the per base content, it suggests that bases 0-12 are actually our feature barcodes (See the distribution of matched barcode positions on read 2).</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_per_base_seq_content_hashtag.png"><img alt="../../../_images/Pyplot_read2_per_base_seq_content_hashtag.png" class="align-center" src="../../../_images/Pyplot_read2_per_base_seq_content_hashtag.png" style="width: 400px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_barcodes_starting_ending_hashtag.png"><img alt="../../../_images/Pyplot_read2_barcodes_starting_ending_hashtag.png" class="align-center" src="../../../_images/Pyplot_read2_barcodes_starting_ending_hashtag.png" style="width: 400px;"/></a>
<p>The detailed qc results are stored in <code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns indicate the matched positions on reads. <code class="docutils literal notranslate"><span class="pre">matching_description</span></code> columns indicate mismatches in <code class="docutils literal notranslate"><span class="pre">substitutions:insertions:deletions</span></code> format.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc qc/feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">NGACGGCGTGTGACGAACGCGCGCCT      GACGCGTGTGCGAAAC        1:18    0:2:1   ACATGTTACCGTCCCATATAAGAAAAGGCGCGCGTTCGT Hashtag_1_ACATGTTACCGT  0:12    0:0:0</span>
<span class="go">NCTACACCACGGTAGAGACCTAGGTC      CACCACTGTGAGTGAC        4:19    2:0:1   AGGACCATCCAACCCATATAAGAAAGACCTAGGTCTCTA Hashtag_A_AGGACCATCCAA  0:12    0:0:0</span>
<span class="go">GCAAACTAGATGGCGTCGACGCTTAG      GCAAACTAGATGGCGT        0:16    0:0:0   AGCTTACTATCCCCCATATAGAAGCTAAGCGTCGACGCC Hashtag_2_AGCTTACTATCC  0:12    0:0:0</span>
<span class="go">CCTTCGAAGTGCCATTCTTTCACTGG      CCTTCGAAGTGCCATT        0:16    0:0:0   TATCACATCGGTCCCATATAAGAAACCAGTGAAAGAATG Hashtag_3_TATCACATCGGT  0:12    0:0:0</span>
<span class="go">NGATCTGGTATGAAACGATCAGGTCA      AGATCTGGTATGAAAC        0:16    1:0:0   AGCTTACTATCCCCCATATAAGAAATGACCTGATCGTTT Hashtag_2_AGCTTACTATCC  0:12    0:0:0</span>
<span class="go">NTCGGGATCTGTGCAAATCGGGTAGT      CTCGGGATCTGTGCAA        0:16    1:0:0   AGCTTACTATCCCCCATATAAGAAACTACCCGATTTGCA Hashtag_2_AGCTTACTATCC  0:12    0:0:0</span>
<span class="go">GCGCGATGTACTTAGCTGCGTAGGTG      GCGCGATGTACTTAGC        0:16    0:0:0   TATCACATCGGTCCCATATAAGAAACACCTACGCAGCTA Hashtag_3_TATCACATCGGT  0:12    0:0:0</span>
<span class="go">AACCGCGCACACAGAGCGTTTGGCCG      AACCGCGCACACAGAG        0:16    0:0:0   ACATGTTACCGTCCCATATAAGAAACGGCCAAACGCTCT Hashtag_1_ACATGTTACCGT  0:12    0:0:0</span>
<span class="go">NAATGAACATGCGCACACGATAGTTT      no_match        NA      NA      TATCACATCGGTCCCATATAAGAAAAAACTATCGTGTGC NA      NA      NA</span>
</pre></div>
</div>
</section>
<section id="barcode-extraction">
<h3>Barcode extraction<a class="headerlink" href="#barcode-extraction" title="Permalink to this headline">¶</a></h3>
<p>The lengths of cell and feature barcodes are all identical (16 and 12, respectively). And based on the <code class="docutils literal notranslate"><span class="pre">qc</span></code> results, the distributions of starting and ending positions of cell and feature barcodes are very uniform. Search ranges are set to <code class="docutils literal notranslate"><span class="pre">0,16</span></code> on read 1 and <code class="docutils literal notranslate"><span class="pre">0,12</span></code> on read 2. One mismatch for cell and feature barcodes (<code class="docutils literal notranslate"><span class="pre">-cb_m</span></code>, <code class="docutils literal notranslate"><span class="pre">-cf_m</span></code>) are allowed. And by default, three ambiguous nucleotides (Ns) for read 1 and read 2 (<code class="docutils literal notranslate"><span class="pre">-cb_n</span></code>, <code class="docutils literal notranslate"><span class="pre">-cf_n</span></code>) are allowed.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba extract <span class="se">\</span>
    -1 SRR8550947_1.fastq.gz <span class="se">\</span>
    -2 SRR8550947_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes_hashtag.tsv <span class="se">\</span>
    -o feature_barcoding_output.tsv.gz <span class="se">\</span>
    -r1_c <span class="m">0</span>,16 <span class="se">\</span>
    -r2_c <span class="m">0</span>,12 <span class="se">\</span>
    -cb_m <span class="m">1</span> <span class="se">\</span>
    -fb_m <span class="m">1</span> <span class="se">\</span>
    -cb_n <span class="m">3</span> <span class="se">\</span>
    -fb_n <span class="m">3</span>
</pre></div>
</div>
<p>Preview of result.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_num_mismatches       read2_seq       feature_barcode fb_num_mismatches</span>
<span class="go">NGACGGCGTGTGACGAacgcgcgcct      TGACGGCGTGTGACGA        1       ACATGTTACCGTcccatataagaaaaggcgcgcgttcgt Hashtag_1_ACATGTTACCGT  0</span>
<span class="go">NCTACACCACGGTAGAgacctaggtc      CCTACACCACGGTAGA        1       AGGACCATCCAAcccatataagaaagacctaggtctcta Hashtag_A_AGGACCATCCAA  0</span>
<span class="go">GCAAACTAGATGGCGTcgacgcttag      GCAAACTAGATGGCGT        0       AGCTTACTATCCcccatatagaagctaagcgtcgacgcc Hashtag_2_AGCTTACTATCC  0</span>
<span class="go">CCTTCGAAGTGCCATTctttcactgg      CCTTCGAAGTGCCATT        0       TATCACATCGGTcccatataagaaaccagtgaaagaatg Hashtag_3_TATCACATCGGT  0</span>
<span class="go">NGATCTGGTATGAAACgatcaggtca      AGATCTGGTATGAAAC        1       AGCTTACTATCCcccatataagaaatgacctgatcgttt Hashtag_2_AGCTTACTATCC  0</span>
<span class="go">NTCGGGATCTGTGCAAatcgggtagt      CTCGGGATCTGTGCAA        1       AGCTTACTATCCcccatataagaaactacccgatttgca Hashtag_2_AGCTTACTATCC  0</span>
<span class="go">GCGCGATGTACTTAGCtgcgtaggtg      GCGCGATGTACTTAGC        0       TATCACATCGGTcccatataagaaacacctacgcagcta Hashtag_3_TATCACATCGGT  0</span>
<span class="go">AACCGCGCACACAGAGcgtttggccg      AACCGCGCACACAGAG        0       ACATGTTACCGTcccatataagaaacggccaaacgctct Hashtag_1_ACATGTTACCGT  0</span>
<span class="go">TCAGATGAGAATGTTGgtggggcttc      TCAGATGAGAATGTTG        0       TATCACATCGGTcccatataagaaagaagccccaccaac Hashtag_3_TATCACATCGGT  0</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>57.0% (4,897,995 out of 8,591,807) of total read pairs have valid cell and feature barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 16:35:05,778 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 16:35:05,778 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 16:35:05,778 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 16:35:05,778 - fba.__main__ - INFO - Using extract subcommand ...</span>
<span class="go">2022-01-08 16:35:05,791 - fba.levenshtein - INFO - Number of reference cell barcodes: 6,871</span>
<span class="go">2022-01-08 16:35:05,791 - fba.levenshtein - INFO - Number of reference feature barcodes: 10</span>
<span class="go">2022-01-08 16:35:05,791 - fba.levenshtein - INFO - Read 1 coordinates to search: [0, 16)</span>
<span class="go">2022-01-08 16:35:05,791 - fba.levenshtein - INFO - Read 2 coordinates to search: [0, 12)</span>
<span class="go">2022-01-08 16:35:05,791 - fba.levenshtein - INFO - Cell barcode maximum number of mismatches: 1</span>
<span class="go">2022-01-08 16:35:05,792 - fba.levenshtein - INFO - Feature barcode maximum number of mismatches: 1</span>
<span class="go">2022-01-08 16:35:05,792 - fba.levenshtein - INFO - Read 1 maximum number of N allowed: 3</span>
<span class="go">2022-01-08 16:35:05,792 - fba.levenshtein - INFO - Read 2 maximum number of N allowed: 3</span>
<span class="go">2022-01-08 16:35:05,984 - fba.levenshtein - INFO - Matching ...</span>
<span class="go">2022-01-08 16:38:39,570 - fba.levenshtein - INFO - Number of read pairs processed: 8,591,807</span>
<span class="go">2022-01-08 16:38:39,572 - fba.levenshtein - INFO - Number of read pairs w/ valid barcodes: 4,897,995</span>
<span class="go">2022-01-08 16:38:39,582 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
</section>
<section id="matrix-generation">
<h3>Matrix generation<a class="headerlink" href="#matrix-generation" title="Permalink to this headline">¶</a></h3>
<p>Only fragments with valid (passed the criteria) cell and feature barcodes are included. UMI deduplication is powered by UMI-tools (<a class="reference external" href="http://www.genome.org/cgi/doi/10.1101/gr.209601.116">Smith, T., et al. 2017. Genome Res. 27, 491–499.</a>). Use <code class="docutils literal notranslate"><span class="pre">-us</span></code> to set the UMI starting position on read 1 (default <code class="docutils literal notranslate"><span class="pre">16</span></code>). Use <code class="docutils literal notranslate"><span class="pre">-ul</span></code> to set the UMI length (default <code class="docutils literal notranslate"><span class="pre">12</span></code>). Fragments with UMI length less than this value are discarded. Use <code class="docutils literal notranslate"><span class="pre">-um</span></code> to set mismatch threshold (default <code class="docutils literal notranslate"><span class="pre">1</span></code>). UMI deduplication method is set by <code class="docutils literal notranslate"><span class="pre">-ud</span></code> (default <code class="docutils literal notranslate"><span class="pre">directional</span></code>).</p>
<p>The generated feature count matrix can be easily imported into well-established single cell analysis packages: <a class="reference external" href="https://satijalab.org/seurat/">Seruat</a> and <a class="reference external" href="https://scanpy.readthedocs.io/en/stable">Scanpy</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba count <span class="se">\</span>
    -i feature_barcoding_output.tsv.gz <span class="se">\</span>
    -o matrix_featurecount.csv.gz <span class="se">\</span>
    -us <span class="m">16</span> <span class="se">\</span>
    -ul <span class="m">10</span> <span class="se">\</span>
    -um <span class="m">1</span> <span class="se">\</span>
    -ud directional
</pre></div>
</div>
<p>Result summary.</p>
<p>31.3% (1,531,088 out of 4,897,995) of read pairs with valid cell and feature barcodes are unique fragments. 17.8% (1,531,088 out of 8,591,807) of total sequenced read pairs contribute to the final matrix with an average of 55 UMIs per cell.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 16:41:49,871 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 16:41:49,871 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 16:41:49,871 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 16:41:49,871 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2022-01-08 16:42:01,202 - fba.__main__ - INFO - Skipping arguments: "-p/--prob"</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Demultiplexing method: 1</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2022-01-08 16:42:01,479 - fba.demultiplex - INFO - Number of cells: 6,871</span>
<span class="go">2022-01-08 16:42:01,479 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2022-01-08 16:42:01,498 - fba.demultiplex - INFO - Number of features: 10 / 10 (after filtering / original in the matrix)</span>
<span class="go">2022-01-08 16:42:01,498 - fba.demultiplex - INFO - Features: Hashtag_1 Hashtag_2 Hashtag_3 Hashtag_A Hashtag_B Hashtag_C Hashtag_D Hashtag_E Hashtag_F Hashtag_G</span>
<span class="go">2022-01-08 16:42:01,499 - fba.demultiplex - INFO - Total UMIs: 1,531,088 / 1,531,088</span>
<span class="go">2022-01-08 16:42:01,507 - fba.demultiplex - INFO - Median number of UMIs per cell: 55.0 / 55.0</span>
<span class="go">2022-01-08 16:42:01,507 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2022-01-08 16:44:11,296 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2022-01-08 16:44:17,314 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2022-01-08 16:44:28,444 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
</section>
<section id="demultiplexing">
<h3>Demultiplexing<a class="headerlink" href="#demultiplexing" title="Permalink to this headline">¶</a></h3>
<p>Cells are classified based on the abundance of features (hashtags, no transcriptome information used). Demultiplexing method <code class="docutils literal notranslate"><span class="pre">1</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is implemented based on the method described in <a class="reference external" href="https://doi.org/10.1186/s13059-018-1603-1">Stoeckius, M., et al. (2018)</a> with some modifications. A cell identity matrix is generated in the output directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>, default <code class="docutils literal notranslate"><span class="pre">demultiplexed</span></code>): 0 means negative, 1 means positive. Use <code class="docutils literal notranslate"><span class="pre">-q</span></code> to set the quantile threshold for demulitplexing (Default <code class="docutils literal notranslate"><span class="pre">0.9999</span></code>). Set <code class="docutils literal notranslate"><span class="pre">-v</span></code> to create visualization plots.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    --output_directory demultiplexed <span class="se">\</span>
    -dm <span class="m">1</span> <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 16:41:49,871 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 16:41:49,871 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 16:41:49,871 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 16:41:49,871 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2022-01-08 16:42:01,202 - fba.__main__ - INFO - Skipping arguments: "-p/--prob"</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Demultiplexing method: 1</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2022-01-08 16:42:01,203 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2022-01-08 16:42:01,479 - fba.demultiplex - INFO - Number of cells: 6,871</span>
<span class="go">2022-01-08 16:42:01,479 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2022-01-08 16:42:01,498 - fba.demultiplex - INFO - Number of features: 10 / 10 (after filtering / original in the matrix)</span>
<span class="go">2022-01-08 16:42:01,498 - fba.demultiplex - INFO - Features: Hashtag_1 Hashtag_2 Hashtag_3 Hashtag_A Hashtag_B Hashtag_C Hashtag_D Hashtag_E Hashtag_F Hashtag_G</span>
<span class="go">2022-01-08 16:42:01,499 - fba.demultiplex - INFO - Total UMIs: 1,531,088 / 1,531,088</span>
<span class="go">2022-01-08 16:42:01,507 - fba.demultiplex - INFO - Median number of UMIs per cell: 55.0 / 55.0</span>
<span class="go">2022-01-08 16:42:01,507 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2022-01-08 16:44:11,296 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2022-01-08 16:44:17,314 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2022-01-08 16:44:28,444 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (hashtags) across all cells. Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_hashtag.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_hashtag.png" style="width: 700px;"/></a>
<p>t-SNE embedding of cells based on the abundance of features (hashtags, no transcriptome information used). Colors indicate the hashtag status for each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed_hashtag.png"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed_hashtag.png" style="width: 500px;"/></a>
<p>Preview the demultiplexing result: the numbers of singlets, multiplets and negative cells. In summary, the numbers of MyLa, Sez4, PBMCs and NIH-3T3 cells demultiplexed are 324, 283, 914 and 4,518 respectively.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">'demultiplexed/matrix_cell_identity.csv.gz'</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">m</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span>
<span class="n">Hashtag_1</span>    <span class="mi">324</span>
<span class="n">Hashtag_2</span>    <span class="mi">283</span>
<span class="n">Hashtag_3</span>    <span class="mi">914</span>
<span class="n">Hashtag_A</span>    <span class="mi">673</span>
<span class="n">Hashtag_B</span>    <span class="mi">771</span>
<span class="n">Hashtag_C</span>    <span class="mi">709</span>
<span class="n">Hashtag_D</span>    <span class="mi">603</span>
<span class="n">Hashtag_E</span>    <span class="mi">707</span>
<span class="n">Hashtag_F</span>    <span class="mi">764</span>
<span class="n">Hashtag_G</span>    <span class="mi">291</span>
<span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="mi">341</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">5</span><span class="p">]:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">5</span><span class="p">]:</span> <span class="mi">491</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
</section>
<section id="protein-tag">
<h2>Protein-tag<a class="headerlink" href="#protein-tag" title="Permalink to this headline">¶</a></h2>
<section id="id2">
<h3>Preparation<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>Prepare feature barcodes (protein-tag sequences, from <a class="reference external" href="https://www.nature.com/articles/s41592-019-0392-0#MOESM1">Supplementary Table 3</a>, legend of <a class="reference external" href="https://www.nature.com/articles/s41592-019-0392-0#MOESM1">Supplementary Figure 1</a>):</p>
<blockquote>
<div><p>All cells were stained with a mix of anti-human CD29 and anti-mouse CD29 antibodies.</p>
</div></blockquote>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat feature_barcodes_CD29.tsv

<span class="go">hCD29   AATAGCGGAGCC</span>
<span class="go">mCD29   CGAAGACCAAGA</span>
</pre></div>
</div>
</section>
<section id="id3">
<h3>QC<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba qc <span class="se">\</span>
    -1 SRR8550946_1.fastq.gz <span class="se">\</span>
    -2 SRR8550946_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes_CD29.tsv <span class="se">\</span>
    --output_directory qc
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 12:29:00,323 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 12:29:00,323 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 12:29:00,323 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 12:29:00,323 - fba.__main__ - INFO - Using qc subcommand ...</span>
<span class="go">2022-01-08 12:29:00,896 - fba.qc - INFO - Summarizing per base read content ...</span>
<span class="go">2022-01-08 12:29:00,896 - fba.qc - INFO - Number of read pairs to analyze: 100,000</span>
<span class="go">2022-01-08 12:29:00,896 - fba.qc - INFO - Output directory: qc</span>
<span class="go">2022-01-08 12:29:01,119 - fba.qc - INFO - Number of reads processed: 100,000</span>
<span class="go">2022-01-08 12:29:03,848 - fba.regex - INFO - regex version: 2.5.91</span>
<span class="go">2022-01-08 12:29:03,852 - fba.regex - INFO - Number of reference cell barcodes: 6,871</span>
<span class="go">2022-01-08 12:29:03,852 - fba.regex - INFO - Number of reference feature barcodes: 2</span>
<span class="go">2022-01-08 12:29:03,852 - fba.regex - INFO - Cell barcode maximum number of mismatches: 3</span>
<span class="go">2022-01-08 12:29:03,852 - fba.regex - INFO - Feature barcode maximum number of mismatches: 3</span>
<span class="go">2022-01-08 12:29:03,852 - fba.regex - INFO - Read 1 maximum number of N allowed: inf</span>
<span class="go">2022-01-08 12:29:03,852 - fba.regex - INFO - Read 2 maximum number of N allowed: inf</span>
<span class="go">2022-01-08 12:29:03,852 - fba.regex - INFO - Number of read pairs to analyze: 100,000</span>
<span class="go">2022-01-08 12:29:05,062 - fba.regex - INFO - Number of threads: 72</span>
<span class="go">2022-01-08 12:29:05,062 - fba.regex - INFO - Chunk size: 50,000</span>
<span class="go">2022-01-08 12:29:05,062 - fba.regex - INFO - Matching ...</span>
<span class="go">2022-01-08 12:31:07,626 - fba.regex - INFO - Read pairs processed: 50,000</span>
<span class="go">2022-01-08 12:33:11,971 - fba.regex - INFO - Read pairs processed: 100,000</span>
<span class="go">2022-01-08 12:33:13,341 - fba.qc - INFO - Summarizing barcode coordinates ...</span>
<span class="go">2022-01-08 12:33:13,341 - fba.qc - INFO - Output directory: qc</span>
<span class="go">2022-01-08 12:33:14,203 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>As for read 2, based on the per base content, it suggests that bases 0-12 are actually our feature barcodes (See the distribution of matched barcode positions on read 2).</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_per_base_seq_content_protein-tag.png"><img alt="../../../_images/Pyplot_read2_per_base_seq_content_protein-tag.png" class="align-center" src="../../../_images/Pyplot_read2_per_base_seq_content_protein-tag.png" style="width: 400px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_barcodes_starting_ending_protein-tag.png"><img alt="../../../_images/Pyplot_read2_barcodes_starting_ending_protein-tag.png" class="align-center" src="../../../_images/Pyplot_read2_barcodes_starting_ending_protein-tag.png" style="width: 400px;"/></a>
<p>The detailed qc results are stored in <code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns indicate the matched positions on reads. <code class="docutils literal notranslate"><span class="pre">matching_description</span></code> columns indicate mismatches in <code class="docutils literal notranslate"><span class="pre">substitutions:insertions:deletions</span></code> format.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc feature_barcoding_output.tsv.gz <span class="p">|</span> head -20

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">NAGCCGATCACGCGGTCTGGTGGGCA      CAGCCGATCACGCGGT        0:16    1:0:0   AATTCCGTCAGATGACCCATATAAGAAATGCCCACCAGA no_match        NA      NA</span>
<span class="go">NACACAAGTCTCCCTAGGCCTGTGAC      CAAGATCTCCCTTGTG        4:18    1:0:2   AATTCCGTCAGATGACCCATATAAGAAAGTCACAGGCCT no_match        NA      NA</span>
<span class="go">TGTATTCAGGAGTCTGAATTGTAATA      CTCGAGGAGATCTGAA        4:18    1:0:2   AATTCCGTCAGATGACCCATATAAGAAATATTACAATTC no_match        NA      NA</span>
<span class="go">NCGTCAAGTGCCTGGTGCTCCTGTAT      ACGTCAAGTGCCTGGT        0:16    1:0:0   CGAAGACCAAGACCCATATAAGAAAATACAGGAGCACCA mCD29_CGAAGACCAAGA      0:12    0:0:0</span>
<span class="go">CTAGAGTAGATCGATACGCGGATGGT      CTAGAGTAGATCTGAA        0:15    1:0:1   CGAAGACCAAGACCCATATAAGAAAACCATCCGCGTATC mCD29_CGAAGACCAAGA      0:12    0:0:0</span>
<span class="go">NATCGGGGTCGAACAGGGAGCGTCAG      AACCGCGAGCGTCAAG        11:26   2:0:1   AATTCCGTCAGATGACCCATATAAGAAACTGACGCTCCC no_match        NA      NA</span>
<span class="go">AAAGCAAAGACAAGCCAGTATTTACG      ACACCAAGTCCAGTAT        7:21    1:0:2   AATTCCGTCAGATGACCCATATAAGAAACGTAAATACTG no_match        NA      NA</span>
<span class="go">AAGCCGCGTCTCAACAACAGACTACG      AAGCCGCGTCTCAACA        0:16    0:0:0   AATTCCGTCAGATGACCCATATAAGAAACGTAGTCTGTT no_match        NA      NA</span>
<span class="go">ACGGAGAAGCGCCTCACTCTATCTTC      ATCCGAAAGCGCCTCA        0:16    1:1:1   AATTCCGTCAGATGACCCATATAAGAAAGAAGATAGAGT no_match        NA      NA</span>
<span class="go">TCGTACCCACCATCCTACACCGGCAC      CGGAGTCCACCATCCT        1:16    2:0:1   AATTCCGTCAGATGACCCATATAAGAAAAGTGCCGGTGT no_match        NA      NA</span>
<span class="go">ACAGCTACAGTATGCTTAAAAACAGG      ACAGCTAAGTACTTGC        0:15    0:1:2   AATTCCGTCAGATGACCCATATAAGAAACCTGTTTTTAA no_match        NA      NA</span>
<span class="go">NTAAGACGTCTAAACCGAGCTGGCAC      CAGCAGCGTCTAAACC        2:16    1:0:2   AATTCCGTCAGATGACCCATATAAGAAAGTGCCAGCTCG no_match        NA      NA</span>
<span class="go">TCTCTAATCCAGTATGCCTCTCTTGA      AATCCAGTCCGCATCT        5:21    3:0:0   AATTCCGTCAGATGACCCATATAAGAAATCAAGAGAGGC mCD29_CGAAGACCAAGA      23:34   2:0:1</span>
<span class="go">GACAGAGTCGCATGATTAAAAATCAA      ACAGCTATCGCATGAT        1:16    2:0:1   AATTCCGTCAGATGACCCATATAAGAAATTGATTTTTAA no_match        NA      NA</span>
<span class="go">NTGAGGGTCTCCAACCGCTTTCTAAT      GATCGCGTCTCCAACC        2:16    1:0:2   AATTCCGTCAGATGACCCATATAAGAAAATTAGAAAGCG no_match        NA      NA</span>
<span class="go">NAGCTGGTCGCCAAATACGTATAACT      GCCAAATTCGATAGAA        9:24    2:0:1   AATTCCGTCAGATGACCCATATAAGAAAAGTTATACGTA no_match        NA      NA</span>
<span class="go">GTGCGGTAGTGTTGAAGGTTTATAAT      AGTAGTCAGAAGGTTT        7:21    1:0:2   AATTCCGTCAGATGACCCATATAAGAAAATTATAAACCT no_match        NA      NA</span>
<span class="go">NGAGCACGTGCCTGTGCTACTAGTAC      ACGCCAGGTGCCTGTG        2:16    1:0:2   AATTCCGTCAGATGACCCATATAAGAAAGTACTAGTAGC no_match        NA      NA</span>
<span class="go">CGGGTCAAGACACTAAAAAACCTGCT      ACACTGACAAACTGCT        9:26    2:1:0   AATTCCGTCAGATGACCCATATAAGAAAAGCAGGTTTTT no_match        NA      NA</span>
</pre></div>
</div>
</section>
<section id="id4">
<h3>Barcode extraction<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba extract <span class="se">\</span>
    -1 SRR8550946_1.fastq.gz <span class="se">\</span>
    -2 SRR8550946_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes_CD29.tsv <span class="se">\</span>
    -o feature_barcoding_output.tsv.gz <span class="se">\</span>
    -r1_c <span class="m">0</span>,16 <span class="se">\</span>
    -r2_c <span class="m">0</span>,12 <span class="se">\</span>
    -cb_m <span class="m">2</span>
</pre></div>
</div>
<p>Preview of result.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_num_mismatches       read2_seq       feature_barcode fb_num_mismatches</span>
<span class="go">NCGTCAAGTGCCTGGTgctcctgtat      ACGTCAAGTGCCTGGT        1       CGAAGACCAAGAcccatataagaaaatacaggagcacca mCD29_CGAAGACCAAGA      0</span>
<span class="go">CTAGAGTAGATCGATAcgcggatggt      CTAGAGTAGATCTGAA        2       CGAAGACCAAGAcccatataagaaaaccatccgcgtatc mCD29_CGAAGACCAAGA      0</span>
<span class="go">GGAAAGCCAATCCGATatcccgtatc      GGAAAGCCAATCCGAT        0       CGAAGACCAAGAcccatataagaaagatacgggatatcg mCD29_CGAAGACCAAGA      0</span>
<span class="go">GCAAACTCAAACAACAaaccttaagg      GCAAACTCAAACAACA        0       CGAAGACCAAGAcccatataagaaaccttaaggtttgtt mCD29_CGAAGACCAAGA      0</span>
<span class="go">GTTACAGGTCTCCACTaatagaaggg      GTTACAGGTCTCCACT        0       AATAGCGGAGCCcccatataagaaacccttctattagtg hCD29_AATAGCGGAGCC      0</span>
<span class="go">CGGACACAGGGCTTCCaaagttttag      CGGACACAGGGCTTCC        0       AATAGCGGAGCCcccatataagaaactaaaactttggaa hCD29_AATAGCGGAGCC      0</span>
<span class="go">TACGGATTCACCACCTcaccctcttg      TACGGATTCACCACCT        0       CGAAGACCAAGAcccatataagaaacaagagggtgaggt mCD29_CGAAGACCAAGA      0</span>
<span class="go">GCTTCCAGTTCCCTTGcagacaagag      GCTTCCAGTTCCCTTG        0       CGAAGACCAAGAcccatataagaaactcttgtctgcaag mCD29_CGAAGACCAAGA      0</span>
<span class="go">CTGCCTAGTGAAATCAatggggaggc      CTGCCTAGTGAAATCA        0       CGAAGACCAAGAcccatataagaaagcctccccattgat mCD29_CGAAGACCAAGA      0</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>5.9% (256,759 out of 4,372,604) of total read pairs have valid cell and feature barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 12:33:14,547 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 12:33:14,547 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 12:33:14,547 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 12:33:14,547 - fba.__main__ - INFO - Using extract subcommand ...</span>
<span class="go">2022-01-08 12:33:14,561 - fba.levenshtein - INFO - Number of reference cell barcodes: 6,871</span>
<span class="go">2022-01-08 12:33:14,561 - fba.levenshtein - INFO - Number of reference feature barcodes: 2</span>
<span class="go">2022-01-08 12:33:14,561 - fba.levenshtein - INFO - Read 1 coordinates to search: [0, 16)</span>
<span class="go">2022-01-08 12:33:14,561 - fba.levenshtein - INFO - Read 2 coordinates to search: [0, 12)</span>
<span class="go">2022-01-08 12:33:14,561 - fba.levenshtein - INFO - Cell barcode maximum number of mismatches: 2</span>
<span class="go">2022-01-08 12:33:14,561 - fba.levenshtein - INFO - Feature barcode maximum number of mismatches: 1</span>
<span class="go">2022-01-08 12:33:14,561 - fba.levenshtein - INFO - Read 1 maximum number of N allowed: 3</span>
<span class="go">2022-01-08 12:33:14,561 - fba.levenshtein - INFO - Read 2 maximum number of N allowed: 3</span>
<span class="go">2022-01-08 12:33:15,856 - fba.levenshtein - INFO - Matching ...</span>
<span class="go">2022-01-08 12:37:37,930 - fba.levenshtein - INFO - Number of read pairs processed: 4,372,604</span>
<span class="go">2022-01-08 12:37:37,931 - fba.levenshtein - INFO - Number of read pairs w/ valid barcodes: 256,759</span>
<span class="go">2022-01-08 12:37:37,985 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
</section>
<section id="id5">
<h3>Matrix generation<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba count <span class="se">\</span>
    -i feature_barcoding_output.tsv.gz <span class="se">\</span>
    -o matrix_featurecount.csv.gz <span class="se">\</span>
    -us <span class="m">16</span> <span class="se">\</span>
    -ul <span class="m">10</span> <span class="se">\</span>
    -um <span class="m">1</span> <span class="se">\</span>
    -ud directional
</pre></div>
</div>
<p>Result summary.</p>
<p>96.2% (246,996 out of 256,759) of read pairs with valid cell and feature barcodes are unique fragments. 5.6% (246,996 out of 4,372,604) of total sequenced read pairs contribute to the final matrix with an average of 29 UMIs per cell.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 12:37:38,233 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 12:37:38,233 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 12:37:38,233 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 12:37:38,233 - fba.__main__ - INFO - Using count subcommand ...</span>
<span class="go">2022-01-08 12:37:39,087 - fba.count - INFO - UMI-tools version: 1.1.1</span>
<span class="go">2022-01-08 12:37:39,090 - fba.count - INFO - UMI starting position on read 1: 16</span>
<span class="go">2022-01-08 12:37:39,090 - fba.count - INFO - UMI length: 10</span>
<span class="go">2022-01-08 12:37:39,090 - fba.count - INFO - UMI-tools deduplication threshold: 1</span>
<span class="go">2022-01-08 12:37:39,090 - fba.count - INFO - UMI-tools deduplication method: directional</span>
<span class="go">2022-01-08 12:37:39,090 - fba.count - INFO - Header line: read1_seq cell_barcode cb_num_mismatches read2_seq feature_barcode fb_num_mismatches</span>
<span class="go">2022-01-08 12:37:39,587 - fba.count - INFO - Number of lines processed: 256,759</span>
<span class="go">2022-01-08 12:37:39,590 - fba.count - INFO - Number of cell barcodes detected: 6,871</span>
<span class="go">2022-01-08 12:37:39,590 - fba.count - INFO - Number of features detected: 2</span>
<span class="go">2022-01-08 12:37:40,917 - fba.count - INFO - Total UMIs after deduplication: 246,996</span>
<span class="go">2022-01-08 12:37:40,926 - fba.count - INFO - Median number of UMIs per cell: 29.0</span>
<span class="go">2022-01-08 12:37:40,983 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>t-SNE embedding of cells based on the abundance of features (hashtags, no transcriptome information used). Colors indicate the hashtag status for each cell, as called by FBA, and the abundance of protein tags. This is a re-creation of <a class="reference external" href="https://www.nature.com/articles/s41592-019-0392-0/figures/1">Fig. 1c</a> in <a class="reference external" href="https://doi.org/10.1038/s41592-019-0392-0">Mimitou, E.P., et al. (2019)</a> (The embedding is based on hashtags, not the transcriptomes).</p>
<a class="reference internal image-reference" href="../../../_images/Rplot_embedding_protein-tag_vertical.png"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Rplot_embedding_protein-tag_vertical.png" style="width: 350px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
</section>
<section id="guide-tag">
<h2>Guide-tag<a class="headerlink" href="#guide-tag" title="Permalink to this headline">¶</a></h2>
<section id="id6">
<h3>Preparation<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<p>Prepare feature barcodes (guide-tag sequences, from <a class="reference external" href="https://www.nature.com/articles/s41592-019-0392-0#MOESM1">Supplementary Table 2</a>, <a class="reference external" href="https://www.nature.com/articles/s41592-019-0392-0#MOESM1">Supplementary Figure 1</a>c).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat feature_barcodes_guide-tag.tsv

<span class="go">mNT1    CGCGGAGCCGAATACCTCG</span>
<span class="go">mNT2    CGTCGAACCTCCGTGAAAG</span>
<span class="go">mNT3    ATCGAGCCGAACTGCAACT</span>
<span class="go">mNT4    AAGGCGTTCGCCTTACACG</span>
<span class="go">mNT5    GACATTTAGTACCCGGAGT</span>
<span class="go">mNT6    CTCGTTCCCTAACGGCGCG</span>
<span class="go">mNT7    CCCGTAGACGGTCGAACAA</span>
<span class="go">mNT8    CCATATCGCACCCGATGGG</span>
<span class="go">mNT9    TTACTAGCAGGTGACGCCC</span>
<span class="go">mNT10   AATACGTTGCGAGTAGAAG</span>
</pre></div>
</div>
</section>
<section id="id7">
<h3>QC<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba qc <span class="se">\</span>
    -1 SRR8550948_1.fastq.gz <span class="se">\</span>
    -2 SRR8550948_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes_guide-tag.tsv <span class="se">\</span>
    --output_directory qc
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 14:39:46,616 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 14:39:46,616 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 14:39:46,617 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 14:39:46,617 - fba.__main__ - INFO - Using qc subcommand ...</span>
<span class="go">2022-01-08 14:39:47,209 - fba.qc - INFO - Summarizing per base read content ...</span>
<span class="go">2022-01-08 14:39:47,209 - fba.qc - INFO - Number of read pairs to analyze: 100,000</span>
<span class="go">2022-01-08 14:39:47,209 - fba.qc - INFO - Output directory: qc</span>
<span class="go">2022-01-08 14:39:47,428 - fba.qc - INFO - Number of reads processed: 100,000</span>
<span class="go">2022-01-08 14:39:50,099 - fba.regex - INFO - regex version: 2.5.91</span>
<span class="go">2022-01-08 14:39:50,103 - fba.regex - INFO - Number of reference cell barcodes: 6,871</span>
<span class="go">2022-01-08 14:39:50,103 - fba.regex - INFO - Number of reference feature barcodes: 10</span>
<span class="go">2022-01-08 14:39:50,103 - fba.regex - INFO - Cell barcode maximum number of mismatches: 3</span>
<span class="go">2022-01-08 14:39:50,103 - fba.regex - INFO - Feature barcode maximum number of mismatches: 3</span>
<span class="go">2022-01-08 14:39:50,103 - fba.regex - INFO - Read 1 maximum number of N allowed: inf</span>
<span class="go">2022-01-08 14:39:50,103 - fba.regex - INFO - Read 2 maximum number of N allowed: inf</span>
<span class="go">2022-01-08 14:39:50,103 - fba.regex - INFO - Number of read pairs to analyze: 100,000</span>
<span class="go">2022-01-08 14:39:51,295 - fba.regex - INFO - Number of threads: 72</span>
<span class="go">2022-01-08 14:39:51,296 - fba.regex - INFO - Chunk size: 50,000</span>
<span class="go">2022-01-08 14:39:51,296 - fba.regex - INFO - Matching ...</span>
<span class="go">2022-01-08 14:40:49,770 - fba.regex - INFO - Read pairs processed: 50,000</span>
<span class="go">2022-01-08 14:41:50,088 - fba.regex - INFO - Read pairs processed: 100,000</span>
<span class="go">2022-01-08 14:41:51,366 - fba.qc - INFO - Summarizing barcode coordinates ...</span>
<span class="go">2022-01-08 14:41:51,366 - fba.qc - INFO - Output directory: qc</span>
<span class="go">2022-01-08 14:41:52,542 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>As for read 2, based on the per base content, it suggests that bases 15-34 are actually our feature barcodes (See the distribution of matched barcode positions on read 2).</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_per_base_seq_content_guide-tag.png"><img alt="../../../_images/Pyplot_read2_per_base_seq_content_guide-tag.png" class="align-center" src="../../../_images/Pyplot_read2_per_base_seq_content_guide-tag.png" style="width: 400px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_barcodes_starting_ending_guide-tag.png"><img alt="../../../_images/Pyplot_read2_barcodes_starting_ending_guide-tag.png" class="align-center" src="../../../_images/Pyplot_read2_barcodes_starting_ending_guide-tag.png" style="width: 400px;"/></a>
<p>The detailed qc results are stored in <code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns indicate the matched positions on reads. <code class="docutils literal notranslate"><span class="pre">matching_description</span></code> columns indicate mismatches in <code class="docutils literal notranslate"><span class="pre">substitutions:insertions:deletions</span></code> format.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">NTAAGAGGTCTGCAATCTATATGCAA      AGCGGTCCATGCAATC        3:17    1:0:2   TTCTAGCTCTAAAACCCCGTAGACGGTCGAACAATCCCC mNT7_CCCGTAGACGGTCGAACAA        15:34   0:0:0</span>
<span class="go">ATCCACCGTCATATCGACATGCCACA      ATCCACCGTCATATCG        0:16    0:0:0   TTCTAGCTCTAAAACCCCGTAGACGGTCGAACAATCCCC mNT7_CCCGTAGACGGTCGAACAA        15:34   0:0:0</span>
<span class="go">TGCCAAACACTGAAGGATGTCGCCAC      ACACTGAAGGATGTAT        6:22    2:0:0   TTCTAGCTCTAAAACGACATTTAGTACCCGGAGTCCCCA mNT5_GACATTTAGTACCCGGAGT        15:34   0:0:0</span>
<span class="go">NTCGAAGAGGGTATCGTGAAGTGCTT      AGGGATGGTGAAGGCT        7:25    1:2:0   TTCTAGCTCTAAAACCCATATCGCACCCGATGGGTCCCC mNT8_CCATATCGCACCCGATGGG        15:34   0:0:0</span>
<span class="go">CCTTACGAGTGGACGTGCAGTCAGGT      CCTTACGAGTGGACGT        0:16    0:0:0   TTCTAGCTCTAAAACCTCGTTCCCTAACGGCGCGGCCCA mNT6_CTCGTTCCCTAACGGCGCG        15:34   0:0:0</span>
<span class="go">GGTACAGGATCGCAAACGCGCAAATT      ACACTGATCGCAAACT        3:18    2:0:1   AATTCCGTCAGATGACCCATATAAGAAATTTGCGCGTTT no_match        NA      NA</span>
<span class="go">NAAATGATCCAAACTGTAAGGGAAGC      AAACCTGTCCAAACTG        1:16    0:1:2   TTCTAGCTCTAAAACGACATTTAGTACCCGGAGTCCCCA mNT5_GACATTTAGTACCCGGAGT        15:34   0:0:0</span>
<span class="go">NCACATAAGGAGTTGCGCAACCGCGA      CACATAGAGTTCGCGC        1:18    0:2:1   TTCTAGCTCTAAAACCCATATCGCACCCGATGGGTCCCC mNT8_CCATATCGCACCCGATGGG        15:34   0:0:0</span>
<span class="go">AGGGTGAAGCGCTTATTAATCGAAGG      AGGGTGAAGCGATGAC        0:16    3:0:0   AATTCCGTCAGATGACCCATATAAGAAACCTTCGATTAA no_match        NA      NA</span>
</pre></div>
</div>
</section>
<section id="id8">
<h3>Barcode extraction<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba extract <span class="se">\</span>
    -1 SRR8550948_1.fastq.gz <span class="se">\</span>
    -2 SRR8550948_2.fastq.gz <span class="se">\</span>
    -w cell_barcodes.txt <span class="se">\</span>
    -f feature_barcodes_guide-tag.tsv <span class="se">\</span>
    -o feature_barcoding_output.tsv.gz <span class="se">\</span>
    -r1_c <span class="m">0</span>,16 <span class="se">\</span>
    -r2_c <span class="m">15</span>,34 <span class="se">\</span>
    -cb_m <span class="m">2</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>63.4% (24,425,023 out of 38,537,829) of total read pairs have valid cell and feature barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 14:41:52,877 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 14:41:52,877 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 14:41:52,877 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 14:41:52,877 - fba.__main__ - INFO - Using extract subcommand ...</span>
<span class="go">2022-01-08 14:41:52,894 - fba.levenshtein - INFO - Number of reference cell barcodes: 6,871</span>
<span class="go">2022-01-08 14:41:52,895 - fba.levenshtein - INFO - Number of reference feature barcodes: 10</span>
<span class="go">2022-01-08 14:41:52,895 - fba.levenshtein - INFO - Read 1 coordinates to search: [0, 16)</span>
<span class="go">2022-01-08 14:41:52,895 - fba.levenshtein - INFO - Read 2 coordinates to search: [15, 34)</span>
<span class="go">2022-01-08 14:41:52,895 - fba.levenshtein - INFO - Cell barcode maximum number of mismatches: 2</span>
<span class="go">2022-01-08 14:41:52,895 - fba.levenshtein - INFO - Feature barcode maximum number of mismatches: 1</span>
<span class="go">2022-01-08 14:41:52,895 - fba.levenshtein - INFO - Read 1 maximum number of N allowed: 3</span>
<span class="go">2022-01-08 14:41:52,895 - fba.levenshtein - INFO - Read 2 maximum number of N allowed: 3</span>
<span class="go">2022-01-08 14:41:54,222 - fba.levenshtein - INFO - Matching ...</span>
<span class="go">2022-01-08 14:54:23,469 - fba.levenshtein - INFO - Read pairs processed: 10,000,000</span>
<span class="go">2022-01-08 15:06:57,696 - fba.levenshtein - INFO - Read pairs processed: 20,000,000</span>
<span class="go">2022-01-08 15:19:24,990 - fba.levenshtein - INFO - Read pairs processed: 30,000,000</span>
<span class="go">2022-01-08 15:30:09,131 - fba.levenshtein - INFO - Number of read pairs processed: 38,537,829</span>
<span class="go">2022-01-08 15:30:09,131 - fba.levenshtein - INFO - Number of read pairs w/ valid barcodes: 24,425,023</span>
<span class="go">2022-01-08 15:30:09,188 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
</section>
<section id="id9">
<h3>Matrix generation<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba count <span class="se">\</span>
    -i feature_barcoding_output.tsv.gz <span class="se">\</span>
    -o matrix_featurecount.csv.gz <span class="se">\</span>
    -us <span class="m">16</span> <span class="se">\</span>
    -ul <span class="m">10</span> <span class="se">\</span>
    -um <span class="m">1</span> <span class="se">\</span>
    -ud directional
</pre></div>
</div>
<p>Result summary.</p>
<p>4.3% (1,050,888 out of 24,425,023) of read pairs with valid cell and feature barcodes are unique fragments. 2.7% (1,050,888 out of 38,537,829) of total sequenced read pairs contribute to the final matrix with an average of 102 UMIs per cell.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-08 15:30:09,447 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-08 15:30:09,447 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-08 15:30:09,447 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2022-01-08 15:30:09,447 - fba.__main__ - INFO - Using count subcommand ...</span>
<span class="go">2022-01-08 15:30:10,313 - fba.count - INFO - UMI-tools version: 1.1.1</span>
<span class="go">2022-01-08 15:30:10,316 - fba.count - INFO - UMI starting position on read 1: 16</span>
<span class="go">2022-01-08 15:30:10,316 - fba.count - INFO - UMI length: 10</span>
<span class="go">2022-01-08 15:30:10,316 - fba.count - INFO - UMI-tools deduplication threshold: 1</span>
<span class="go">2022-01-08 15:30:10,316 - fba.count - INFO - UMI-tools deduplication method: directional</span>
<span class="go">2022-01-08 15:30:10,316 - fba.count - INFO - Header line: read1_seq cell_barcode cb_num_mismatches read2_seq feature_barcode fb_num_mismatches</span>
<span class="go">2022-01-08 15:30:53,780 - fba.count - INFO - Number of lines processed: 24,425,023</span>
<span class="go">2022-01-08 15:30:53,785 - fba.count - INFO - Number of cell barcodes detected: 6,867</span>
<span class="go">2022-01-08 15:30:53,785 - fba.count - INFO - Number of features detected: 10</span>
<span class="go">2022-01-08 15:31:18,853 - fba.count - INFO - Total UMIs after deduplication: 1,050,888</span>
<span class="go">2022-01-08 15:31:18,862 - fba.count - INFO - Median number of UMIs per cell: 102.0</span>
<span class="go">2022-01-08 15:31:19,003 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>t-SNE embedding of cells based on the abundance of features (hashtags, no transcriptome information used). Colors indicate the guide tag abundance for each cell, as caculated by FBA. This is a re-creation of <a class="reference external" href="https://www.nature.com/articles/s41592-019-0392-0/figures/1">Fig. 1c (iv)</a>  in <a class="reference external" href="https://doi.org/10.1038/s41592-019-0392-0">Mimitou, E.P., et al. (2019)</a> (The embedding is based on hashtags, not the transcriptomes).</p>
<a class="reference internal image-reference" href="../../../_images/Rplot_embedding_guide.png"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Rplot_embedding_guide.png" style="width: 350px;"/></a>
<div class="line-block">
<div class="line"><br/></div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../../targeted_transcript_enrichment/index.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Targeted transcript enrichment</div>
              </div>
              <svg><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../index.html">
              <svg><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">ECCITE-Seq</div>
                
              </div>
            </a>
        </div>

        <div class="related-information">
              Copyright &#169; 2020-2022, Jialei Duan
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            Contents
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs</a><ul>
<li><a class="reference internal" href="#preparation">Preparation</a></li>
<li><a class="reference internal" href="#hashtag">Hashtag</a><ul>
<li><a class="reference internal" href="#id1">Preparation</a></li>
<li><a class="reference internal" href="#qc">QC</a></li>
<li><a class="reference internal" href="#barcode-extraction">Barcode extraction</a></li>
<li><a class="reference internal" href="#matrix-generation">Matrix generation</a></li>
<li><a class="reference internal" href="#demultiplexing">Demultiplexing</a></li>
</ul>
</li>
<li><a class="reference internal" href="#protein-tag">Protein-tag</a><ul>
<li><a class="reference internal" href="#id2">Preparation</a></li>
<li><a class="reference internal" href="#id3">QC</a></li>
<li><a class="reference internal" href="#id4">Barcode extraction</a></li>
<li><a class="reference internal" href="#id5">Matrix generation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#guide-tag">Guide-tag</a><ul>
<li><a class="reference internal" href="#id6">Preparation</a></li>
<li><a class="reference internal" href="#id7">QC</a></li>
<li><a class="reference internal" href="#id8">Barcode extraction</a></li>
<li><a class="reference internal" href="#id9">Matrix generation</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/scripts/main.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    </body>
</html>