.. fba documentation master file, created by
   sphinx-quickstart on Tue Apr  6 18:44:42 2021.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

Welcome to fba's documentation!
===============================


fba: Tools for feature barcoding analysis


.. image:: https://img.shields.io/pypi/v/fba.svg
   :target: https://pypi.python.org/pypi/fba

.. image:: https://img.shields.io/badge/license-MIT-yellow.svg
   :target: https://github.com/jlduan/fba/blob/master/LICENSE

.. image:: https://github.com/jlduan/fba/actions/workflows/fba.yml/badge.svg?branch=master
   :target: https://github.com/jlduan/fba/actions/workflows/fba.yml

.. image:: https://readthedocs.org/projects/fba/badge/?version=latest
   :target: https://fba.readthedocs.io/en/latest/

.. image:: https://codecov.io/gh/jlduan/fba/branch/master/graph/badge.svg?token=H3189R59G0
   :target: https://codecov.io/gh/jlduan/fba

.. image:: https://zenodo.org/badge/DOI/10.5281/zenodo.4642815.svg
   :target: https://doi.org/10.5281/zenodo.4642815

Quickstart
----------

Installation
++++++++++++

.. code-block::

   pip install fba


Usage
+++++

.. code-block::

   $ fba

   usage: fba [-h]  ...

   Tools for feature barcoding analyses

   optional arguments:
   -h, --help        show this help message and exit

   functions:

      extract         extract cell and feature barcodes
      map             map enriched transcripts
      filter          filter extracted barcodes
      count           count feature barcodes per cell
      demultiplex     demultiplex cells based on feature abundance
      qc              quality control of feature barcoding assay
      kallisto_wrapper
                     deploy kallisto/bustools for feature barcoding
                     quantification

* **extract**: cell and feature barcodes from paired fastq files. For single cell assays, read 1 usually contains cell partitioning and UMI information, and read 2 contains feature information.
* **map**: quantify enriched transcripts (through hybridization or PCR amplification) from parent single cell libraries. Read 1 contains cell partitioning and UMI information, and read 2 contains transcribed regions of enriched/targeted transcripts of interest. BWA (Li, H. 2013) or Bowtie2 (Langmead, B., et al. 2012) is used for read 2 alignment. The quantification (UMI deduplication) of enriched/targeted transcripts is powered by UMI-tools (Smith, T., et al. 2017).
* **filter**: filter extracted cell and feature barcodes (output of ``extract`` or ``qc``). Additional fragment filter/selection can be applied through ``-cb_seq`` and/or ``-fb_seq``.
* **count**: count UMIs per feature per cell (UMI deduplication), powered by UMI-tools (Smith, T., et al. 2017). Take the output of ``extract`` or ``filter`` as input.
* **demultiplex**: demultiplex cells based on the abundance of features (matrix generated by ``count`` as input).
* **qc**: generate diagnostic information. If ``-1`` is omitted, bulk mode is enabled and only read 2 will be analyzed.
* **kallisto_wrapper**: deploy kallisto/bustools for feature barcoding quantification (just a wrapper) (Bray, N.L., et al. 2016).


Contents
--------

.. :caption: Contents

.. toctree::
   :maxdepth: 2

   tutorial
   LICENSE
   help

Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
