############
 User Guide
############

|

******************
 What is ``fba``?
******************

``fba`` is a flexible and streamlined toolbox for quality control,
quantification, demultiplexing of various single-cell feature barcoding
assays. It can be applied to customized feature barcoding
specifications, including different CRISPR constructs or targeted
enriched transcripts. ``fba`` allows users to customize a wide range of
parameters for the quantification and demultiplexing process. ``fba``
also has a user-friendly quality control module, which is helpful in
troubleshooting feature barcoding experiments.

**********
 Workflow
**********

.. figure:: https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/37/22/10.1093_bioinformatics_btab375/2/btab375f1.jpeg?Expires=1674792986&Signature=BtQPKObtXmv1XfGAp0GVyWPFZ5zTJuHyc9yPaZ29Di1iphZd2SJKlU6mLtBbLeZmuWMIYPl06w~h0Wmz724E5vbbnPFxm8d5IyrNAOa-0QCwoYrHxd4~SCeDu0QkQHeIBVGxHpVt7qjz-1mSl5XaTVpl8~2QoZwc1FB~y1fS3l9aR0j6~KnTBjDJ9n9hkoOqmt94R~qCnmEEyboxbsujClNWUJTt-g1n8PTVOFNF7o-Co5QisbmFze28YaCpzNUTS6Yhwap6Q4g2YC6geQw5CsKNLwHcRcf2aZZMLA2r94~2Bw2xMDxzPOTnPKx32vSFoUJPQ~y95G3ynX8HHPsXxA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA
   :target: https://doi.org/10.1093/bioinformatics/btab375
   :alt: FBA workflow
   :width: 500px
   :align: center

   The workflow of FBA: a flexible and streamlined package for single-cell feature barcoding assays.

*******
 Usage
*******

.. code:: console

   $ fba

   usage: fba [-h]  ...

   Tools for single-cell feature barcoding analysis

   optional arguments:
   -h, --help         show this help message and exit

   functions:

      extract         extract cell and feature barcodes
      map             map enriched transcripts
      filter          filter extracted barcodes
      count           count feature barcodes per cell
      demultiplex     demultiplex cells based on feature abundance
      qc              quality control of feature barcoding assay
      kallisto_wrapper
                      deploy kallisto/bustools for feature barcoding
                      quantification

-  **extract**: extract cell and feature barcodes from paired fastq
   files. For single cell assays, read 1 usually contains cell
   partitioning and UMI information, and read 2 contains feature
   information.

-  **map**: quantify enriched transcripts (through hybridization or PCR
   amplification) from parent single cell libraries. Read 1 contains
   cell partitioning and UMI information, and read 2 contains
   transcribed regions of enriched/targeted transcripts of interest. BWA
   (Li, H. 2013) or Bowtie2 (Langmead, B., et al. 2012) is used for read
   2 alignment. The quantification (UMI deduplication) of
   enriched/targeted transcripts is powered by UMI-tools (Smith, T., et
   al. 2017).

-  **filter**: filter extracted cell and feature barcodes (output of
   ``extract`` or ``qc``). Additional fragment filter/selection can be
   applied through ``-cb_seq`` and/or ``-fb_seq``.

-  **count**: count UMIs per feature per cell (UMI deduplication),
   powered by UMI-tools (Smith, T., et al. 2017). Take the output of
   ``extract`` or ``filter`` as input.

-  **demultiplex**: demultiplex cells based on the abundance of features
   (matrix generated by ``count`` as input).

-  **qc**: generate diagnostic information. If ``-1`` is omitted, bulk
   mode is enabled and only read 2 will be analyzed.

-  **kallisto_wrapper**: deploy kallisto/bustools for feature barcoding
   quantification (just a wrapper) (Bray, N.L., et al. 2016).

|
