<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />
<meta property="og:title" content="10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://jlduan.github.io/fba/_build/html/tutorials/crispr_screening/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html" />
<meta property="og:site_name" content="fba" />
<meta property="og:description" content="Dataset: 10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed The detailed description of this dataset can be found here., Preparation: Download fastq files. Combine reads of different lanes. Download cell barcode info. These are the cell-associated barcodes in this si..." />
<meta property="og:image" content="https://raw.githubusercontent.com/jlduan/fba/gh-pages/docs/_static/logo.svg" />
<meta property="og:image:alt" content="fba" />
<meta name="description" content="Dataset: 10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed The detailed description of this dataset can be found here., Preparation: Download fastq files. Combine reads of different lanes. Download cell barcode info. These are the cell-associated barcodes in this si..." />
<meta property="og:ignore_canonical" content="true" />
<meta property="article:author" content="Jialei Duan" />
<meta property="og:site_name" content="FBA" />
<meta name="twitter:card" content="summary"></meta>
<meta name="twitter:creator" content="@jlduan" />
<meta name="twitter:image" content="https://github.com/jlduan.png" />
<link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" /><link rel="next" title="CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)" href="../PRJNA358686/tutorial.html" /><link rel="prev" title="CRISPR screening" href="../index.html" />

    <link rel="shortcut icon" href="../../../_static/logo.ico"/><!-- Generated with Sphinx 5.3.0 and Furo 2022.12.07 -->
        <title>10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA - FBA</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?digest=91d0f0d1c444bdcb17a68e833c7a53903343c195" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #eeeedd;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #272822;
  --color-code-foreground: #f8f8f2;
  
      }
    }
  }
</style>


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-X0MEL2T23M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-X0MEL2T23M');
</script>

</head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">FBA</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo only-light" src="../../../_static/logo.svg" alt="Light Logo"/>
    <img class="sidebar-logo only-dark" src="../../../_static/logo_dark.svg" alt="Dark Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user_guide.html">User Guide</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current has-children"><a class="reference internal" href="../index.html">CRISPR screening</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../PRJNA358686/tutorial.html">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../PRJNA609688/tutorial.html">Direct-capture Perturb-seq; CRISPRi-based Screen of Unfolded Protein Response (UPR) Using 3' sgRNA-CR1&lt;sup&gt;cs1&lt;/sup&gt;</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_surface_protein_labeling/index.html">Cell surface protein labeling</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA393315/tutorial.html">CITE-seq; 8k Cord Blood Mononuclear Cells with 13 Antibodies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/PRJNA658075/tutorial.html">ASAP-seq; Multiplexed CRISPR Perturbations in Primary T Cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cell_surface_protein_labeling/SC3_v3_NextGem_DI_PBMC_CSP_1K/tutorial.html">1k Human PBMCs Stained with a Panel of TotalSeq B Antibodies, Dual Indexed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../eccite-seq/index.html">ECCITE-seq</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../eccite-seq/PRJNA521522/tutorial.html">6k Single-cell Multimodal Readout of NIH-3T3, MyLa, Sez4 and PBMCs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../phage-atac/index.html">PHAGE-ATAC</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../phage-atac/PRJNA661457/tutorial.html">PHAGE-ATAC; Anti-CD8 Phage Hashing Single-cell ATAC-seq Using CD8 T Cells from Four Human Donors</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cellplex/index.html">CellPlex</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../cellplex/SC3_v3_NextGem_DI_CellPlex_Nuclei_30K_Multiplex/tutorial.html">30k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Nuclei Multiplexed, 12 CMOs</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../cell_hashing/index.html">Cell hashing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../cell_hashing/PRJNA423077/tutorial.html">Peripheral Blood Mononuclear cells with 8 Antibodies</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../multi-seq/index.html">MULTI-seq</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../multi-seq/PRJNA531855/tutorial.html">15k HEK293 and 40k HMECs Multiplexed by Lipid- and Cholesterol-tagged Indices</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../targeted_transcript_enrichment/index.html">Targeted transcript enrichment</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../targeted_transcript_enrichment/Targeted_NGSC3_DI_HodgkinsLymphoma_GeneSignature/tutorial.html">Hodgkin’s Lymphoma, Dissociated Tumor: Targeted, Gene Signature Panel</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../pseudo-bulk/index.html">Pseudo-bulk</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.html">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA, Dual Indexed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../pseudo-bulk/SC3_v3_NextGem_DI_CellPlex_Jurkat_Raji_10K_Multiplex/tutorial.html">10k 1:1 Mixture of Raji and Jurkat Cells Multiplexed, 2 CMOs</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/jlduan/fba">GitHub Repository</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/jlduan/fba/edit/gh-pages/docs/tutorials/crispr_screening/SC3_v3_NextGem_DI_CRISPR_10K/tutorial.rst" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="k-a375-cells-transduced-with-1-non-target-and-1-target-sgrna">
<span id="tutorial-crispr-screening-sc3-v3-nextgem-di-crispr-10k"></span><h1>10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA<a class="headerlink" href="#k-a375-cells-transduced-with-1-non-target-and-1-target-sgrna" title="Permalink to this heading">#</a></h1>
<p><strong>Dataset</strong>: 10k A375 Cells Transduced with (1) Non-Target and (1)
Target sgRNA, Dual Indexed</p>
<p>The detailed description of this dataset can be found <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/4.0.0/SC3_v3_NextGem_DI_CRISPR_10K">here</a>.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="preparation">
<h2>Preparation<a class="headerlink" href="#preparation" title="Permalink to this heading">#</a></h2>
<p>Download fastq files.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>wget https://cg.10xgenomics.com/samples/cell-exp/4.0.0/SC3_v3_NextGem_DI_CRISPR_10K/SC3_v3_NextGem_DI_CRISPR_10K_fastqs.tar

<span class="gp">$ </span>tar xvf SC3_v3_NextGem_DI_PBMC_CSP_1K/SC3_v3_NextGem_DI_PBMC_CSP_1K_fastqs.tar
</pre></div>
</div>
<p>Combine reads of different lanes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat SC3_v3_NextGem_DI_CRISPR_10K_fastqs/SC3_v3_NextGem_DI_CRISPR_10K_crispr_fastqs/SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_L00?_R1_001.fastq.gz &gt; SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_combined_R1_001.fastq.gz

<span class="gp">$ </span>cat SC3_v3_NextGem_DI_CRISPR_10K_fastqs/SC3_v3_NextGem_DI_CRISPR_10K_crispr_fastqs/SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_L00?_R2_001.fastq.gz &gt; SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_combined_R2_001.fastq.gz
</pre></div>
</div>
<p>Download cell barcode info. These are the cell-associated barcodes in
this single cell RNA-Seq library.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>wget https://cf.10xgenomics.com/samples/cell-exp/4.0.0/SC3_v3_NextGem_DI_CRISPR_10K/SC3_v3_NextGem_DI_CRISPR_10K_filtered_feature_bc_matrix.tar.gz

<span class="gp">$ </span>tar zxvf SC3_v3_NextGem_DI_CRISPR_10K_filtered_feature_bc_matrix.tar.gz
</pre></div>
</div>
<p>Inspect cell barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc filtered_feature_bc_matrix/barcodes.tsv.gz <span class="p">|</span> head

<span class="go">AAACCCACACATAGCT-1</span>
<span class="go">AAACCCACATCATGAC-1</span>
<span class="go">AAACCCAGTCATCGGC-1</span>
<span class="go">AAACCCAGTCGCACAC-1</span>
<span class="go">AAACCCAGTGCAACGA-1</span>
<span class="go">AAACCCATCAAGCGTT-1</span>
<span class="go">AAACCCATCAGATGCT-1</span>
<span class="go">AAACCCATCATCTACT-1</span>
<span class="go">AAACCCATCCTTATGT-1</span>
<span class="go">AAACCCATCTCGGCTT-1</span>
</pre></div>
</div>
<p>Prepare feature barcodes.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>wget https://cf.10xgenomics.com/samples/cell-exp/4.0.0/SC3_v3_NextGem_DI_CRISPR_10K/SC3_v3_NextGem_DI_CRISPR_10K_feature_ref.csv
</pre></div>
</div>
<p>Inspect feature barcode info.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat SC3_v3_NextGem_DI_CRISPR_10K_feature_ref.csv

<span class="go">id,name,read,pattern,sequence,feature_type,target_gene_id,target_gene_name</span>
<span class="go">RAB1A-2,RAB1A-2,R2,(BC)GTTTAAGAGCTAAGCTGGAA,GCCGGCGAACCAGGAAATA,CRISPR Guide Capture,ENSG00000138069,RAB1A</span>
<span class="go">NON_TARGET-1,NON_TARGET-1,R2,(BC)GTTTAAGAGCTAAGCTGGAA,AACGTGCTGACGATGCGGGC,CRISPR Guide Capture,Non-Targeting,Non-Targeting</span>
</pre></div>
</div>
<p>Clean up.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cut -d <span class="s1">&#39;,&#39;</span> -f1,5 SC3_v3_NextGem_DI_CRISPR_10K_feature_ref.csv <span class="p">|</span> sed <span class="s1">&#39;s/,/\t/g&#39;</span> <span class="p">|</span> tail -2 &gt; SC3_v3_NextGem_DI_CRISPR_10K_feature_ref.tsv

<span class="gp">$ </span>cat SC3_v3_NextGem_DI_CRISPR_10K_feature_ref.tsv

<span class="go">RAB1A-2 GCCGGCGAACCAGGAAATA</span>
<span class="go">NON_TARGET-1    AACGTGCTGACGATGCGGGC</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="qc">
<h2>QC<a class="headerlink" href="#qc" title="Permalink to this heading">#</a></h2>
<p>Sample the first 20,000 (set by <code class="docutils literal notranslate"><span class="pre">-n</span></code>, default <code class="docutils literal notranslate"><span class="pre">100,000</span></code>) read pairs
for quality control. Use <code class="docutils literal notranslate"><span class="pre">-t</span></code> to set the number of threads. By
default, the diagnostic results and plots are generated in the <code class="docutils literal notranslate"><span class="pre">qc</span></code>
directory (set by <code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>), and full length of read 1 and
read 2 are searched against reference cell and feature barcodes,
respectively. The per base content of both read pairs and the
distribution of matched barcode positions are summarized. Use <code class="docutils literal notranslate"><span class="pre">-r1_c</span></code>
and/or <code class="docutils literal notranslate"><span class="pre">-r2_c</span></code> to limit the search range. Use <code class="docutils literal notranslate"><span class="pre">-cb_n</span></code> and/or
<code class="docutils literal notranslate"><span class="pre">-fb_n</span></code> to set the mismatch tolerance for cell and feature barcode
matching (default <code class="docutils literal notranslate"><span class="pre">3</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba qc <span class="se">\</span>
    -1 SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_combined_R1_001.fastq.gz <span class="se">\</span>
    -2 SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_combined_R2_001.fastq.gz <span class="se">\</span>
    -w filtered_feature_bc_matrix/barcodes.tsv.gz <span class="se">\</span>
    -f SC3_v3_NextGem_DI_CRISPR_10K_feature_ref.tsv <span class="se">\</span>
    -r1_c <span class="m">0</span>,16 <span class="se">\</span>
    -n <span class="m">20000</span>
</pre></div>
</div>
<p>This library is built using the Chromium Next GEM Single Cell 3ʹ Reagent
Kits v3.1 (Dual Index) with Feature Barcode technology for CRISPR
Screening and sequenced on Illumina NovaSeq 6000. The first 16 bases are
cell barcodes and the following 12 bases are UMIs. Based on the base
content plot, the GC content of cell barcodes are quite even. The UMIs
are slightly T enriched.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_barcodes_starting_ending4.webp"><img alt="../../../_images/Pyplot_read1_barcodes_starting_ending4.webp" class="align-center" src="../../../_images/Pyplot_read1_barcodes_starting_ending4.webp" style="width: 350px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read1_per_base_seq_content6.webp"><img alt="../../../_images/Pyplot_read1_per_base_seq_content6.webp" class="align-center" src="../../../_images/Pyplot_read1_per_base_seq_content6.webp" style="width: 350px;" /></a>
<p>As for read 2, based on the per base content, it suggests that bases
0-31 are constant and we can almost read the bases. They are actually
<a class="reference external" href="https://assets.ctfassets.net/an68im79xiti/4HEC1M6tFbBJXXTv7jVVo1/a271ac8d5fa73180e603df21030f9e9a/CG000316_ChromiumNextGEMSingleCell3__v3.1_CRISPR_Screening_Dual_Index_RevA.pdf">Template Switch Oligo (TSO) sequence</a>. Starting from base 32, it seems
there are two genotypes for the reads we have sampled.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_per_base_seq_content6.webp"><img alt="../../../_images/Pyplot_read2_per_base_seq_content6.webp" class="align-center" src="../../../_images/Pyplot_read2_per_base_seq_content6.webp" style="width: 800px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_read2_barcodes_starting_ending6.webp"><img alt="../../../_images/Pyplot_read2_barcodes_starting_ending6.webp" class="align-center" src="../../../_images/Pyplot_read2_barcodes_starting_ending6.webp" style="width: 800px;" /></a>
<p>The detailed <code class="docutils literal notranslate"><span class="pre">qc</span></code> results are stored in
<code class="docutils literal notranslate"><span class="pre">feature_barcoding_output.tsv.gz</span></code> file. <code class="docutils literal notranslate"><span class="pre">matching_pos</span></code> columns
indicate the matched positions on reads. <code class="docutils literal notranslate"><span class="pre">matching_description</span></code>
columns indicate mismatches in substitutions:insertions:deletions
format.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc qc/feature_barcoding_output.tsv.gz <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_matching_pos cb_matching_description read2_seq       feature_barcode fb_matching_pos fb_matching_description</span>
<span class="go">CNCCACACACGTGTTAatgagtactagc    CCTCACACACGTAGTT        0:15    2:0:1   AAGCAGTGGTATCAACGCAGAGTACATGGGATAGGTTTGGTCCTAGCCTTTCTATTAGCTCTTAGTAAGATTACACATGCAAGCATCCCC    no_match        NA      NA</span>
<span class="go">GNCGCGATCAGCATTActtttgtcaccc    GTCGCGAAGAGCATTA        0:16    3:0:0   AAGCAGTGGTATCAACGCAGAGTACATGGGGACTGTTGCTGGTGTGTACTTGCTAAGGTTTATGTCAGTTCAAGATTATAAGCCCCCCAG    no_match        NA      NA</span>
<span class="go">TNGGAAGGTAAGTGTAatcgagggaaca    TGGGAAGCAAAGTGTA        0:16    3:0:0   AAGCAGTGGTATCAACGCAGAGTACATGGGGGCCGGCGAACCAGGAAATAGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAAT    RAB1A-2_GCCGGCGAACCAGGAAATAG    31:51   0:0:0</span>
<span class="go">CNCCCAAGTCGATAGGgagcgcaagcat    CCCAACTCACGATAGG        2:16    1:0:2   AAGCAGTGGTATCAACGCAGAGTACATGGGGGCCGGCGAACCAGGAAATAGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAAT    RAB1A-2_GCCGGCGAACCAGGAAATAG    31:51   0:0:0</span>
<span class="go">CNCACTGCAAACGGTGggcgtaaatgag    CTCACTGGTAACGGTG        0:16    3:0:0   AAGCAGTGGTATCAACGCAGAGTACATGGGGGCCGGCGAACCAGGAAATAGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAAT    RAB1A-2_GCCGGCGAACCAGGAAATAG    31:51   0:0:0</span>
<span class="go">ANCATCACAGGCGCTTgtcccactatat    AGCATCAGTGGCGCTT        0:16    3:0:0   AAGCAGTGGTATCAACGCAGAGTACATGGGGGCCGGCGAACCAGGAAATAGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAAT    RAB1A-2_GCCGGCGAACCAGGAAATAG    31:51   0:0:0</span>
<span class="go">ANACGAACACTTTCATccaaaagaagtt    AAACGAAGTCTTTCAT        0:16    3:0:0   AAGCAGTGGTATCAACGCAGAGTACATGGGGGCCGGCGAACCAGGAAATAGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAAT    RAB1A-2_GCCGGCGAACCAGGAAATAG    31:51   0:0:0</span>
<span class="go">ANCAACCAGTATCGTTgaaatcctggta    AACAACCTCTATCGTT        0:16    3:0:0   AAGCAGTGGTATCAACGCAGAGTACATGGGGAACGTGCTGACGATGCGGGCGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAA    NON_TARGET-1_AACGTGCTGACGATGCGGGC       31:51   0:0:0</span>
<span class="go">GNAGCCCGTACCACATgggcccagtatg    GAAGCCCCAACCACAT        0:16    3:0:0   AAGCAGTGGTATCAACGCAGAGTACATGGGGGCCGGCGAACCAGGAAATAGTTTAAGAGCTAAGCTGGAAACAGCATAGCAAGTTTAAAT    RAB1A-2_GCCGGCGAACCAGGAAATAG    31:51   0:0:0</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="barcode-extraction">
<h2>Barcode extraction<a class="headerlink" href="#barcode-extraction" title="Permalink to this heading">#</a></h2>
<p>Although the lengths of the two feature barcodes are one base different,
they all start at the same position on read 2. For the purpose of
feature barcode identification, let’s include one extra downstream base
(G) for the RAB1A-2 feature barcode to make their lengths equal.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>cat SC3_v3_NextGem_DI_CRISPR_10K_feature_ref_edited.tsv

<span class="go">RAB1A-2 GCCGGCGAACCAGGAAATAG</span>
<span class="go">NON_TARGET-1    AACGTGCTGACGATGCGGGC</span>
</pre></div>
</div>
<p>Search ranges are set to <code class="docutils literal notranslate"><span class="pre">0,16</span></code> on read 1 and <code class="docutils literal notranslate"><span class="pre">31,51</span></code> on read 2. Two
mismatches for cell and feature barcodes (<code class="docutils literal notranslate"><span class="pre">-cb_m</span></code>, <code class="docutils literal notranslate"><span class="pre">-cf_m</span></code>) are
allowed.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba extract <span class="se">\</span>
    -1 SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_combined_R1_001.fastq.gz <span class="se">\</span>
    -2 SC3_v3_NextGem_DI_CRISPR_10K_crispr_S1_combined_R2_001.fastq.gz <span class="se">\</span>
    -w filtered_feature_bc_matrix/barcodes.tsv.gz <span class="se">\</span>
    -f SC3_v3_NextGem_DI_CRISPR_10K_feature_ref_edited.tsv <span class="se">\</span>
    -o feature_barcoding_output.tsv.gz <span class="se">\</span>
    -r1_c <span class="m">0</span>,16 <span class="se">\</span>
    -r2_c <span class="m">31</span>,51 <span class="se">\</span>
    -cb_m <span class="m">2</span> <span class="se">\</span>
    -fb_m <span class="m">2</span>
</pre></div>
</div>
<p>Preview of result.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gzip -dc feature_barcoding_output.tsv.gz  <span class="p">|</span> head

<span class="go">read1_seq       cell_barcode    cb_num_mismatches       read2_seq       feature_barcode fb_num_mismatches</span>
<span class="go">GGCAGTCTCCGTTACTtatccagccttc    GGCAGTCTCGGTAACT        2       aagcagtggtatcaacgcagagtacatggggGCCGGCGAACCAGGAAATAGtttaagagctaagctggaaacagcatagcaagtttaaat    RAB1A-2_GCCGGCGAACCAGGAAATAG     0</span>
<span class="go">TTACGTTGTGAATCGGgtggggctcttc    TTACGTTCAGAATCGG        2       aagcagtggtatcaacgcagagtacatggggAACGTGCTGACGATGCGGGCgtttaagagctaagctggaaacagcatagcaagtttaaa    NON_TARGET-1_AACGTGCTGACGATGCGGGC        0</span>
<span class="go">TCGGGCAAGGATTGGTttctactcggaa    TCGGGCATCGATTGGT        2       aagcagtggtatcaacgcagagtacatgggaACGTGCTGACGATGCGGGCGtttaagagctaagctggaaacagcatagcaagtttaaat    NON_TARGET-1_AACGTGCTGACGATGCGGGC        2</span>
<span class="go">ACAACCACACATCTAGcggcatcatact    ACAACCAGTCATCTAG        2       aagcagtggtatcaacgcagagtacatggggCCGGCGAACCAGGAAATAGTttaagagctaagctggaaacagcatagcaagtttaaata    RAB1A-2_GCCGGCGAACCAGGAAATAG     2</span>
<span class="go">AGACTCAAGTGCTAGAacagaactggtg    AGACTCATCTGCTAGA        2       aagcagtggtatcaacgcagagtacatggggAACGTGCTGACGATGCGGGCgtttaagagctaagctggaaacagcatagcaagtttaaa    NON_TARGET-1_AACGTGCTGACGATGCGGGC        0</span>
<span class="go">GAGTTGTTCGAACATTctgcccgacgtc    GAGTTGTAGGAACATT        2       aagcagtggtatcaacgcagagtacatggggAACGTGCTGACGATGCGGGCgtttaagagctaagctggaaacagcatagcaagtttaaa    NON_TARGET-1_AACGTGCTGACGATGCGGGC        0</span>
<span class="go">AGACTCAGTGGCACAAtgtcagaattca    AGACTCACAGGCACAA        2       aagcagtggtatcaacgcagagtacatggggGCCGGCGAACCAGGAAATAGtttaagagctaagctggaaacagcatagcaagtttaaat    RAB1A-2_GCCGGCGAACCAGGAAATAG     0</span>
<span class="go">TGCACGGAGGATAACCcgtgcacgtaca    TGCACGGTCGATAACC        2       aagcagtggtatcaacgcagagtacatggggGCCGGCGAACCAGGAAATAGtttaagagctaagctggaaacagcatagcaagtttaaat    RAB1A-2_GCCGGCGAACCAGGAAATAG     0</span>
<span class="go">CGTAGTAGTAACACGGaagagggaactg    CGTAGTAGTAACGCGA        2       aagcagtggtatcaacgcagagtacatggggAACGTGCTGACGATGCGGGCgtttaagagctaagctggaaacagcatagcaagtttaaa    NON_TARGET-1_AACGTGCTGACGATGCGGGC        0</span>
</pre></div>
</div>
<p>Result summary.</p>
<p>64.7% (93,795,979 out of 145,032,428) of total read pairs have valid
cell and feature barcodes. Majority of fragments in this library have
correct structure.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-02-15 01:51:59,262 - fba.__main__ - INFO - fba version: 0.0.7</span>
<span class="go">2021-02-15 01:51:59,262 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-02-15 01:51:59,262 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2021-02-15 01:51:59,262 - fba.__main__ - INFO - Using extract subcommand ...</span>
<span class="go">2021-02-15 01:51:59,276 - fba.levenshtein - INFO - Number of reference cell barcodes: 11,791</span>
<span class="go">2021-02-15 01:51:59,276 - fba.levenshtein - INFO - Number of reference feature barcodes: 2</span>
<span class="go">2021-02-15 01:51:59,276 - fba.levenshtein - INFO - Read 1 coordinates to search: [0, 16)</span>
<span class="go">2021-02-15 01:51:59,276 - fba.levenshtein - INFO - Read 2 coordinates to search: [31, 51)</span>
<span class="go">2021-02-15 01:51:59,276 - fba.levenshtein - INFO - Cell barcode maximum number of mismatches: 2</span>
<span class="go">2021-02-15 01:51:59,276 - fba.levenshtein - INFO - Feature barcode maximum number of mismatches: 2</span>
<span class="go">2021-02-15 01:51:59,276 - fba.levenshtein - INFO - Read 1 maximum number of N allowed: 3</span>
<span class="go">2021-02-15 01:51:59,276 - fba.levenshtein - INFO - Read 2 maximum number of N allowed: 3</span>
<span class="go">2021-02-15 01:52:02,510 - fba.levenshtein - INFO - Matching ...</span>
<span class="go">2021-02-15 02:20:39,807 - fba.levenshtein - INFO - Read pairs processed: 10,000,000</span>
<span class="go">2021-02-15 02:49:04,142 - fba.levenshtein - INFO - Read pairs processed: 20,000,000</span>
<span class="go">2021-02-15 03:17:27,422 - fba.levenshtein - INFO - Read pairs processed: 30,000,000</span>
<span class="go">2021-02-15 03:45:54,615 - fba.levenshtein - INFO - Read pairs processed: 40,000,000</span>
<span class="go">2021-02-15 04:14:23,049 - fba.levenshtein - INFO - Read pairs processed: 50,000,000</span>
<span class="go">2021-02-15 04:42:49,377 - fba.levenshtein - INFO - Read pairs processed: 60,000,000</span>
<span class="go">2021-02-15 05:11:15,736 - fba.levenshtein - INFO - Read pairs processed: 70,000,000</span>
<span class="go">2021-02-15 05:39:43,011 - fba.levenshtein - INFO - Read pairs processed: 80,000,000</span>
<span class="go">2021-02-15 06:08:09,940 - fba.levenshtein - INFO - Read pairs processed: 90,000,000</span>
<span class="go">2021-02-15 06:36:39,658 - fba.levenshtein - INFO - Read pairs processed: 100,000,000</span>
<span class="go">2021-02-15 07:05:08,115 - fba.levenshtein - INFO - Read pairs processed: 110,000,000</span>
<span class="go">2021-02-15 07:33:32,101 - fba.levenshtein - INFO - Read pairs processed: 120,000,000</span>
<span class="go">2021-02-15 08:02:01,233 - fba.levenshtein - INFO - Read pairs processed: 130,000,000</span>
<span class="go">2021-02-15 08:30:29,660 - fba.levenshtein - INFO - Read pairs processed: 140,000,000</span>
<span class="go">2021-02-15 08:44:47,038 - fba.levenshtein - INFO - Number of read pairs processed: 145,032,428</span>
<span class="go">2021-02-15 08:44:47,038 - fba.levenshtein - INFO - Number of read pairs w/ valid barcodes: 93,795,979</span>
<span class="go">2021-02-15 08:44:47,153 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="matrix-generation">
<h2>Matrix generation<a class="headerlink" href="#matrix-generation" title="Permalink to this heading">#</a></h2>
<p>Only fragments with correct (passed the criteria) cell and feature
barcodes are included. UMI removal is powered by UMI-tools (<a class="reference external" href="http://www.genome.org/cgi/doi/10.1101/gr.209601.116">Smith, T.,
et al. 2017. Genome Res. 27, 491–499.</a>). Use <code class="docutils literal notranslate"><span class="pre">-us</span></code> to set the UMI
starting position on read 1 (default <code class="docutils literal notranslate"><span class="pre">16</span></code>). Use <code class="docutils literal notranslate"><span class="pre">-ul</span></code> to set the UMI
length (default <code class="docutils literal notranslate"><span class="pre">12</span></code>). Fragments with UMI length less than this value
are discarded. UMI deduplication method is set by <code class="docutils literal notranslate"><span class="pre">-ud</span></code> (default
<code class="docutils literal notranslate"><span class="pre">directional</span></code>). Use <code class="docutils literal notranslate"><span class="pre">-um</span></code> to set UMI deduplication mismatch
threshold (default <code class="docutils literal notranslate"><span class="pre">1</span></code>).</p>
<p>The generated feature count matrix can be easily imported into
well-established single cell analysis packages: <a class="reference external" href="https://satijalab.org/seurat/">Seruat</a> and <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba count <span class="se">\</span>
    -i feature_barcoding_output.tsv.gz <span class="se">\</span>
    -o matrix_featurecount.csv.gz <span class="se">\</span>
    -us <span class="m">16</span> <span class="se">\</span>
    -ul <span class="m">12</span> <span class="se">\</span>
    -um <span class="m">1</span> <span class="se">\</span>
    -ud directional
</pre></div>
</div>
<p>Result summary.</p>
<p>7.6% (7,145,799 out of 93,795,979) of read pairs with valid cell and
feature barcodes are unique fragments. 4.9% (7,143,943 out of
145,032,428) of total sequenced read pairs contribute to the final
matrix.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2020-10-20 04:47:32,738 - fba.__main__ - INFO - fba version: 0.0.7</span>
<span class="go">2020-10-20 04:47:32,738 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2020-10-20 04:47:32,738 - fba.__main__ - INFO - Python version: 3.7</span>
<span class="go">2020-10-20 04:47:32,738 - fba.__main__ - INFO - Using count subcommand ...</span>
<span class="go">2020-10-20 04:47:32,738 - fba.count - INFO - UMI-tools version: 1.0.1</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - UMI starting position on read 1: 16</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - UMI length: 12</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - UMI-tools deduplication threshold: 1</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - UMI-tools deduplication method: directional</span>
<span class="go">2020-10-20 04:47:32,795 - fba.count - INFO - Header line: read1_seq cell_barcode cb_num_mismatches read2_seq feature_barcode fb_num_mismatches</span>
<span class="go">2020-10-20 04:51:50,886 - fba.count - INFO - Number of lines processed: 93,795,979</span>
<span class="go">2020-10-20 04:51:50,893 - fba.count - INFO - Number of cell barcodes detected: 11,758</span>
<span class="go">2020-10-20 04:51:50,894 - fba.count - INFO - Number of features detected: 2</span>
<span class="go">2020-10-20 05:00:42,298 - fba.count - INFO - Total UMIs after deduplication: 7,145,799</span>
<span class="go">2020-10-20 05:00:42,320 - fba.count - INFO - Median number of UMIs per cell: 477.0</span>
<span class="go">2020-10-20 05:00:42,434 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="demultiplexing">
<h2>Demultiplexing<a class="headerlink" href="#demultiplexing" title="Permalink to this heading">#</a></h2>
<section id="negative-binomial-distribution">
<h3>Negative binomial distribution<a class="headerlink" href="#negative-binomial-distribution" title="Permalink to this heading">#</a></h3>
<p>Cells are classified based on the feature count matrix. Demultiplexing
method <code class="docutils literal notranslate"><span class="pre">1</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is implemented based on the method
described by <a class="reference external" href="https://doi.org/10.1186/s13059-018-1603-1">Stoeckius, M., et al. (2018)</a> with some modifications. A
cell identity matrix is generated in the output directory (set by
<code class="docutils literal notranslate"><span class="pre">--output_directory</span></code>, default <code class="docutils literal notranslate"><span class="pre">demultiplexed</span></code>): 0 means negative, 1
means positive. Use <code class="docutils literal notranslate"><span class="pre">-q</span></code> to set the quantile threshold for
demulitplexing. Set <code class="docutils literal notranslate"><span class="pre">-v</span></code> to create visualization plots.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    --output_directory demultiplexed <span class="se">\</span>
    -dm <span class="m">1</span> <span class="se">\</span>
    -q <span class="m">0</span>.75 <span class="se">\</span>
    -v
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed1.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed1.png" style="width: 700px;" /></a>
<p>t-SNE embedding of cells based on the abundance of features (sgRNAs, no
transcriptome information used). Colors indicate the sgRNA status for
each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed.webp"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed.webp" style="width: 500px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="gaussian-mixture-model">
<h3>Gaussian mixture model<a class="headerlink" href="#gaussian-mixture-model" title="Permalink to this heading">#</a></h3>
<p>The implementation of demultiplexing method <code class="docutils literal notranslate"><span class="pre">2</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is
inspired by the method described on <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/crispr">10x Genomics’ website</a>. Use <code class="docutils literal notranslate"><span class="pre">-p</span></code>
to set the probability threshold for demulitplexing (default <code class="docutils literal notranslate"><span class="pre">0.9</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    -dm <span class="m">2</span> <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-10-04 14:14:15,659 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2021-10-04 14:14:15,659 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-10-04 14:14:15,659 - fba.__main__ - INFO - Python version: 3.8</span>
<span class="go">2021-10-04 14:14:15,659 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2021-10-04 14:14:36,166 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;</span>
<span class="go">2021-10-04 14:14:36,166 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2021-10-04 14:14:36,166 - fba.demultiplex - INFO - Demultiplexing method: 2</span>
<span class="go">2021-10-04 14:14:36,166 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2021-10-04 14:14:36,167 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2021-10-04 14:14:36,167 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2021-10-04 14:14:36,167 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2021-10-04 14:14:37,875 - fba.demultiplex - INFO - Number of cells: 11,758</span>
<span class="go">2021-10-04 14:14:37,875 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2021-10-04 14:14:37,920 - fba.demultiplex - INFO - Number of features: 2 / 2 (after filtering / original in the matrix)</span>
<span class="go">2021-10-04 14:14:37,920 - fba.demultiplex - INFO - Features: NON_TARGET-1 RAB1A-2</span>
<span class="go">2021-10-04 14:14:37,920 - fba.demultiplex - INFO - Total UMIs: 7,145,799 / 7,145,799</span>
<span class="go">2021-10-04 14:14:37,942 - fba.demultiplex - INFO - Median number of UMIs per cell: 477.0 / 477.0</span>
<span class="go">2021-10-04 14:14:37,942 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2021-10-04 14:14:38,418 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2021-10-04 14:14:42,078 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2021-10-04 14:15:24,288 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_gm3.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_gm3.png" style="width: 700px;" /></a>
<p>t-SNE embedding of cells based on the abundance of features (sgRNAs, no
transcriptome information used). Colors indicate the sgRNA status for
each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed_gm.webp"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed_gm.webp" style="width: 500px;" /></a>
<p>UMI distribution and model fitting threshold:</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_gm_NON_TARGET-1_0.9.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_gm_NON_TARGET-1_0.9.webp" style="width: 800px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_gm_RAB1A-2_0.9.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_gm_RAB1A-2_0.9.webp" style="width: 800px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="poisson-gaussian-mixture-model">
<h3>Poisson-Gaussian mixture model<a class="headerlink" href="#poisson-gaussian-mixture-model" title="Permalink to this heading">#</a></h3>
<p>The implementation of demultiplexing method <code class="docutils literal notranslate"><span class="pre">3</span></code> (set by <code class="docutils literal notranslate"><span class="pre">-dm</span></code>) is
inspired by <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2021.12.16.473013">Replogle, M., et al. (2021)</a>. Use <code class="docutils literal notranslate"><span class="pre">-p</span></code> to set the
probability threshold for demulitplexing (default <code class="docutils literal notranslate"><span class="pre">0.5</span></code>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    -dm <span class="m">3</span> <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-12-20 00:13:17,443 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2021-12-20 00:13:17,443 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-12-20 00:13:17,443 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2021-12-20 00:13:17,443 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2021-12-20 00:13:19,774 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;</span>
<span class="go">2021-12-20 00:13:19,774 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2021-12-20 00:13:19,774 - fba.demultiplex - INFO - Demultiplexing method: 3</span>
<span class="go">2021-12-20 00:13:19,774 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2021-12-20 00:13:19,774 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2021-12-20 00:13:19,774 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2021-12-20 00:13:19,774 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2021-12-20 00:13:20,479 - fba.demultiplex - INFO - Number of cells: 11,758</span>
<span class="go">2021-12-20 00:13:20,479 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2021-12-20 00:13:20,497 - fba.demultiplex - INFO - Number of features: 2 / 2 (after filtering / original in the matrix)</span>
<span class="go">2021-12-20 00:13:20,497 - fba.demultiplex - INFO - Features: NON_TARGET-1 RAB1A-2</span>
<span class="go">2021-12-20 00:13:20,497 - fba.demultiplex - INFO - Total UMIs: 7,145,799 / 7,145,799</span>
<span class="go">2021-12-20 00:13:20,506 - fba.demultiplex - INFO - Median number of UMIs per cell: 477.0 / 477.0</span>
<span class="go">2021-12-20 00:13:20,506 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2021-12-20 00:13:21,930 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2021-12-20 00:13:23,070 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2021-12-20 00:13:41,271 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_pgm1.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_pgm1.png" style="width: 700px;" /></a>
<p>t-SNE embedding of cells based on the abundance of features (sgRNAs, no
transcriptome information used). Colors indicate the sgRNA status for
each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed_pgm1.webp"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed_pgm1.webp" style="width: 500px;" /></a>
<p>UMI distribution and model fitting threshold:</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_pgm_NON_TARGET-1_0.5.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_pgm_NON_TARGET-1_0.5.webp" style="width: 800px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_pgm_RAB1A-2_0.5.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_pgm_RAB1A-2_0.5.webp" style="width: 800px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="kernel-density-estimation">
<h3>Kernel density estimation<a class="headerlink" href="#kernel-density-estimation" title="Permalink to this heading">#</a></h3>
<p>CRISPR perturbatons are demultiplexed based on the abundance of
features. Demultiplexing method <code class="docutils literal notranslate"><span class="pre">4</span></code> is implemented based on the method
described in <a class="reference external" href="https://doi.org/10.1038/s41592-019-0433-8">McGinnis, C., et al. (2019)</a> with some modifications.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    -dm <span class="m">4</span> <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-12-26 18:11:16,685 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2021-12-26 18:11:16,685 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-12-26 18:11:16,686 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2021-12-26 18:11:16,686 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2021-12-26 18:11:19,633 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;, &quot;-p/--prob&quot;</span>
<span class="go">2021-12-26 18:11:19,633 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2021-12-26 18:11:19,633 - fba.demultiplex - INFO - Demultiplexing method: 4</span>
<span class="go">2021-12-26 18:11:19,633 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2021-12-26 18:11:19,633 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2021-12-26 18:11:19,633 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2021-12-26 18:11:19,633 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2021-12-26 18:11:19,745 - fba.demultiplex - INFO - Number of cells: 11,758</span>
<span class="go">2021-12-26 18:11:19,745 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2021-12-26 18:11:19,762 - fba.demultiplex - INFO - Number of features: 2 / 2 (after filtering / original in the matrix)</span>
<span class="go">2021-12-26 18:11:19,762 - fba.demultiplex - INFO - Features: NON_TARGET-1 RAB1A-2</span>
<span class="go">2021-12-26 18:11:19,762 - fba.demultiplex - INFO - Total UMIs: 7,145,799 / 7,145,799</span>
<span class="go">2021-12-26 18:11:19,771 - fba.demultiplex - INFO - Median number of UMIs per cell: 477.0 / 477.0</span>
<span class="go">2021-12-26 18:11:19,771 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2021-12-26 18:11:22,049 - fba.demultiplex - INFO - Quantile cutoff: 18</span>
<span class="go">2021-12-26 18:11:23,703 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2021-12-26 18:11:24,911 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2021-12-26 18:11:44,219 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_kde1.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_kde1.png" style="width: 700px;" /></a>
<p>t-SNE embedding of cells based on the abundance of features (sgRNAs, no
transcriptome information used). Colors indicate the sgRNA status for
each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed_kde.webp"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed_kde.webp" style="width: 500px;" /></a>
<p>UMI distribution and model fitting threshold:</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_kde_NON_TARGET-1_18.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_kde_NON_TARGET-1_18.webp" style="width: 800px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_kde_RAB1A-2_18.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_kde_RAB1A-2_18.webp" style="width: 800px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="knee-point">
<h3>Knee point<a class="headerlink" href="#knee-point" title="Permalink to this heading">#</a></h3>
<section id="method-1">
<h4>Method 1<a class="headerlink" href="#method-1" title="Permalink to this heading">#</a></h4>
<p>Cells are demultiplexed based on the abundance of features (sgRNAs).
Demultiplexing method <code class="docutils literal notranslate"><span class="pre">5-2019</span></code> is our previous implementation, which
tries to determine perturbations in the cells through the detection of
inflection point on the feature UMI saturation curve (<a class="reference external" href="https://doi.org/10.1016/j.celrep.2019.10.073">Xie, S., et al.
(2019)</a>).</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    -dm <span class="m">5</span>-2019 <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2022-01-02 13:57:51,792 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2022-01-02 13:57:51,792 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2022-01-02 13:57:51,792 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2022-01-02 13:57:51,792 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2022-01-02 13:57:54,328 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;, &quot;-p/--prob&quot;</span>
<span class="go">2022-01-02 13:57:54,329 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2022-01-02 13:57:54,329 - fba.demultiplex - INFO - Demultiplexing method: 5-2019</span>
<span class="go">2022-01-02 13:57:54,329 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2022-01-02 13:57:54,329 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2022-01-02 13:57:54,329 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2022-01-02 13:57:54,329 - fba.demultiplex - INFO - Loading feature count matrix: raw/m2_2020-10-20/matrix_featurecount.csv.gz ...</span>
<span class="go">2022-01-02 13:57:54,444 - fba.demultiplex - INFO - Number of cells: 11,758</span>
<span class="go">2022-01-02 13:57:54,444 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2022-01-02 13:57:54,464 - fba.demultiplex - INFO - Number of features: 2 / 2 (after filtering / original in the matrix)</span>
<span class="go">2022-01-02 13:57:54,464 - fba.demultiplex - INFO - Features: NON_TARGET-1 RAB1A-2</span>
<span class="go">2022-01-02 13:57:54,464 - fba.demultiplex - INFO - Total UMIs: 7,145,799 / 7,145,799</span>
<span class="go">2022-01-02 13:57:54,474 - fba.demultiplex - INFO - Median number of UMIs per cell: 477.0 / 477.0</span>
<span class="go">2022-01-02 13:57:54,474 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2022-01-02 13:57:55,509 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2022-01-02 13:57:56,717 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2022-01-02 13:58:12,014 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_knee-2019.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_knee-2019.png" style="width: 700px;" /></a>
<p>t-SNE embedding of cells based on the abundance of features (sgRNAs, no
transcriptome information used). Colors indicate the sgRNA status for
each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed_knee-2019.webp"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed_knee-2019.webp" style="width: 500px;" /></a>
<p>UMI distribution and model fitting threshold:</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_knee-2019_NON_TARGET-1.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_knee-2019_NON_TARGET-1.webp" style="width: 400px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_knee-2019_RAB1A-2.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_knee-2019_RAB1A-2.webp" style="width: 400px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="method-2">
<h4>Method 2<a class="headerlink" href="#method-2" title="Permalink to this heading">#</a></h4>
<p>Cells are demultiplexed based on the abundance of features (sgRNAs).
Demultiplexing method <code class="docutils literal notranslate"><span class="pre">5</span></code> is implemented to use the local maxima on
the difference curve to detemine the knee point on the UMI saturation
curve.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>fba demultiplex <span class="se">\</span>
    -i matrix_featurecount.csv.gz <span class="se">\</span>
    -dm <span class="m">5</span> <span class="se">\</span>
    -v
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">2021-12-29 22:55:34,303 - fba.__main__ - INFO - fba version: 0.0.x</span>
<span class="go">2021-12-29 22:55:34,303 - fba.__main__ - INFO - Initiating logging ...</span>
<span class="go">2021-12-29 22:55:34,303 - fba.__main__ - INFO - Python version: 3.9</span>
<span class="go">2021-12-29 22:55:34,303 - fba.__main__ - INFO - Using demultiplex subcommand ...</span>
<span class="go">2021-12-29 22:55:36,774 - fba.__main__ - INFO - Skipping arguments: &quot;-q/--quantile&quot;, &quot;-cm/--clustering_method&quot;, &quot;-p/--prob&quot;</span>
<span class="go">2021-12-29 22:55:36,774 - fba.demultiplex - INFO - Output directory: demultiplexed</span>
<span class="go">2021-12-29 22:55:36,774 - fba.demultiplex - INFO - Demultiplexing method: 5</span>
<span class="go">2021-12-29 22:55:36,774 - fba.demultiplex - INFO - UMI normalization method: clr</span>
<span class="go">2021-12-29 22:55:36,774 - fba.demultiplex - INFO - Visualization: On</span>
<span class="go">2021-12-29 22:55:36,774 - fba.demultiplex - INFO - Visualization method: tsne</span>
<span class="go">2021-12-29 22:55:36,774 - fba.demultiplex - INFO - Loading feature count matrix: matrix_featurecount.csv.gz ...</span>
<span class="go">2021-12-29 22:55:36,886 - fba.demultiplex - INFO - Number of cells: 11,758</span>
<span class="go">2021-12-29 22:55:36,886 - fba.demultiplex - INFO - Number of positive cells for a feature to be included: 200</span>
<span class="go">2021-12-29 22:55:36,904 - fba.demultiplex - INFO - Number of features: 2 / 2 (after filtering / original in the matrix)</span>
<span class="go">2021-12-29 22:55:36,904 - fba.demultiplex - INFO - Features: NON_TARGET-1 RAB1A-2</span>
<span class="go">2021-12-29 22:55:36,904 - fba.demultiplex - INFO - Total UMIs: 7,145,799 / 7,145,799</span>
<span class="go">2021-12-29 22:55:36,913 - fba.demultiplex - INFO - Median number of UMIs per cell: 477.0 / 477.0</span>
<span class="go">2021-12-29 22:55:36,913 - fba.demultiplex - INFO - Demultiplexing ...</span>
<span class="go">2021-12-29 22:55:37,415 - fba.demultiplex - INFO - Generating heatmap ...</span>
<span class="go">2021-12-29 22:55:38,576 - fba.demultiplex - INFO - Embedding ...</span>
<span class="go">2021-12-29 22:55:57,485 - fba.__main__ - INFO - Done.</span>
</pre></div>
</div>
<p>Heatmap of the relative abundance of features (sgRNAs) across all cells.
Each column represents a single cell.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_heatmap_cells_demultiplexed_knee1.png"><img alt="Heatmap" class="align-center" src="../../../_images/Pyplot_heatmap_cells_demultiplexed_knee1.png" style="width: 700px;" /></a>
<p>t-SNE embedding of cells based on the abundance of features (sgRNAs, no
transcriptome information used). Colors indicate the sgRNA status for
each cell, as called by FBA.</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_embedding_cells_demultiplexed_knee.webp"><img alt="t-SNE embedding" class="align-center" src="../../../_images/Pyplot_embedding_cells_demultiplexed_knee.webp" style="width: 500px;" /></a>
<p>UMI distribution and model fitting threshold:</p>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_knee_NON_TARGET-1.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_knee_NON_TARGET-1.webp" style="width: 400px;" /></a>
<a class="reference internal image-reference" href="../../../_images/Pyplot_feature_umi_distribution_knee_RAB1A-2.webp"><img alt="UMI distribution" class="align-center" src="../../../_images/Pyplot_feature_umi_distribution_knee_RAB1A-2.webp" style="width: 400px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../PRJNA358686/tutorial.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">CROP-seq; 1:1:1 Mixture of DNMT3B, MBD1, and TET2 Knockout Cell Lines (HEK293T)</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">CRISPR screening</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2020-2023, Jialei Duan
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/jlduan/fba" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">10k A375 Cells Transduced with (1) Non-Target and (1) Target sgRNA</a><ul>
<li><a class="reference internal" href="#preparation">Preparation</a></li>
<li><a class="reference internal" href="#qc">QC</a></li>
<li><a class="reference internal" href="#barcode-extraction">Barcode extraction</a></li>
<li><a class="reference internal" href="#matrix-generation">Matrix generation</a></li>
<li><a class="reference internal" href="#demultiplexing">Demultiplexing</a><ul>
<li><a class="reference internal" href="#negative-binomial-distribution">Negative binomial distribution</a></li>
<li><a class="reference internal" href="#gaussian-mixture-model">Gaussian mixture model</a></li>
<li><a class="reference internal" href="#poisson-gaussian-mixture-model">Poisson-Gaussian mixture model</a></li>
<li><a class="reference internal" href="#kernel-density-estimation">Kernel density estimation</a></li>
<li><a class="reference internal" href="#knee-point">Knee point</a><ul>
<li><a class="reference internal" href="#method-1">Method 1</a></li>
<li><a class="reference internal" href="#method-2">Method 2</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/scripts/furo.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    </body>
</html>